---
title: "Reclustering Analysis of Cancer-Associated Fibroblast Using SCISSORS"
subtitle: "Jack Leary"
author: 
  - "University of North Carolina at Chapel Hill - Lineberger Comprehensive Cancer Center"
  - "University of Florida - Department of Biostatistics"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: paper
    highlight: tango
    df_print: kable
    toc: true
    toc_float: true
    code_folding: show
    code_download: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE, 
                      warning = FALSE, 
                      results = "hold", 
                      fig.align = "center")
reticulate::use_virtualenv("~/Desktop/Python/science/venv/", required = TRUE)
set.seed(629)
```

# Introduction

In 2017, the Tuveson lab at Cold Spring Harbor Cancer Center published a paper written by Elyada *et al* (2017) that detailed the discovery of cancer-associated fibroblasts (CAFs) in mice. The subtypes were then validated in human samples affected with PDAC in a subsequent paper released in 2019. Here we will use SCISSORS to identify the CAF subtypes within the larger fibroblast population.

# Libraries

## R

```{r}
library(pals)        # high N discrete color palettes
library(VAM)         # single cell GSEA
library(dplyr)       # tidy data 
library(Seurat)      # single cell infrastructure
library(ggplot2)     # plots
library(SingleR)     # cell type assignment
library(decoderr)    # de novo deconvolution 
library(SCISSORS)    # our package
library(mixtools)    # Gaussian mixture model estimation
library(patchwork)   # plot grids
library(paletteer)   # more color palettes
library(CONICSmat)   # CNV estimation
library(reticulate)  # Python interface
library(wesanderson) # even more color palettes
```

## Python

```{python}
import numpy as np
from openTSNE import TSNEEmbedding
from openTSNE import initialization
from openTSNE.affinity import Multiscale
from openTSNE.affinity import PerplexityBasedNN
```

# Data

First we load in the $\text{gene} \times \text{cell}$ counts matrix, then create a `Seurat` object to hold it in. Next, we add samplename, tissue type, and patient sex metadata taken from the publicly available dataset.

```{r}
raw_counts <- Read10X(data.dir = "~/Desktop/Data/Elyada Raw/All Human/")
pdac <- CreateSeuratObject(raw_counts, 
                           project = "Elyada", 
                           min.cells = 3, 
                           min.features = 200)
pdac@meta.data$sample <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "SRR9274536", 
                                   grepl("-2", rownames(pdac@meta.data)) ~ "SRR9274537", 
                                   grepl("-3", rownames(pdac@meta.data)) ~ "SRR9274538", 
                                   grepl("-4", rownames(pdac@meta.data)) ~ "SRR9274539",
                                   grepl("-5", rownames(pdac@meta.data)) ~ "SRR9274540", 
                                   grepl("-6", rownames(pdac@meta.data)) ~ "SRR9274541", 
                                   grepl("-7", rownames(pdac@meta.data)) ~ "SRR9274542", 
                                   grepl("-8", rownames(pdac@meta.data)) ~ "SRR9274543", 
                                   grepl("-9", rownames(pdac@meta.data)) ~ "SRR9274544")
pdac@meta.data$condition <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-2", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-3", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-4", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-5", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-6", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-7", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-8", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-9", rownames(pdac@meta.data)) ~ "PDAC")
pdac@meta.data$sex <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-2", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-3", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-4", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-5", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-6", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-7", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-8", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-9", rownames(pdac@meta.data)) ~ "female")
```

```{r, echo=FALSE}
rm(raw_counts)
```

# Preprocessing

We run the typical single cell pre-processing steps on our cells - normalization, dimension reduction, and clustering. The t-SNE embedding looks decent, but could definitely be improved. Globally, the clusters are arranged in a blob - which isn't very informative - though the local structure seems to have been preserved fairly well. Clusters 5 and 10 are split in two, which we would prefer not to have happen. 

```{r, warning=FALSE}
pdac <- PrepareData(pdac, 
                    n.variable.genes = 4000, 
                    n.PC = 20, 
                    which.dim.reduc = "tsne", 
                    initial.resolution = .5, 
                    random.seed = 629)
p0 <- DimPlot(pdac, reduction = "bh_tsne") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "t-SNE 1", y = "t-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p0
```

## Optimize Dimension Reduction

I think the two-dimensional visualization of the cells could be improved. We'll try using UMAP and the Fast Fourier Transform-accelerated Fit-SNE (as implemented in the `openTSNE` library) to improve the embedding.

### UMAP

It seems like UMAP does a good job of clearly separating our clusters and preserving the global structure of the data. However, clusters 5 and 7 are separated, and it's difficult to see local structure within some of the clusters due to their density.

```{r}
pdac <- RunUMAP(pdac, 
                reduction = "pca", 
                dims = 1:20, 
                umap.method = "uwot", 
                n.components = 2, 
                n.epochs = 750, 
                n.neighbors = 50, 
                metric = "cosine", 
                seed.use = 629, 
                verbose = FALSE)
p1 <- DimPlot(pdac, reduction = "umap") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "UMAP 1", y = "UMAP 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p1
```

### Fit-SNE

For information on how to install the `openTSNE` implementation of Fit-SNE and how to run the algorithm, please visit [the excellent GitHub repository of Pavlin Policar](https://github.com/pavlin-policar/openTSNE).

First we'll run a simple, standard Fit-SNE embedding. It's necessary to make the PCA embeddings accessible by Python.

```{r}
pc_df <- Embeddings(pdac, reduction = "pca")
```

```{python}
# import data
pc_df = np.array(r.pc_df)
# run Fit-SNE
affin = PerplexityBasedNN(pc_df, perplexity=50, metric='cosine', random_state=629)
init = initialization.pca(pc_df, random_state=629)
tsne1 = TSNEEmbedding(init, affin, negative_gradient_method='fft')
embed1 = tsne1.optimize(n_iter=350, exaggeration=12, momentum=0.6) 
embed2 = embed1.optimize(n_iter=1000, momentum=0.8)
```

The embedding looks good, but the subclusters within clusters 5 and 10 are still clearly separated, and a small part of cluster 7 is far from the rest of the cells in the cluster.

```{r}
embed <- as.matrix(py$embed2)
rownames(embed) <- colnames(pdac)
pdac@reductions$fitsne <- CreateDimReducObject(embeddings = embed, 
                                               key = "FitSNE_", 
                                               assay = "SCT", 
                                               global = TRUE)
p2 <- DimPlot(pdac, reduction = "fitsne") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p2
```

We use the ability of Fit-SNE to consider multiple perplexities in an attempt to provide a more accurate representation of both local and global structure in our data.

```{python}
affin_anneal = PerplexityBasedNN(pc_df, perplexity=150, metric='cosine', random_state=629)
tsne2 = TSNEEmbedding(init, affin_anneal, negative_gradient_method='fft')
embed3 = tsne2.optimize(n_iter=350, exaggeration=12, momentum=0.5)
embed4 = embed1.optimize(n_iter=1000, exaggeration=1, momentum=0.8)
affin_anneal.set_perplexity(30)
embed5 = embed4.optimize(n_iter=500, momentum=0.8)
```

The perplexity annealing hasn't really improved or even changed the embedding much at all - cluster 4 is still scattered over the two-dimensional space, and the outlier subgroup in cluster 7 hasn't moved any closer to the rest of the cluster. 

```{r}
embed_multi <- as.matrix(py$embed5)
rownames(embed_multi) <- colnames(pdac)
pdac@reductions$fitsne_multi <- CreateDimReducObject(embeddings = embed_multi, 
                                                     key = "FitSNE_", 
                                                     assay = "SCT", 
                                                     global = TRUE)
p3 <- DimPlot(pdac, reduction = "fitsne_multi") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p3
```

Finally, we'll use a multiscale kernel to represent the data in the high-dimensional space instead of the default Gaussian. This in tandem with the multiple perplexity values will preserve as much as possible of both local and global structure.

```{python}
affin_multi = Multiscale(pc_df, perplexities=[30, 100], metric='cosine', random_state=629)
tsne3 = TSNEEmbedding(init, affin_multi, negative_gradient_method='fft')
embed6 = tsne3.optimize(n_iter=450, exaggeration=12, momentum=0.6)
embed7 = embed6.optimize(n_iter=1200, exaggeration=1, momentum=0.8)
```

The multiscale trick doesn't really fix the embedding issues either, but it does clean up a bit of the noise in the plot, and it does place the two subclusters in cluster 10 closer together. Even though clusters 5 & 7 are still split up, we'll use this embedding going forward and make sure to examine those clusters for subgroups using `SCISSORS`.

```{r}
embed_kern <- as.matrix(py$embed7)
rownames(embed_kern) <- colnames(pdac)
pdac@reductions$fitsne_kern<- CreateDimReducObject(embeddings = embed_kern, 
                                                   key = "FitSNE_", 
                                                   assay = "SCT", 
                                                   global = TRUE)

p4 <- DimPlot(pdac, reduction = "fitsne_kern") +
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p4
```

Looking at the embedding, we can see that clusters 3, 4, 5, 6, 7, and 10 contain clear small subclusters. We'll be sure to investigate each of these in turn, and annotate the groups using marker genes.

Due to the way `Seurat` accesses cell embeddings, we'll need to replace our original t-SNE dimension reduction in our `Seurat` object with the new Fit-SNE version. We'll keep the original Barnes-Hut t-SNE embedding under a separate name.

```{r}
pdac@reductions$bh_tsne <- pdac@reductions$tsne
pdac@reductions$tsne <- pdac@reductions$fitsne_kern
```

```{r, echo=FALSE}
rm(embed, embed_kern, embed_multi, pc_df)
```

# SingleR Cell Type Identification

## Single Cell RNA-seq Reference Data

Here we use the `SingleR` package to identify broad cell types. The reference dataset we load is an `sctransform`-normalized version of the raw counts available in `scRNAseq::BaronPancreasData()`, which consists of normal pancreas cells that were sequenced and annotated by the researchers.

```{r}
sc_ref <- readRDS("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/Seurat/single_cell_ref_normalized.Rds")
sc_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                    ref = sc_ref, 
                    labels = sc_ref$label, 
                    method = "cluster", 
                    clusters = pdac$seurat_clusters, 
                    de.method = "wilcox")
pdac[["SingleR.labels.sc"]] <- sc_preds$labels[match(pdac[[]][["seurat_clusters"]], 
                                                     rownames(sc_preds))]
pdac$SingleR.labels.sc <- case_when(pdac$SingleR.labels.sc == "acinar" ~ "Acinar", 
                                    pdac$SingleR.labels.sc == "activated_stellate" ~ "Activated Stellate",
                                    pdac$SingleR.labels.sc == "ductal" ~ "Ductal", 
                                    pdac$SingleR.labels.sc == "macrophage" ~ "Macrophage",
                                    pdac$SingleR.labels.sc == "t_cell" ~ "T",)
```

We can see that there's a large immune population, as well as smaller ductal, fibroblast (denoted activated stellate in the reference dataset), and acinar groups. These broad cell types line up with what we expected to see given the authors' original cell cluster annotations. 

```{r}
p5 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.sc") + 
      scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p5
```

## Bulk Tissue RNA-seq Reference Data

This dataset is composed of labeled & log-normalized bulk RNA-seq samples from the Human Primary Cell Atlas.

```{r}
bulk_ref <- HumanPrimaryCellAtlasData()
bulk_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                      ref = bulk_ref, 
                      labels = bulk_ref$label.main, 
                      method = "cluster", 
                      clusters = pdac$seurat_clusters, 
                      de.method = "wilcox")
pdac[["SingleR.labels.bulk"]] <- bulk_preds$labels[match(pdac[[]][["seurat_clusters"]], 
                                                         rownames(bulk_preds))]
pdac$SingleR.labels.bulk <- case_when(pdac$SingleR.labels.bulk == "B_cell" ~ "B", 
                                      pdac$SingleR.labels.bulk == "Epithelial_cells" ~ "Epithelial", 
                                      pdac$SingleR.labels.bulk == "Pre-B_cell_CD34-" ~ "Pre-B CD34-", 
                                      pdac$SingleR.labels.bulk == "T_cells" ~ "T", 
                                      TRUE ~ pdac$SingleR.labels.bulk)
```

The bulk reference gives us somewhat more granular labels for the immune cells, and confirms the identities of the ductal / epithelial and fibroblast (denoted here as mesenchymal stem cells) clusters. 

```{r}
p6 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.bulk") + 
      scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p6
```

```{r, echo=FALSE}
rm(sc_preds, bulk_preds, bulk_ref, sc_ref)
```

One shouldn't use `SingleR` as the final authority for cell types, but we were able to confirm the identities of the ductal and fibroblast clusters, which is important for this dataset as the cells we are most interested in are the cancer-associated fibroblasts (CAFs). 

One interesting label was that of cluster 5, which was labeled as containing macrophages by the scRNA-seq reference, and as pre-B CD34- by the bulk RNA-seq reference. It is split in half in Fit-SNE space, with half of the cells adjoined to T cells and the other half touching macrophage and B cell clusters. I think I'll probably lump cluster 

# CONICSmat CNV Estimation

Next we'll attempt to identify malignant cells using single-cell copy number variation estimation as implemented in the `CONCISmat` package. Details of the GMM methodology used can be found [at the Diaz Lab's GitHub repository](https://github.com/diazlab/CONICS). Note: this step is memory-intensive because 1) it requires the sparse counts matrix to be cast to a dense matrix and 2) a lot of Gaussian mixture models get estimated. If your machine doesn't have a lot of RAM it might be best to skip this and manually annotate the malignant cells instead through the usage of known canonical markers or the `VAM` single-cell GSEA methodology. 

```{r}
chrom_regions <- read.table("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/chrom_arm_positions.txt", 
                            sep = "\t", 
                            row.names = 1, 
                            header = TRUE)
gene_pos <- getGenePositions(rownames(pdac))
cpm <- t(t(as.matrix(pdac@assays$SCT@counts)) / colSums(as.matrix(pdac@assays$SCT@counts))) * 10^5
cpm <- log2(cpm + 1)
norm_factor <- calcNormFactors(cpm)
cnv_est <- plotAll(mat = cpm, 
                   normFactor = norm_factor, 
                   regions = chrom_regions, 
                   gene_pos = gene_pos, 
                   fname = "Elyada_CNVs")
```

## Visualizing CNVs

After estimating CNVs, we cluster the cells into $k = 3$ clusters, with the hope of finding one large cluster of normal cells and two smaller clusters composed of CAFs and PDAC cells. 

```{r}
bic_table <- read.table("./Elyada_CNVs_BIC_LR.txt", 
                        sep = "\t", 
                        row.names = 1, 
                        header = TRUE, 
                        check.names = FALSE)
cand_regions <- rownames(bic_table[bic_table$`BIC difference` > 1000 & bic_table$`LRT adj. p-val` < .01, ])
hist1 <- plotHistogram(cnv_est[, cand_regions], 
                       cpm, 
                       clusters = 3, 
                       zscoreThreshold = 3, 
                       patients = pdac$sample)
```

We add the normal vs. malignant cell labels in to our `Seurat` object's metadata, then visualize the results. As expected, the malignant cells are located in the clusters identified by `SingleR` as ductal cells and fibroblasts. This indicates that `CONICSmat` did a solid job of estimating the CNVs - no easy feat with sparse, noisy single-cell data. 

```{r}
normal <- which(hist1 == 1)
malignant <- which(hist1 != 1)
pdac@meta.data$malig <- ifelse(rownames(pdac@meta.data) %in% names(normal), "Normal", "Malignant")
p7 <- DimPlot(pdac, reduction = "tsne", group.by = "malig") + 
      scale_color_manual(values = wes_palette("Zissou1", n = 5)[c(5, 2)]) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p7
```

```{r, echo=FALSE}
rm(cpm, bic_table, chrom_regions, cnv_est, gene_pos, cand_regions, hist1, normal, norm_factor, malignant)
gc()  # not recommended, but it could help this document compile, so ...
```

# DECODER

Next, we use [Dr. Xianlu Peng's DECODER](https://github.com/laurapeng/decoderr) in order to deconvolve the dataset and assign weights to each cell using non-negative matrix factorization. We calculate basal & classical PDAC, normal & activated stroma, immune, and endocrine & exocrine pancreas compartment weights.

```{r}
sample_wts_unscaled <- Decon_single_sample("TCGA_RNAseq_PAAD", pdac@assays$SCT@data, "geneSymbol")
sample_wts <- Norm_PDAC_weights(sample_wts_unscaled)
pdac <- AddMetaData(pdac, sample_wts$Immune, "immune")
pdac <- AddMetaData(pdac, sample_wts$bcRatio, "bc_ratio")
pdac <- AddMetaData(pdac, sample_wts$Exocrine, "exocrine")
pdac <- AddMetaData(pdac, sample_wts$Endocrine, "endocrine")
pdac <- AddMetaData(pdac, sample_wts_unscaled[, 9], "basal")
pdac <- AddMetaData(pdac, sample_wts_unscaled[, 5], "classical")
pdac <- AddMetaData(pdac, sample_wts$NormalStroma, "norm_stroma")
pdac <- AddMetaData(pdac, sample_wts$ActivatedStroma, "act_stroma")
```

## Visualizing DECODER Weights

### Basal PDAC

The basal weights are highest in a subcluster of the ductal group identified through `SingleR`. This is interesting as the authors did not find evidence of the basal subtype in their paper. 

```{r}
p8 <- FeaturePlot(pdac, reduction = "tsne", features = "basal") + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p8
```

### Classical PDAC

The classical weights are highest in another subcluster of the ductal cluster, and cells with high classical weights do not collocate with those that have high basal weights. The putative classical and basal PDAC cells also align closely with the cells identified through `CONCISmat` as malignant. 

```{r}
p9 <- FeaturePlot(pdac, reduction = "tsne", features = "classical") + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p9
```

### Exocrine Pancreas

The cluster identified through `SingleR` as acinar cells is the only cluster with high exocrine pancreas weights. 

```{r}
p10 <- FeaturePlot(pdac, reduction = "tsne", features = "exocrine") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p10
```

### Endocrine Pancreas

No cells have high endocrine pancreas weights. 

```{r}
p11 <- FeaturePlot(pdac, reduction = "tsne", features = "endocrine") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p11
```

### Immune

Once again, we confirm the largeness of the immune cell population in this dataset. 

```{r}
p12 <- FeaturePlot(pdac, reduction = "tsne", features = "immune") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p12
```

### Normal Stroma

Cells with high normal stroma stroma weights are located in the cluster identified by `SingleR` as being composed of fibroblasts. 

```{r}
p13 <- FeaturePlot(pdac, reduction = "tsne", features = "norm_stroma") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p13
```

### Activated Stroma

Cells with high activated stroma weights are also located in the fibroblast cluster, but do not intersect with the cells that have high normal stroma scores. This indicates that `SCISSORS` will most likely perform well on the fibroblast cluster and be able to quickly tease out the cell subtypes.

```{r}
p14 <- FeaturePlot(pdac, reduction = "tsne", features = "act_stroma") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p14
```

# SCISSORS

Now that we have rough labels from `SingleR`, CNVs from `CONICSmat`, and compartment weights from `DECODER`, we should have more than enough cell-level metadata to look for and annotate cell subtypes using `SCISSORS`. 

## Fibroblasts

The fibroblast marker genes provided by @elyada2017 match the `SingleR` results defining cluster 7 as containing fibroblasts. Interestingly, the outlier subcluster within cluster 7 does not express any of the fibroblast markers. 

```{r}
p15 <- FeaturePlot(pdac, reduction = "tsne", features = "COL1A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL1A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p16 <- FeaturePlot(pdac, reduction = "tsne", features = "COL3A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL3A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p17 <- FeaturePlot(pdac, reduction = "tsne", features = "LUM") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "LUM") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p18 <- FeaturePlot(pdac, reduction = "tsne", features = "DCN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "DCN") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p15 | p16) / (p17 | p18)
```

### Reclustering

Here we use `ReclusterCells()` to identify subclusters within cluster 7. We find five distinct subclusters, which are clearly visible on the t-SNE re-embedding.

```{r, warning=FALSE}
fibro <- ReclusterCells(pdac, 
                        which.clust = 7,
                        n.variable.genes = 4000, 
                        n.PC = 20, 
                        resolution.vals = c(.1, .2, .3, .4), 
                        k.vals = c(5, 10, 20, 30), 
                        which.dim.reduc = "tsne", 
                        redo.embedding = TRUE, 
                        do.plot = FALSE)
fibro_pc <- Embeddings(fibro, "pca")
```

We'll again run Fit-SNE on the reclustered cells, for consistencies sake. 

```{python}
# import data
fibro_pc = r.fibro_pc
# run Fit-SNE
affin_fibro = PerplexityBasedNN(fibro_pc, perplexity=40, metric='cosine', random_state=629)
init = initialization.pca(fibro_pc, random_state=629)
tsne_f = TSNEEmbedding(init, affin_fibro, negative_gradient_method='fft')
embed_f1 = tsne_f.optimize(n_iter=250, exaggeration=12, momentum=0.6) 
embed_f2 = embed_f1.optimize(n_iter=750, momentum=0.8)
```

Pulling the results back into R and visualizing them shows clear separation between the subclusters. There's some noise in subcluster 0, but other than that the reembedding looks solid. 

```{r}
embed_fibro <- as.matrix(py$embed_f2)
rownames(embed_fibro) <- colnames(fibro)
fibro@reductions$bh_tsne <- fibro@reductions$tsne
fibro@reductions$tsne<- CreateDimReducObject(embeddings = embed_fibro, 
                                             key = "FitSNE_", 
                                             assay = "SCT", 
                                             global = TRUE)
p19 <- DimPlot(fibro, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p19
```

### Cell Type Identification

Next, we use the `VAM` method of single cell gene set enrichment analysis to determine which clusters are enriched for the iCAF and myCAF marker genes, as well as the general pan-CAF marker set. We use the marker genes identified by the authors. 

```{r}
icaf_genes <- c("IL6", "PDGFRA", "CXCL12", "CFD", "LMNA", 
                "AGTR1", "HAS1", "CXCL1", "CXCL2", "CCL2", "IL8")
mycaf_genes <- c("ACTA2", "TAGLN", "MMP11", "MYL9", 
                 "HOPX", "POSTN", "TPM1", "TPM2")
pan_caf_genes <- c("COL1A1", "FAP", "PDPN", "DCN", "VIM")
gene_sets <- list(icaf_genes, mycaf_genes, pan_caf_genes)
names(gene_sets) <- c("iCAF", "myCAF", "Pan-CAF")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
```

#### Pan-CAF 

First, we examine Pan-CAF gene set enrichment. We can see that only clusters 0 and 2 are highly enriched for the CAF marker genes. This indicates that the other three clusters are composed of cells that are not fibroblasts, but that have similar gene expression patterns to fibroblasts.

```{r}
p20 <- FeaturePlot(fibro, reduction = "tsne", features = "Pan-CAF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "Pan-CAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p20 / p19
```

#### iCAFs & myCAFs

We can easily define cluster 0 as the myCAF population, and cluster 2 as the slightly smaller iCAF population. Finally, though cluster 1 shows some enrichment for both pathways, it doesn't have a consistent enough pattern to label it another CAF cluster.

```{r}
p21 <- FeaturePlot(fibro, reduction = "tsne", features = "iCAF") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "iCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p22 <- FeaturePlot(fibro, reduction = "tsne", features = "myCAF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "myCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p21 | p22) / p19
```

#### Perivascular Cells

By examining the marker genes provided in the Elyada paper, we define cluster 1 as being composed of perivascular cells. This intuitively makes sense, as the perivascular cells were positioned near to the fibroblast cells in the main t-SNE figure from @elyada2017.

```{r}
DefaultAssay(fibro) <- "SCT"
p23 <- FeaturePlot(fibro, reduction = "tsne", features = "IGFBP7") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "IGFBP7") + 
       theme_yehlab() + 
       NoLegend() +
       theme(axis.title = element_blank())
p24 <- FeaturePlot(fibro, reduction = "tsne", features = "ACTA2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "ACTA2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p25 <- FeaturePlot(fibro, reduction = "tsne", features = "RGS5") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "RGS5") + 
       theme_yehlab() + 
       NoLegend() +
       theme(axis.title = element_blank())
(p23 | p24 | p25) / p19
```

#### Endothelial Cells

Next, we use the gene markers PLVAP and VWF (again provided in @elyada2017) to identify cluster 3 as the endothelial cells.

```{r}
p26 <- FeaturePlot(fibro, reduction = "tsne", features = "PLVAP") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "PLVAP") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p27 <- FeaturePlot(fibro, reduction = "tsne", features = "VWF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "VWF") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p26 | p27) / p19
```

#### apCAFs

Interestingly, we have a small cluster of 22 cells that does not appear to express any of the fibroblast, CAF, perivascular, or endothelial markers. We'll use differential expression testing to find out what makes it unique, and hopefully annotate it. After performing differential expression analysis, we found that the top 3 markers for cluster 4 are CLU, CD74, and CRYAB. Interestingly, CLU and CD74 were found to be differentially expressed in the apCAF population discovered in the KPC mouse models of CAFs in @elyada2017.

```{r}
fibro_markers <- FindAllMarkers(fibro, 
                                assay = "SCT", 
                                logfc.threshold = 1.5, 
                                test.use = "wilcox", 
                                only.pos = TRUE, 
                                random.seed = 629, 
                                verbose = FALSE)
fibro_markers[fibro_markers$cluster == 4, ] %>% arrange(desc(avg_logFC))
```

We re-run GSEA, again using the `VAM` package and the differentially expressed genes for the apCAF population as defined in @elyada2017 (with the mouse gene names converted to HGNC symbols). We can see that the apCAF pathway is strongly enriched in cluster 4 as compared to the other CAF clusters.

```{r}
apcaf_genes <- c("HLA-DQB1", "CD74", "SAA3P", "SLPI")
gene_sets <- list(icaf_genes, mycaf_genes, apcaf_genes, pan_caf_genes)
names(gene_sets) <- c("iCAF", "myCAF", "apCAF", "Pan-CAF")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
p28 <- FeaturePlot(fibro, reduction = "tsne", features = "apCAF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "apCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p28 / p19
```

### Visualization

Finally, we add cell labels to our identified clusters.

```{r}
fibro@meta.data$label <- case_when(fibro@meta.data$seurat_clusters == 0 ~ "myCAF", 
                                   fibro@meta.data$seurat_clusters == 1 ~ "Perivascular", 
                                   fibro@meta.data$seurat_clusters == 2 ~ "iCAF", 
                                   fibro@meta.data$seurat_clusters == 3 ~ "Endothelial", 
                                   fibro@meta.data$seurat_clusters == 4 ~ "apCAF")
p29 <- DimPlot(fibro, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank())
p29
```

## T Cells

Going back to the main `Seurat` object, we should have a large population of T and NK cells that warrants further investigation. Using CD3D expression we can easily identify clusters 0, 3, and 9 as the mixed T & NK cells. We already see some good separation, so reclustering the cells should have good results. We considered including cluster 5 in the T & NK cell reclustering, since part of it is located directly adjacent to the T / NK  clusters, but since it does not express CD3D at all we conclude that it most likely contains myeloid-type cells. 

```{r}
p30 <- FeaturePlot(pdac, reduction = "tsne", features = "CD3D") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p30 / p4
```

### Tumor

```{r}
nkt_tumor <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 9) & condition == "PDAC")
```

#### Reclustering

Here we run `ReclusterCells()`, while treating the NK & T cell clusters as one large group. This will hopefully allow use to elucidate T cell subtypes. We use fewer PCs for these cells since the differences between immune cells are subtle, and adding more PCs will most likely only contribute noise to our analyses. 

```{r}
nkt_tumor <- ReclusterCells(nkt_tumor, 
                            which.clust = list(0, 3, 9), 
                            merge.clusters = TRUE, 
                            n.variable.genes = 4000, 
                            n.PC = 15, 
                            k.vals = c(30, 40, 50, 60), 
                            resolution.vals = c(.2, .3, .4), 
                            which.dim.reduc = "tsne", 
                            redo.embedding = TRUE, 
                            random.seed = 629)
nkt_tumor_pc <- Embeddings(nkt_tumor, "pca")
```

Once again we'll run Fit-SNE on the reclustered cells. 

```{python}
# import data
nkt_tumor_pc = r.nkt_tumor_pc
# run Fit-SNE
affin_nkt_tumor = PerplexityBasedNN(nkt_tumor_pc, perplexity=80, metric='cosine', random_state=629)
init = initialization.pca(nkt_tumor_pc, random_state=629)
tsne_nkt_tumor = TSNEEmbedding(init, affin_nkt_tumor, negative_gradient_method='fft')
embed_nkt_tumor1 = tsne_nkt_tumor.optimize(n_iter=250, exaggeration=12, momentum=0.5) 
embed_nkt_tumor2 = embed_nkt_tumor1.optimize(n_iter=750, momentum=0.8)
affin_nkt_tumor.set_perplexity(20)
embed_nkt_tumor3 = embed_nkt_tumor2.optimize(n_iter=500)
```

The reembedding isn't perfect, which we somewhat expected as immune cells are difficult to tell apart based on the transcriptome alone. 

```{r}
embed_nkt_tumor <- as.matrix(py$embed_nkt_tumor3)
rownames(embed_nkt_tumor) <- colnames(nkt_tumor)
nkt_tumor@reductions$bh_tsne <- nkt_tumor@reductions$tsne
nkt_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_tumor, 
                                                 key = "FitSNE_", 
                                                 assay = "SCT", 
                                                 global = TRUE)
p31 <- DimPlot(nkt_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p31
```

#### Cell Type Identification

##### CD4+ T

CLusters 0 and 1 contain our CD4+ T cells, which we characterize using IL7R as we did in the PBMC3k vignette. It's outside of our scope here to determine which subtype each cluster belongs to, so we'll simply assign both clusters the same label and move on. 

```{r}
p32 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL7R") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p33 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD69") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p32 | p33) / p31
```

##### T-regs

Cluster 3 contains the regulatory T cells. 

```{r}
p34 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL2RA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p35 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "FOXP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p34 | p35) / p31
```

##### Proliferating T-regs

We can find the proliferating T-regs in cluster 7. 

```{r}
p36 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TOP2A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p36 / p31
```

##### Mast

Mast cells can be identified using TPSAB1 expression in cluster 6. 

```{r}
p37 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TPSAB1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p37 / p31
```

##### NK

NKG7 and PRF1 show us the NK cells in cluster 5. 

```{r}
p38 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "NKG7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p39 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "PRF1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p38 | p39) / p31
```

##### CD8+ T

We use CD8A and CD2 to reveal the CD8+ T cells within clusters 2 and 4. 

```{r}
p40 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD8A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p41 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p40 | p41) / p31
```

#### Visualization

We add labels to our T cell `Seurat` object and visualize the results. 

```{r}
nkt_tumor$label <- case_when(nkt_tumor$seurat_clusters == 0 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 1 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 2 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 3 ~ "T-reg", 
                             nkt_tumor$seurat_clusters == 4 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 5 ~ "NK", 
                             nkt_tumor$seurat_clusters == 6 ~ "Mast", 
                             nkt_tumor$seurat_clusters == 7 ~ "Proliferating T-reg")
p42 <- DimPlot(nkt_tumor, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p42
```

### Adjacent Normal

```{r}
nkt_norm <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 9) & condition == "AdjNorm")
```

#### Reclustering

```{r}
nkt_norm <- ReclusterCells(nkt_norm, 
                           which.clust = list(0, 3, 9), 
                           merge.clusters = TRUE, 
                           n.variable.genes = 4000, 
                           n.PC = 10, 
                           k.vals = c(10, 20, 30, 40), 
                           resolution.vals = c(.1, .2, .3, .4), 
                           which.dim.reduc = "tsne", 
                           redo.embedding = TRUE, 
                           random.seed = 629)
nkt_norm_pc <- Embeddings(nkt_norm, "pca")
```

As with the other reclusterings, we'll run Fit-SNE in order to (hopefully) obtain a better low-dimensional embedding of our cells.

```{python}
# import data
nkt_norm_pc = r.nkt_norm_pc
# run Fit-SNE
affin_nkt_norm = PerplexityBasedNN(nkt_norm_pc, perplexity=150, metric='cosine', random_state=629)
init = initialization.pca(nkt_norm_pc, random_state=629)
tsne_nkt_norm = TSNEEmbedding(init, affin_nkt_norm, negative_gradient_method='fft')
embed_nkt_norm1 = tsne_nkt_norm.optimize(n_iter=250, exaggeration=13, momentum=0.7) 
embed_nkt_norm2 = embed_nkt_norm1.optimize(n_iter=650, exaggeration=1, momentum=0.8)
affin_nkt_norm.set_perplexity(20)
embed_nkt_norm3 = embed_nkt_norm2.optimize(n_iter=500, exaggeration=1, momentum=0.8)
```

```{r}
embed_nkt_norm <- as.matrix(py$embed_nkt_norm3)
rownames(embed_nkt_norm) <- colnames(nkt_norm)
nkt_norm@reductions$bh_tsne <- nkt_norm@reductions$tsne
nkt_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_norm, 
                                                key = "FitSNE_", 
                                                assay = "SCT", 
                                                global = TRUE)
p43 <- DimPlot(nkt_norm, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p43
```

#### Cell Type Identification

##### CD4+ T

We can use IL7R expression to identify the CD4+ T cells in cluster 0. 

```{r}
p44 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "C1QA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p44 / p43
```

##### CD8+ T

Next we reveal the CD8+ T cells in clusterS 1 and 2 with CD8A, as per usual. 

```{r}
p45 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "CD8A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p45 / p43
```

##### T-reg

The T-reg cluster, cluster 3, is identified using IL2RA and FOXP3. 

```{r}
p46 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "IL2RA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p47 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "FOXP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p46 | p47) / p43
```

##### NK

The natural killers can be found in cluster 4 through their expression of PRF1.

```{r}
p48 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "PRF1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p48 / p43
```

##### Proliferating T-reg

The tiny proliferating T-reg population in cluster 5 is characterized by TOP2A. 

```{r}
p49 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TOP2A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p49 / p43
```

#### Visualization

We add final cell labels to our 9 cell clusters. 

```{r}
nkt_norm$label <- case_when(nkt_norm$seurat_clusters == 0 ~ "CD4+ T", 
                            nkt_norm$seurat_clusters == 1 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 2 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 3 ~ "T-reg",
                            nkt_norm$seurat_clusters == 4 ~ "NK",
                            nkt_norm$seurat_clusters == 5 ~ "Proliferating T-reg")
p50 <- DimPlot(nkt_norm, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p50
```

## Ductal Cells

### Reclustering

Next we'll examine the three ductal clusters identified by `SingleR` - clusters 6, 11, and 13.

```{r}
ductal <- ReclusterCells(pdac, 
                         which.clust = c(6, 11, 13), 
                         n.variable.genes = 4000, 
                         merge.clusters = TRUE, 
                         n.PC = 20, 
                         k.vals = c(10, 20, 30, 40), 
                         resolution.vals = c(.1, .2, .3, .4), 
                         which.dim.reduc = "tsne", 
                         redo.embedding = TRUE, 
                         do.plot = FALSE, 
                         random.seed = 629)
# classical 1 - 0 & 1 (TFF2)
# classical 2 - 5 (CRISP3)
# basal - 3 (DECODER basal & bc_ratio)
# secretory - 2 (SOD3 & CFTR)
# lipid proc - 4 (ANPEP)
```

Again, we'll run Fit-SNE on the reclustered cells. 

```{r}
duct_pc <- Embeddings(ductal, reduction = "pca")
```

```{python}
# import data
duct_pc = r.duct_pc
# run Fit-SNE
affin_duct = PerplexityBasedNN(duct_pc, perplexity=120, metric='cosine', random_state=629)
init = initialization.pca(duct_pc, random_state=629)
tsne_duct = TSNEEmbedding(init, affin_duct, negative_gradient_method='fft')
embed_d1 = tsne_duct.optimize(n_iter=250, exaggeration=8, momentum=0.6)
embed_d2 = embed_d1.optimize(n_iter=750, exaggeration=2, momentum=0.8)
affin_duct.set_perplexity(20)
embed_d3 = embed_d2.optimize(n_iter=400, exaggeration=1, momentum=0.5)
```

We pull the results into R, making sure to save the Barnes-Hut t-SNE results in another reduction slot.

```{r}
embed_duct <- as.matrix(py$embed_d3)
rownames(embed_duct) <- colnames(ductal)
# ductal_reclust@reductions$bh_tsne <- ductal_reclust@reductions$tsne
ductal@reductions$tsne<- CreateDimReducObject(embeddings = embed_duct, 
                                              key = "FitSNE_", 
                                              assay = "SCT", 
                                              global = TRUE)
p51 <- DimPlot(ductal, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p51
```

### Cell Type Identification 

We run a differential expression test to determine which genes characterize each cluster. 

```{r}
ductal_markers <- FindAllMarkers(ductal, 
                                 assay = "SCT", 
                                 logfc.threshold = .75, 
                                 test.use = "wilcox", 
                                 only.pos = TRUE, 
                                 verbose = FALSE, 
                                 random.seed = 629) %>% 
                  subset(p_val_adj < .05) %>% 
                  group_by(cluster) %>% 
                  top_n(n = 10, wt = avg_logFC)
ductal_markers %>% 
  top_n(n = 2, wt = avg_logFC) %>% 
  print()
```

#### Lipid Processing

We use ANPEP expression to identify the lipid processing ductal cells in cluster 4. 

```{r}
p52 <- FeaturePlot(ductal, reduction = "tsne", features = "ANPEP") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p52 / p51
```

#### Secretory

Expression of SOD3 reveals the secretory cells in cluster 2.

```{r}
p53 <- FeaturePlot(ductal, reduction = "tsne", features = "SOD3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p53 / p51
```

#### Classical 1

We use the classical PDAC compartment weights from `DECODER` to identify the classical epithelial cells in clusters 0, 1, and 5. One of these clusters will be further identified as the classical 2 subtype found by the authors. It is out of our scope to determine what's driving the differences in gene expression between the two classical 1 clusters.

```{r}
p54 <- FeaturePlot(ductal, reduction = "tsne", features = "classical") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Classical 1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p54 / p51
```

#### Classical 2

The classical 2 cells are located in cluster 5, as evidenced by their expression of CRISP3. With that informmation, we'll define clusters 0 and 1 as classical 1. 

```{r}
p55 <- FeaturePlot(ductal, reduction = "tsne", features = "CRISP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p55 / p51
```

#### Basal-like

The basal compartment weights from `DECODER` are highest in cluster 3, which we will denote as being composed of basal-like PDAC. 

```{r}
p56 <- FeaturePlot(ductal, reduction = "tsne", features = "basal") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Basal-like") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p56 / p51
```

#### Classical 3

```{r}
ductal_markers %>% 
  filter(cluster == 6) %>% 
  top_n(n = 5, wt = avg_logFC)
DimPlot(ductal, reduction = "tsne", group.by = "sample")
FeaturePlot(ductal, reduction = "tsne", features = "classical")
```


### Visualization

We add final cluster labels to our `Seurat` object and visualize the results.

```{r}
ductal$label <- case_when(ductal$seurat_clusters == 0 ~ "Classical 1", 
                          ductal$seurat_clusters == 1 ~ "Classical 1", 
                          ductal$seurat_clusters == 2 ~ "Secretory", 
                          ductal$seurat_clusters == 3 ~ "Basal-like", 
                          ductal$seurat_clusters == 4 ~ "Lipid Proc.", 
                          ductal$seurat_clusters == 5 ~ "Classical 2", 
                          ductal$seurat_clusters == 6 ~ "Classical 3")
p58 <- DimPlot(ductal, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p58
```

## Plasma Cells & Plasmacytoid DCs

We'll run `SCISSORS` on cluster 10 in order to reveal its subgroups, the plasma cells and plasmacytoid DCs, as defined by their expression of IGJ (aka JCHAIN) and IRF7, respectively. 

```{r}
p59 <- FeaturePlot(pdac, reduction = "tsne", features = "JCHAIN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "JCHAIN / IGJ") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p60 <- FeaturePlot(pdac, reduction = "tsne", features = "IRF7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p59 | p60) / p4
```

### Reclustering

We run `SCISSORS` with default parameters. 

```{r}
plas <- ReclusterCells(pdac, 
                       which.clust = 10, 
                       n.PC = 20, 
                       which.dim.reduc = "tsne",
                       redo.embedding = TRUE, 
                       do.plot = FALSE, 
                       random.seed = 629)
plas_pc <- Embeddings(plas, reduction = "pca")
```

Even though in this case it's probably not necessary, we'll run Fit-SNE on the cells just so that our axis labels are consistent. 

```{python}
# import data
plas_pc = r.plas_pc
# run Fit-SNE
affin_plas = PerplexityBasedNN(plas_pc, perplexity=200, metric='cosine', random_state=629)
init = initialization.pca(plas_pc, random_state=629)
tsne_plas = TSNEEmbedding(init, affin_plas, negative_gradient_method='fft')
embed_p1 = tsne_plas.optimize(n_iter=250, exaggeration=4, momentum=0.6)
embed_p2 = embed_p1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_plas.set_perplexity(30)
embed_p3 = embed_p2.optimize(n_iter=400)
```

```{r}
embed_plas <- as.matrix(py$embed_p3)
rownames(embed_plas) <- colnames(plas)
# plas@reductions$bh_tsne <- plas@reductions$tsne
plas@reductions$tsne<- CreateDimReducObject(embeddings = embed_plas, 
                                            key = "FitSNE_", 
                                            assay = "SCT", 
                                            global = TRUE)
p61 <- DimPlot(plas, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p61
```

### Cell Type Identification

We use the aforementioned markers to identify our two clusters. 

```{r}
p62 <- FeaturePlot(plas, reduction = "tsne", features = "JCHAIN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "JCHAIN") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p63 <- FeaturePlot(plas, reduction = "tsne", features = "IRF7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p62 | p63) / p61
```

### Visualization

As before, we add cell labels and visualize the results. 

```{r}
plas$label <- case_when(plas$seurat_clusters == 0 ~ "Plasma", 
                        plas$seurat_clusters == 1 ~ "pDC")
p64 <- DimPlot(plas, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p64
```

## B Cells

We can identify the B cells in cluster 4 using joint expression of MS4A1 and CD79A.  

```{r}
p65 <- FeaturePlot(pdac, reduction = "tsne", features = "MS4A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MS4A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p66 <- FeaturePlot(pdac, reduction = "tsne", features = "CD79A") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD79A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p65 | p66) / p4
```

## Myeloid Cells

Lastly, we'll split up the myeloid population by tissue condition (PDAC vs. adjacent normal) just like we did with the NK / T cells. We'll run `SCISSORS`, annotate the clusters, and visualize the results. 

### Tumor 

First we'll attempt to assign broad cell type labels to each of the four putative myeloid clusters. We'll use the marker genes from Elyada *et al* once again. 

```{r}
DimPlot(pdac, reduction = "tsne")
FeaturePlot(pdac, reduction = "tsne", features = "CD79A")
```


```{r}
myo_tumor <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 5, 8) & condition == "PDAC")
p67 <- DimPlot(myo_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p67
```

Cluster 1 appears to be composed of our resident and alternatively activated macrophages due to its expression of CD14 and C1QA. 

```{r}
p68 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "CD14") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p69 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "C1QA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p68 | p69) / p67
```

We can use LYZ and S100A8 expression to reveal the classic monocytes and neutrophils in cluster 2. 

```{r}
p70 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "LYZ") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LYZ") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p71 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "S100A8") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p70 | p71) / p67
```

Lastly, we show that cluster 8 contains our various DC subtypes through its expression of FCER1A, a canonical dendritic cell marker. 

```{r}
p72 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "FCER1A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FCER1A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p72 / p67
```


```{r}
FeaturePlot(myo_tumor, reduction = "tsne", features = "NKG7")
DimPlot(myo_tumor, reduction = "tsne", group.by = "SingleR.labels.sc")
```


#### Reclustering

Since we've already broadly defined our three clusters in terms of what types of cells they contain, we won't merge the three clusters into one group before reclustering using `SCISSORS`. 

```{r}
myo_tumor_reclust <- ReclusterCells(myo_tumor, 
                                    which.clust = c(1, 2, 5, 8), 
                                    n.PC = 20, 
                                    merge.clusters = TRUE, 
                                    k.vals = c(20, 30, 40), 
                                    resolution.vals = c(.1, .2, .3), 
                                    n.variable.genes = 4000, 
                                    which.dim.reduc = "tsne", 
                                    redo.embedding = TRUE, 
                                    do.plot = FALSE, 
                                    random.seed = 629)


# clust 4 - Alt. macrophage
# 
DimPlot(myo_tumor_reclust, reduction = "tsne")
FeaturePlot(myo_tumor_reclust, reduction = "tsne", features = "CD1A")
```

For the last time, we'll run Fit-SNE on our reclustered cells. 

```{r}
macro_tumor_pc <- Embeddings(macro_tumor, "pca")
mono_tumor_pc <- Embeddings(mono_tumor, "pca")
dc_tumor_pc <- Embeddings(dc_tumor, "pca")
```

```{python}
# import data
macro_pc = r.macro_tumor_pc
mono_pc = r.mono_tumor_pc
dc_pc = r.dc_tumor_pc
# Fit-SNE - macrophages
affin_macro = PerplexityBasedNN(macro_pc, perplexity=100, metric='cosine', random_state=629)
init = initialization.pca(macro_pc, random_state=629)
tsne_macro = TSNEEmbedding(init, affin_macro, negative_gradient_method='fft')
embed_mac1 = tsne_macro.optimize(n_iter=250, exaggeration=12, momentum=0.6)
embed_mac2 = embed_mac1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_macro.set_perplexity(30)
embed_mac3 = embed_mac2.optimize(n_iter=250)
# Fit-SNE - monocytes
affin_mono = PerplexityBasedNN(mono_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(mono_pc, random_state=629)
tsne_mono = TSNEEmbedding(init, affin_mono, negative_gradient_method='fft')
embed_mono1 = tsne_mono.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_mono2 = embed_mono1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
# Fit-SNE - DC
affin_dc = PerplexityBasedNN(dc_pc, perplexity=60, metric='cosine', random_state=629)
init = initialization.pca(dc_pc, random_state=629)
tsne_dc = TSNEEmbedding(init, affin_dc, negative_gradient_method='fft')
embed_dc1 = tsne_dc.optimize(n_iter=250, exaggeration=8, momentum=0.6)
embed_dc2 = embed_dc1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_dc.set_perplexity(20)
embed_dc3 = embed_dc2.optimize(n_iter=100)
```

We pull the results back into R and visualize them. 

```{r}
embed_macro <- as.matrix(py$embed_mac3)
rownames(embed_macro) <- colnames(macro_tumor)
macro_tumor@reductions$bh_tsne <- macro_tumor@reductions$tsne
macro_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_macro, 
                                                   key = "FitSNE_", 
                                                   assay = "SCT",
                                                   global = TRUE)
embed_mono <- as.matrix(py$embed_mono2)
rownames(embed_mono) <- colnames(mono_tumor)
mono_tumor@reductions$bh_tsne <- mono_tumor@reductions$tsne
mono_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_mono, 
                                                  key = "FitSNE_", 
                                                  assay = "SCT",
                                                  global = TRUE)
embed_dc <- as.matrix(py$embed_dc3)
rownames(embed_dc) <- colnames(dc_tumor)
dc_tumor@reductions$bh_tsne <- dc_tumor@reductions$tsne
dc_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_dc, 
                                                key = "FitSNE_", 
                                                assay = "SCT",
                                                global = TRUE)
```

```{r}
p89 <- DimPlot(macro_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p90 <- DimPlot(mono_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p91 <- DimPlot(dc_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p89
p90
p91
```

#### Cell Type Identification

##### Monoctyes & Neutrophils

First we ID the neutrophils in cluster 1 using S100A8 and S100A9 - marker genes used by Elyada *et al*. 

```{r}
p92 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "S100A8") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p93 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "S100A9") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A9") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p92 | p93) / p90
```

We can identify the CD14+ classical monocytes in cluster 2 through expression of, you guessed it, CD14, and the CD14+ CD16+ nonclassical monocytes in cluster 0 through their additional expression of CD16 aka FCGR3A. The nonclassical monocytes are proinflammatory, which follows from our knowledge of PDAC as an inflammatory disease. 

```{r}
p94 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "CD14") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p95 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "FCGR3A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD16") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p94 | p95) / p90
```

We add cell labels to our `Seurat` object, then we're off to the DCs. 

```{r}
mono_tumor$label <- case_when(mono_tumor$seurat_clusters == 0 ~ "FCGR3A+ Monocyte", 
                              mono_tumor$seurat_clusters == 1 ~ "Neutrophil", 
                              mono_tumor$seurat_clusters == 2 ~ "Classical Monocyte")
```

##### Dendritic Cells

We can use CLEC9A to annotate the convectional DC1 cells in cluster 4. 

```{r}
p96 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CLEC9A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CLEC9A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p96 / p91
```

Joint expression of CD1A and CD207 reveals the Langerhans-like DCB (highly expressed) and Langerhans-like DCA (lowly expressed) cells in clusters 3 and 5, respectively. 

```{r}
p97 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CD1A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD1A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p98 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CD207") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD207") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p97 | p98) / p91
```

Next we use LAMP3 to identify the activated DCs in cluster 6. 

```{r}
p99 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "LAMP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LAMP3") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p99 / p91
```

Joint expression of MS4A7 and FCGR3A reveals a group of FCGR3A+ monocytes in cluster 0. 

```{r}
p100 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "MS4A7") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MS4A7") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p101 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "FCGR3A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FCGR3A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p100 | p101) / p91
```

Next, we can use VEGFA expression to define cluster 1 as containing M2d macrophages. 

```{r}
p102 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "VEGFA") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "VEGFA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p102 / p91
```

And lastly we'll use joint expression of CD163 and CD31 (aka PECAM1) to identify the tumor-associated macrophages in cluster 2. 

```{r}
p103 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CD163") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD163") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p104 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "PECAM1") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "PECAM1") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p103 | p104) / p91
```

We add final subcluster cell type labels to our `Seurat` object, and then we're off to the macrophages. 

```{r}
dc_tumor$label <- case_when(dc_tumor$seurat_clusters == 0 ~ "FCGR3A+ Monocyte", 
                            dc_tumor$seurat_clusters == 1 ~ "M2d", 
                            dc_tumor$seurat_clusters == 2 ~ "TAM", 
                            dc_tumor$seurat_clusters == 3 ~ "Langerhans-like DCB", 
                            dc_tumor$seurat_clusters == 4 ~ "Conventional DC", 
                            dc_tumor$seurat_clusters == 5 ~ "Langerhans-like DCA", 
                            dc_tumor$seurat_clusters == 6 ~ "Activated DC")
```

##### Macrophages 

First we use a Wilcox test for differential expression to determine markers for each subcluster. 

```{r}
macro_tumor_markers <- FindAllMarkers(macro_tumor, 
                                      assay = "SCT", 
                                      logfc.threshold = .5, 
                                      test.use = "wilcox", 
                                      only.pos = TRUE, 
                                      verbose = FALSE, 
                                      random.seed = 629, 
                                      min.pct = .5, 
                                      min.diff.pct = .1) %>% 
                       subset(p_val_adj < .01) %>% 
                       group_by(cluster) %>% 
                       top_n(n = 5, wt = avg_logFC)
macro_tumor_markers

FeaturePlot(macro_tumor, reduction = "tsne", features = "SPP1") 
```

We can use MARCO expression to identify the alternatively activated (M2a) macrophages in cluster 1. 

```{r}
p105 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "MARCO") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MARCO") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p105 / p89
```

From here on out things get a bit more complicated, as we're delving beyond the cell types identified in Elyada *et al*, and macrophage subtypes can be difficult to tell apart. 

We'll use joint expression of CD86 and HLA-DQA2 (involved in MCH class II receptor activity) to identify our M2b macrophages in clusters 0 and 4. 

```{r}
p106 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "CD86") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD86") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p107 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "HLA-DQA2") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "HLA-DQA2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p106 | p107) / p89
```

Next, we'll isolate the resident macrophages in cluster 2 through their expression of C1QA, APOE and SPP1. .

```{r}
p108 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "C1QA") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p109 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "APOE") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "APOE") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p110 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "SPP1") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "SPP1") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p108 | p109 | p110) / p89
```

Next we'll annotate cluster 3 as tumor-associated macrophages using CD163, CD68, MRC1 (aka CD206), and VEGFA expression. 

```{r}
p111 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "CD163") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD163") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p112 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "CD68") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD68") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p113 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "MRC1") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MRC1") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
((p111 | p112) / (p113 | p109)) / p89
```

Next up are the TCR+ macrophages in cluster 6. We'll use LAT, LCK, CD68, and FYN expression. 

```{r}
p114 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "LAT") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LAT") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p115 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "LCK") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LCKL") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p116 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "FYN") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FYN") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
((p114 | p112) / (p115 | p116)) / p89
```

Finally, we have one cluster left that evades definition. It does however, highly express the gene TYMS at a a statistically significant level. [This gene](https://www.genecards.org/cgi-bin/carddisp.pl?gene=TYMS&keywords=TYMS) has polymorphisms associated with neoplasia and chemotherapy response, so we have reason to believe that it might be important in the context of PDAC.  

```{r}
p117 <- FeaturePlot(macro_tumor, reduction = "tsne", features = "TYMS") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "TYMS") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p117 / p89
```

We'll add cell labels to our `Seurat` object. 

```{r}
macro_tumor$label <- case_when(macro_tumor$seurat_clusters == 0 ~ "M2b", 
                               macro_tumor$seurat_clusters == 1 ~ "M2a", 
                               macro_tumor$seurat_clusters == 2 ~ "Resident Macrophage", 
                               macro_tumor$seurat_clusters == 3 ~ "TAM", 
                               macro_tumor$seurat_clusters == 4 ~ "M2b", 
                               macro_tumor$seurat_clusters == 5 ~ "TYMS+ Macrophage", 
                               macro_tumor$seurat_clusters == 6 ~ "TCR+ Macrophage")
```

#### Visualization

##### Monocytes & Neutrophils

```{r}
p118 <- DimPlot(mono_tumor, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p118
```

##### DCs

```{r}
p119 <- DimPlot(dc_tumor, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p119
```

##### Macrophages

```{r}
p120 <- DimPlot(macro_tumor, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p120
```

### Adjacent Normal

Since there's a lot fewer cells in the normal tissue myeloid population, we'll give merging the clusters together a shot. 

```{r}
myo_norm <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 8) & condition == "AdjNorm")
```

#### Reclustering

```{r}
myo_norm <- ReclusterCells(myo_norm, 
                           which.clust = c(1, 2, 8), 
                           merge.clusters = TRUE, 
                           n.variable.genes = 4000, 
                           n.PC = 20, 
                           which.dim.reduc = "tsne", 
                           redo.embedding = TRUE, 
                           random.seed = 629)
myo_norm_pc <- Embeddings(myo_norm, reduction = "pca")
```

For the last time, we run Fit-SNE in order to obtain a better embedding. 

```{python}
# import data
myo_norm_pc = r.myo_norm_pc
# run Fit-SNE
affin_myo_norm = PerplexityBasedNN(myo_norm_pc, perplexity=100, metric='cosine', random_state=629)
init = initialization.pca(myo_norm_pc, random_state=629)
tsne_myo_norm = TSNEEmbedding(init, affin_myo_norm, negative_gradient_method='fft')
embed_m1 = tsne_myo_norm.optimize(n_iter=350, exaggeration=12, momentum=0.6)
embed_m2 = embed_m1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_myo_norm.set_perplexity(30)
embed_m3 = embed_m2.optimize(n_iter=500, exaggeration=1, momentum=0.7)
```

We fetch the results from Python and add them to our `Seurat` object. 

```{r}
embed_myo <- as.matrix(py$embed_m3)
rownames(embed_myo) <- colnames(myo_norm)
myo_norm@reductions$bh_tsne <- myo_norm@reductions$tsne
myo_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_myo, 
                                                key = "FitSNE_", 
                                                assay = "SCT",
                                                global = TRUE)
p121 <- DimPlot(myo_norm, reduction = "tsne") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p121
```

#### Cell Type Identification

##### Classic Monocytes

We use high LYZ & low CD14 expression to denote the classic monocytes in cluster 0. 

```{r}
p122 <- FeaturePlot(myo_norm, reduction = "tsne", features = "LYZ") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LYZ") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p123 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD14") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p122 | p123) / p121
```

##### Resident Macrophages

Next, we define cluster 1 as containing our resident macrophages using C1QA and APOE.

```{r}
p124 <- FeaturePlot(myo_norm, reduction = "tsne", features = "C1QA") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p125 <- FeaturePlot(myo_norm, reduction = "tsne", features = "APOE") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "APOE") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p124 | p125) / p121
```

##### Conventional DC1

Cluster 6 contains our conventional dendritic cells, as evidenced by their expression of CIRF8. 

```{r}
p126 <- FeaturePlot(myo_norm, reduction = "tsne", features = "IRF8") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF8") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p126 / p121
```

##### CD8+ T

We find a surprise cluster of CD8+ T cells - we can be confident in that annotation, though, due to its expression of CD8A, CD2, and NKG7. 

```{r}
p127 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD8A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD8A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p128 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD2") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p129 <- FeaturePlot(myo_norm, reduction = "tsne", features = "NKG7") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "NKG7") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p127 | p128 | p129) / p121
```

##### Tumor-Associated Macrophages

As we did with the tumor myeloid cells, we find a group of IL10+ VEGFA+ CD163+ tumor-associated macrophages in cluster 4. 

```{r}
p130 <- FeaturePlot(myo_norm, reduction = "tsne", features = "IL10") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IL10") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p131 <- FeaturePlot(myo_norm, reduction = "tsne", features = "VEGFA") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "VEGFA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p132 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD163") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD163") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p130 | p131 | p132) / p121
```

##### Langerhans-like DCA

We use HLA-DRA, CD1A, and CD207 to identify the Langerhans-like type A dendritic cells in cluster 2. 

```{r}
p133 <- FeaturePlot(myo_norm, reduction = "tsne", features = "HLA-DRA") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "HLA-DRA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p134 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD1A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD1A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p135 <- FeaturePlot(myo_norm, reduction = "tsne", features = "CD207") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD207") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p133 | p134 | p135) / p121
```

Last but not least, we find another group of neutrophils in cluster 3 using the same marker genes as we did earlier. 

```{r}
p136 <- FeaturePlot(myo_norm, reduction = "tsne", features = "S100A8") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p137 <- FeaturePlot(myo_norm, reduction = "tsne", features = "S100A9") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A9") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p136 | p137) / p121
```

We add labels to our clusters. 

```{r}
myo_norm$label <- case_when(myo_norm$seurat_clusters == 0 ~ "Classic Monocyte", 
                            myo_norm$seurat_clusters == 1 ~ "Resident Macrophage", 
                            myo_norm$seurat_clusters == 2 ~ "Langerhans-like DCA", 
                            myo_norm$seurat_clusters == 3 ~ "Neutrophil", 
                            myo_norm$seurat_clusters == 4 ~ "TAM", 
                            myo_norm$seurat_clusters == 5 ~ "CD8+ T", 
                            myo_norm$seurat_clusters == 6 ~ "Conventional DC1")
```

#### Visualization

Here are the final annotations for the adjacent normal tissue myeloid population. 

```{r}
p138 <- DimPlot(myo_norm, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p138
```

# Conclusions

# Save Figures & Data

We'll create a quick convenience function to help us save the figures.

```{r}
saveSCISSORS <- function(plot = NULL, 
                         name = NULL, 
                         border = TRUE, 
                         pub.ready = FALSE) {
  if (is.null(plot) | is.null(name)) stop("You forgot some arguments.")
  if (pub.ready) {
    dir <- "~/Desktop/R/SCISSORS/vignettes/figures_pub/PBMC"
    if (!border) {
      plot <- plot + 
              theme(panel.border = element_blank(), 
                    axis.title = element_blank(), 
                    legend.position = "none")
    } else {
      plot <- plot + 
              theme(axis.title = element_blank(), 
                    legend.position = "none")
    }
    ggsave(filename = paste0(name, ".pdf"), 
           device = "pdf", 
           units = "in",
           path = dir, 
           height = 5, 
           width = 5) 
  } else {
    dir <- "~/Desktop/R/SCISSORS/vignettes/figures_supp/PBMC"
    ggsave(filename = paste0(name, ".pdf"), 
           device = "pdf", 
           units = "in",
           path = dir, 
           height = 5, 
           width = 5) 
  }
}
```

This section isn't worth reading; it's here solely to prove that the figures we present in our publication were dynamically generated during the knitting of this document.

```{r}
saveSCISSORS(plot = p0, name = "Seurat_Clusters_tSNE.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p1, name = "Seurat_Clusters_UMAP.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p4, name = "Seurat_Clusters_FitSNE.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p5, name = "SingleR_Annos_sc_ref.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p6, name = "SingleR_Annos_bulk_ref.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p7, name = "CONICSmat_Annos.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p8, name = "DECODER_basal.pdf")
saveSCISSORS(plot = p9, name = "DECODER_classical.pdf")
saveSCISSORS(plot = p10, name = "DECODER_exocrine.pdf")
saveSCISSORS(plot = p11, name = "DECODER_endocrine.pdf")
saveSCISSORS(plot = p12, name = "DECODER_immune.pdf")
saveSCISSORS(plot = p13, name = "DECODER_normal_stroma.pdf")
saveSCISSORS(plot = p14, name = "DECODER_act_stroma.pdf")
saveSCISSORS(plot = p15, name = "Fibro_COL1A1.pdf")
saveSCISSORS(plot = p16, name = "Fibro_COL3A1.pdf")
saveSCISSORS(plot = p17, name = "Fibro_LUM.pdf")
saveSCISSORS(plot = p18, name = "Fibro_DCN.pdf")
saveSCISSORS(plot = p19, name = "Fibro_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p20, name = "Fibro_PanCAF_VAM.pdf")
saveSCISSORS(plot = p21, name = "Fibro_iCAF_VAM.pdf")
saveSCISSORS(plot = p22, name = "Fibro_myCAF_VAM.pdf")
saveSCISSORS(plot = p23, name = "Perivascular_IGFBP7.pdf")
saveSCISSORS(plot = p24, name = "Perivascular_ACTA2.pdf")
saveSCISSORS(plot = p25, name = "Perivascular_RGS5.pdf")
saveSCISSORS(plot = p26, name = "Endothelial_PLVAP.pdf")
saveSCISSORS(plot = p27, name = "Endothelial_VWF.pdf")
saveSCISSORS(plot = p28, name = "Fibro_apCAF_VAM.pdf")
saveSCISSORS(plot = p29, name = "Fibro_SCISSORS_annos.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p30, name = "NKT_All_CD3D.pdf")
saveSCISSORS(plot = p31, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p32, name = "NKT_Tumor_IL7R_CD4T.pdf")
saveSCISSORS(plot = p33, name = "NKT_Tumor_CD69_CD4T.pdf")
saveSCISSORS(plot = p34, name = "NKT_Tumor_IL2RA_Treg.pdf")
saveSCISSORS(plot = p35, name = "NKT_Tumor_FOXP3_Treg.pdf")
saveSCISSORS(plot = p36, name = "NKT_Tumor_TOP2A_Prolif_Treg.pdf")
saveSCISSORS(plot = p37, name = "NKT_Tumor_TPSAB1_Mast.pdf")
saveSCISSORS(plot = p38, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p39, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p40, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p41, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p42, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p43, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p44, name = "NKT_Tumor_SCISSORS_clusts.pdf")
saveSCISSORS(plot = p45, name = "NKT_Tumor_SCISSORS_clusts.pdf")
```

And of course:

```{r}
sessionInfo()
```
