---
title: "Reclustering Analysis of Cancer-Associated Fibroblast Using SCISSORS"
subtitle: "Jack Leary"
author: 
  - "University of North Carolina at Chapel Hill - Lineberger Comprehensive Cancer Center"
  - "University of Florida - Department of Biostatistics"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: paper
    highlight: tango
    df_print: kable
    toc: true
    toc_float: true
    code_folding: show
    code_download: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE, 
                      warning = FALSE, 
                      results = "hold", 
                      fig.align = "center")
reticulate::use_virtualenv("~/Desktop/Python/science/venv/", required = TRUE)
set.seed(629)
```

# Introduction

In 2017, the Tuveson Lab at Cold Spring Harbor Cancer Center published a paper written by Elyada *et al* (2017) that detailed the discovery of cancer-associated fibroblasts (CAFs) in mice. The subtypes were then validated in human samples affected with PDAC in a subsequent paper released in 2019. Here we will use SCISSORS to identify the CAF subtypes within the larger fibroblast population, immune cell types within broadly-defined immune clusters, and ductal & PDAC subtypes within the ductal group. 

# Libraries

## R

```{r}
library(pals)        # high N discrete color palettes
library(VAM)         # single cell GSEA
library(dplyr)       # tidy data 
library(Seurat)      # single cell infrastructure
library(ggplot2)     # plots
library(SingleR)     # cell type assignment
library(decoderr)    # de novo deconvolution 
library(SCISSORS)    # our package
library(mixtools)    # Gaussian mixture model estimation
library(patchwork)   # plot grids
library(paletteer)   # more color palettes
library(CONICSmat)   # CNV estimation
library(reticulate)  # Python interface
library(wesanderson) # even more color palettes
```

## Python

```{python}
import numpy as np
from openTSNE import TSNEEmbedding
from openTSNE import initialization
from openTSNE.affinity import Multiscale
from openTSNE.affinity import PerplexityBasedNN
```

# Data

First we load in the $\text{gene} \times \text{cell}$ counts matrix, then create a `Seurat` object to hold it in. Next, we add samplename, tissue type, and patient sex metadata taken from the publicly available dataset.

```{r}
raw_counts <- Read10X(data.dir = "~/Desktop/Data/Elyada Raw/All Human/")
pdac <- CreateSeuratObject(raw_counts, 
                           project = "Elyada", 
                           min.cells = 3, 
                           min.features = 500)
pdac@meta.data$sample <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "SRR9274536", 
                                   grepl("-2", rownames(pdac@meta.data)) ~ "SRR9274537", 
                                   grepl("-3", rownames(pdac@meta.data)) ~ "SRR9274538", 
                                   grepl("-4", rownames(pdac@meta.data)) ~ "SRR9274539",
                                   grepl("-5", rownames(pdac@meta.data)) ~ "SRR9274540", 
                                   grepl("-6", rownames(pdac@meta.data)) ~ "SRR9274541", 
                                   grepl("-7", rownames(pdac@meta.data)) ~ "SRR9274542", 
                                   grepl("-8", rownames(pdac@meta.data)) ~ "SRR9274543", 
                                   grepl("-9", rownames(pdac@meta.data)) ~ "SRR9274544")
pdac@meta.data$condition <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-2", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-3", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-4", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-5", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-6", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-7", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-8", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-9", rownames(pdac@meta.data)) ~ "PDAC")
pdac@meta.data$sex <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-2", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-3", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-4", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-5", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-6", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-7", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-8", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-9", rownames(pdac@meta.data)) ~ "female")
```

```{r, echo=FALSE}
rm(raw_counts)
```

# Preprocessing

We run the typical single cell pre-processing steps on our cells - normalization, dimension reduction, and clustering. The t-SNE embedding looks decent, but could definitely be improved. Globally, the clusters are arranged in a blob - which isn't very informative - though the local structure seems to have been preserved fairly well. Clusters 5 and 11 are split in two, which we would prefer not to have happen. 

```{r, warning=FALSE}
pdac <- PrepareData(pdac, 
                    n.variable.genes = 4000, 
                    n.PC = 20, 
                    which.dim.reduc = "tsne", 
                    initial.resolution = .5, 
                    random.seed = 629)
p0 <- DimPlot(pdac, reduction = "tsne") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "t-SNE 1", y = "t-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p0
```

## Optimize Dimension Reduction

I think the two-dimensional visualization of the cells could be improved. We'll try using UMAP and the Fast Fourier Transform-accelerated Fit-SNE (as implemented in the `openTSNE` library) to improve the embedding.

### UMAP

It seems like UMAP does a good job of clearly separating our clusters and preserving the global structure of the data. However,  difficult to see local structure within some of the clusters due to their density, and cluster 5 is split across the 2D UMAP space.

```{r}
pdac <- RunUMAP(pdac, 
                reduction = "pca", 
                dims = 1:20, 
                umap.method = "uwot", 
                n.components = 2, 
                n.epochs = 750, 
                n.neighbors = 50, 
                metric = "cosine", 
                seed.use = 629, 
                verbose = FALSE)
p1 <- DimPlot(pdac, reduction = "umap") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "UMAP 1", y = "UMAP 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p1
```

### Fit-SNE

For information on how to install the `openTSNE` implementation of Fit-SNE and how to run the algorithm, please visit [the excellent GitHub repository of Pavlin Policar](https://github.com/pavlin-policar/openTSNE).

First we'll run a simple, standard Fit-SNE embedding. It's necessary to make the PCA embeddings accessible by Python.

```{r}
pc_df <- Embeddings(pdac, reduction = "pca")
```

```{python}
# import data
pc_df = np.array(r.pc_df)
# run Fit-SNE
affin = PerplexityBasedNN(pc_df, perplexity=50, metric='cosine', random_state=629)
init = initialization.pca(pc_df, random_state=629)
tsne1 = TSNEEmbedding(init, affin, negative_gradient_method='fft')
embed1 = tsne1.optimize(n_iter=350, exaggeration=12, momentum=0.6) 
embed2 = embed1.optimize(n_iter=1000, momentum=0.8)
```

The embedding looks good, but the subclusters within clusters 5 and 11 are still clearly separated, and a small part of cluster 12 is far from the rest of the cells in the cluster.

```{r}
embed <- as.matrix(py$embed2)
rownames(embed) <- colnames(pdac)
pdac@reductions$fitsne <- CreateDimReducObject(embeddings = embed, 
                                               key = "FitSNE_", 
                                               assay = "SCT", 
                                               global = TRUE)
p2 <- DimPlot(pdac, reduction = "fitsne") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p2
```

We use the ability of Fit-SNE to consider multiple perplexities in an attempt to provide a more accurate representation of both local and global structure in our data.

```{python}
affin_anneal = PerplexityBasedNN(pc_df, perplexity=150, metric='cosine', random_state=629)
tsne2 = TSNEEmbedding(init, affin_anneal, negative_gradient_method='fft')
embed3 = tsne2.optimize(n_iter=350, exaggeration=12, momentum=0.5)
embed4 = embed1.optimize(n_iter=1000, exaggeration=1, momentum=0.8)
affin_anneal.set_perplexity(30)
embed5 = embed4.optimize(n_iter=500, momentum=0.8)
```

The perplexity annealing hasn't really improved or even changed the embedding much at all.

```{r}
embed_multi <- as.matrix(py$embed5)
rownames(embed_multi) <- colnames(pdac)
pdac@reductions$fitsne_multi <- CreateDimReducObject(embeddings = embed_multi, 
                                                     key = "FitSNE_", 
                                                     assay = "SCT", 
                                                     global = TRUE)
p3 <- DimPlot(pdac, reduction = "fitsne_multi") + 
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p3
```

Finally, we'll use a multiscale kernel to represent the data in the high-dimensional space instead of the default Gaussian. This in tandem with the multiple perplexity values will preserve as much as possible of both local and global structure.

```{python}
affin_multi = Multiscale(pc_df, perplexities=[30, 200], metric='cosine', random_state=629)
tsne3 = TSNEEmbedding(init, affin_multi, negative_gradient_method='fft')
embed6 = tsne3.optimize(n_iter=450, exaggeration=12, momentum=0.6)
embed7 = embed6.optimize(n_iter=1200, exaggeration=1, momentum=0.8)
```

The multiscale trick doesn't completely fix the embedding issues either, but it does clean up a bit of the noise in the plot, and it does place the subclusters in clusters 11 and 5 closer together. We'll use this embedding going forward and make sure to examine those clusters for subgroups using `SCISSORS`.

```{r}
embed_kern <- as.matrix(py$embed7)
rownames(embed_kern) <- colnames(pdac)
pdac@reductions$fitsne_kern<- CreateDimReducObject(embeddings = embed_kern, 
                                                   key = "FitSNE_", 
                                                   assay = "SCT", 
                                                   global = TRUE)

p4 <- DimPlot(pdac, reduction = "fitsne_kern") +
      scale_color_manual(values = unname(tableau20())) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p4
```

Looking at the embedding, we can see that clusters 3, 4, 5, 10, and 11 contain clear small subclusters. I'm also intrigued by the shape of cluster 8.  We'll be sure to investigate each of these in turn, and annotate the groups using marker genes.

Due to the way `Seurat` accesses cell embeddings, we'll need to replace our original t-SNE dimension reduction in our `Seurat` object with the new Fit-SNE version. We'll keep the original Barnes-Hut t-SNE embedding under a separate name.

```{r}
pdac@reductions$bh_tsne <- pdac@reductions$tsne
pdac@reductions$tsne <- pdac@reductions$fitsne_kern
```

```{r, echo=FALSE}
rm(embed, embed_kern, embed_multi, pc_df)
```

# SingleR Cell Type Identification

## Single Cell RNA-seq Reference Data

Here we use the `SingleR` package to identify broad cell types. The reference dataset we load is an `sctransform`-normalized version of the raw counts available in `scRNAseq::BaronPancreasData()`, which consists of normal pancreas cells that were sequenced and annotated by the researchers.

```{r}
sc_ref <- readRDS("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/Seurat/single_cell_ref_normalized.Rds")
sc_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                    ref = sc_ref, 
                    labels = sc_ref$label, 
                    method = "cluster", 
                    clusters = pdac$seurat_clusters, 
                    de.method = "wilcox")
pdac[["SingleR.labels.sc"]] <- sc_preds$labels[match(pdac[[]][["seurat_clusters"]], rownames(sc_preds))]
pdac$SingleR.labels.sc <- case_when(pdac$SingleR.labels.sc == "acinar" ~ "Acinar", 
                                    pdac$SingleR.labels.sc == "activated_stellate" ~ "Activated Stellate",
                                    pdac$SingleR.labels.sc == "ductal" ~ "Ductal", 
                                    pdac$SingleR.labels.sc == "macrophage" ~ "Macrophage",
                                    pdac$SingleR.labels.sc == "t_cell" ~ "T", 
                                    pdac$SingleR.labels.sc == "quiescent_stellate" ~ "Quiescent Stellate")
```

We can see that there's a large immune population, as well as smaller ductal, fibroblast (denoted activated and quiescent stellate in the reference dataset), and acinar groups. These broad cell types line up with what we expected to see given the authors' original cell cluster annotations. 

```{r}
p5 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.sc") + 
      scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p5
```

## Bulk Tissue RNA-seq Reference Data

This dataset is composed of labeled & log-normalized bulk RNA-seq samples from the Human Primary Cell Atlas.

```{r}
bulk_ref <- HumanPrimaryCellAtlasData()
bulk_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                      ref = bulk_ref, 
                      labels = bulk_ref$label.main, 
                      method = "cluster", 
                      clusters = pdac$seurat_clusters, 
                      de.method = "wilcox")
pdac[["SingleR.labels.bulk"]] <- bulk_preds$labels[match(pdac[[]][["seurat_clusters"]], 
                                                         rownames(bulk_preds))]
pdac$SingleR.labels.bulk <- case_when(pdac$SingleR.labels.bulk == "B_cell" ~ "B", 
                                      pdac$SingleR.labels.bulk == "Epithelial_cells" ~ "Epithelial", 
                                      pdac$SingleR.labels.bulk == "B_cell-" ~ "B", 
                                      pdac$SingleR.labels.bulk == "T_cells" ~ "T", 
                                      pdac$SingleR.labels.bulk == "Endothelial_cells" ~ "Endothelial", 
                                      TRUE ~ pdac$SingleR.labels.bulk)
```

The bulk reference gives us somewhat more granular labels for the immune cells, and confirms the identities of the ductal / epithelial and stroma clusters. We see that the clusters denoted quiescent stellate by the single cell reference are here labeled as endothelial cells, and the activated stellate cells are confirmed to be fibroblasts.  

```{r}
p6 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.bulk") + 
      scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p6
```

```{r, echo=FALSE}
rm(sc_preds, bulk_preds, bulk_ref, sc_ref)
```

One shouldn't use `SingleR` as the final authority for cell types, but we were able to confirm the identities of the ductal and fibroblast clusters, which is important for this dataset as the cells we are most interested in are the cancer-associated fibroblasts (CAFs). 

# CONICSmat CNV Estimation

Next we'll attempt to identify malignant cells using single-cell copy number variation estimation as implemented in the `CONCISmat` package. Details of the GMM methodology used can be found [at the Diaz Lab's GitHub repository](https://github.com/diazlab/CONICS). Note: this step is memory-intensive because 1) it requires the sparse counts matrix to be cast to a dense matrix and 2) a lot of Gaussian mixture models get estimated. If your machine doesn't have a lot of RAM it might be best to skip this and manually annotate the malignant cells instead through the usage of known canonical markers or the `VAM` single-cell GSEA methodology. 

```{r}
chrom_regions <- read.table("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/chrom_arm_positions.txt", 
                            sep = "\t", 
                            row.names = 1, 
                            header = TRUE)
gene_pos <- getGenePositions(rownames(pdac))
cpm <- t(t(as.matrix(pdac@assays$SCT@counts)) / colSums(as.matrix(pdac@assays$SCT@counts))) * 10^5
cpm <- log2(cpm + 1)
norm_factor <- calcNormFactors(cpm)
cnv_est <- plotAll(mat = cpm, 
                   normFactor = norm_factor, 
                   regions = chrom_regions, 
                   gene_pos = gene_pos, 
                   fname = "Elyada")
```

## Visualizing CNVs

After estimating CNVs, we cluster the cells into $k = 3$ clusters, with the hope of finding one large cluster of normal cells and two smaller clusters composed of CAFs and PDAC cells. 

```{r}
bic_table <- read.table("./Elyada_BIC_LR.txt", 
                        sep = "\t", 
                        row.names = 1, 
                        header = TRUE, 
                        check.names = FALSE)
cand_regions <- rownames(bic_table[bic_table$`BIC difference` > 1000 & bic_table$`LRT adj. p-val` < .01, ])
hist1 <- plotHistogram(cnv_est[, cand_regions], 
                       cpm, 
                       clusters = 3, 
                       zscoreThreshold = 3, 
                       celltypes = pdac$SingleR.labels.bulk, 
                       patients = pdac$sample)
```

We add the normal vs. malignant cell labels in to our `Seurat` object's metadata, then visualize the results. As expected, the malignant cells are located in the clusters identified by `SingleR` as ductal cells and fibroblasts. This indicates that `CONICSmat` did a solid job of estimating the CNVs - no easy feat with sparse, noisy single-cell data. 

```{r}
normal <- which(hist1 == 2)
malignant <- which(hist1 != 2)
pdac@meta.data$malig <- ifelse(rownames(pdac@meta.data) %in% names(normal), "Normal", "Malignant")
p7 <- DimPlot(pdac, reduction = "tsne", group.by = "malig") + 
      scale_color_manual(values = wes_palette("Zissou1", n = 5)[c(5, 2)]) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p7
```

```{r, echo=FALSE}
rm(cpm, bic_table, chrom_regions, cnv_est, gene_pos, cand_regions, hist1, normal, norm_factor, malignant)
gc()  # not recommended, but it could help this document compile, so ...
```

# DECODER

Next, we use [Dr. Xianlu Peng's DECODER](https://github.com/laurapeng/decoderr) in order to deconvolve the dataset and assign weights to each cell using non-negative matrix factorization. We calculate basal & classical PDAC, normal & activated stroma, immune, and endocrine & exocrine pancreas compartment weights.

```{r}
sample_wts_unscaled <- Decon_single_sample("TCGA_RNAseq_PAAD", pdac@assays$SCT@data, "geneSymbol")
sample_wts <- Norm_PDAC_weights(sample_wts_unscaled)
pdac <- AddMetaData(pdac, sample_wts$Immune, "immune")
pdac <- AddMetaData(pdac, sample_wts$bcRatio, "bc_ratio")
pdac <- AddMetaData(pdac, sample_wts$Exocrine, "exocrine")
pdac <- AddMetaData(pdac, sample_wts$Endocrine, "endocrine")
pdac <- AddMetaData(pdac, sample_wts_unscaled[, 9], "basal")
pdac <- AddMetaData(pdac, sample_wts_unscaled[, 5], "classical")
pdac <- AddMetaData(pdac, sample_wts$NormalStroma, "norm_stroma")
pdac <- AddMetaData(pdac, sample_wts$ActivatedStroma, "act_stroma")
```

## Visualizing DECODER Weights

### Basal PDAC

The basal weights are highest in a subcluster of the ductal group identified through `SingleR`. This is interesting as the authors did not find evidence of the basal subtype in their paper. 

```{r}
p8 <- FeaturePlot(pdac, reduction = "tsne", features = "basal") + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p8
```

### Classical PDAC

The classical weights are highest in another subcluster of the ductal cluster, and cells with high classical weights do not collocate with those that have high basal weights. The putative classical and basal PDAC cells also align closely with the cells identified through `CONCISmat` as malignant. 

```{r}
p9 <- FeaturePlot(pdac, reduction = "tsne", features = "classical") + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p9
```

### Exocrine Pancreas

The cluster identified through `SingleR` as acinar cells is the only cluster with high exocrine pancreas weights. 

```{r}
p10 <- FeaturePlot(pdac, reduction = "tsne", features = "exocrine") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p10
```

### Endocrine Pancreas

No cells have high endocrine pancreas weights. 

```{r}
p11 <- FeaturePlot(pdac, reduction = "tsne", features = "endocrine") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p11
```

### Immune

Once again, we confirm the largeness of the immune cell population in this dataset. 

```{r}
p12 <- FeaturePlot(pdac, reduction = "tsne", features = "immune") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p12
```

### Normal Stroma

Cells with high normal stroma stroma weights are located in the cluster identified by `SingleR` as being composed of endothelial cells. 

```{r}
p13 <- FeaturePlot(pdac, reduction = "tsne", features = "norm_stroma") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p13
```

### Activated Stroma

Cells with high activated stroma weights are also located in the fibroblast cluster, and do not intersect with the cells that have high normal stroma scores. This indicates that `SCISSORS` will most likely perform well on the fibroblast cluster and be able to quickly tease out the cell subtypes.

```{r}
p14 <- FeaturePlot(pdac, reduction = "tsne", features = "act_stroma") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p14
```

# SCISSORS

Now that we have rough labels from `SingleR`, CNVs from `CONICSmat`, and compartment weights from `DECODER`, we should have more than enough cell-level metadata to look for and annotate cell subtypes using `SCISSORS`. 

## Fibroblasts

The fibroblast marker genes provided by Elyada *et al* match the `SingleR` results defining cluster 8 as containing fibroblasts. 

```{r}
p15 <- FeaturePlot(pdac, reduction = "tsne", features = "COL1A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL1A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p16 <- FeaturePlot(pdac, reduction = "tsne", features = "COL3A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL3A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p17 <- FeaturePlot(pdac, reduction = "tsne", features = "LUM") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "LUM") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p18 <- FeaturePlot(pdac, reduction = "tsne", features = "DCN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "DCN") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p15 | p16) / (p17 | p18)
```

### Reclustering

Here we use `ReclusterCells()` to identify subclusters within cluster 7. We find four distinct subclusters.

```{r, warning=FALSE}
fibro <- ReclusterCells(pdac,
                        which.clust = 8,
                        n.variable.genes = 4000, 
                        n.PC = 20, 
                        resolution.vals = c(.1, .2, .3, .4), 
                        k.vals = c(5, 10, 20, 30), 
                        which.dim.reduc = "tsne", 
                        redo.embedding = TRUE, 
                        do.plot = FALSE)
fibro_pc <- Embeddings(fibro, "pca")
```

We'll again run Fit-SNE on the reclustered cells, for consistencies sake. 

```{python}
# import data
fibro_pc = r.fibro_pc
# run Fit-SNE
affin_fibro = PerplexityBasedNN(fibro_pc, perplexity=40, metric='cosine', random_state=629)
init = initialization.pca(fibro_pc, random_state=629)
tsne_f = TSNEEmbedding(init, affin_fibro, negative_gradient_method='fft')
embed_f1 = tsne_f.optimize(n_iter=250, exaggeration=5, momentum=0.4) 
embed_f2 = embed_f1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

Pulling the results back into R and visualizing them shows clear separation between the subclusters. There's some noise in subcluster 0, but other than that the reembedding looks solid. 

```{r}
embed_fibro <- as.matrix(py$embed_f2)
rownames(embed_fibro) <- colnames(fibro)
fibro@reductions$bh_tsne <- fibro@reductions$tsne
fibro@reductions$tsne<- CreateDimReducObject(embeddings = embed_fibro, 
                                             key = "FitSNE_", 
                                             assay = "SCT", 
                                             global = TRUE)
p19 <- DimPlot(fibro, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p19
```

### Cell Type Identification

Next, we use the `VAM` method of single cell gene set enrichment analysis to determine which clusters are enriched for the iCAF and myCAF marker genes, as well as the general pan-CAF marker set. We use the marker genes identified by the authors. 

```{r}
icaf_genes <- c("IL6", "PDGFRA", "CXCL12", "CFD", "LMNA", 
                "AGTR1", "HAS1", "CXCL1", "CXCL2", "CCL2", "IL8")
mycaf_genes <- c("ACTA2", "TAGLN", "MMP11", "MYL9", 
                 "HOPX", "POSTN", "TPM1", "TPM2")
pan_caf_genes <- c("COL1A1", "FAP", "PDPN", "DCN", "VIM")
gene_sets <- list(icaf_genes, mycaf_genes, pan_caf_genes)
names(gene_sets) <- c("iCAF", "myCAF", "Pan-CAF")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
```

#### iCAFs & myCAFs

We can easily define cluster 0 as the myCAF population, and clusters 1 and 2 as the slightly smaller iCAF population. Interestingly, cluster 3 shows no enrichment whatsoever for either the iCAF or myCAF gene sets. 

```{r}
p20 <- FeaturePlot(fibro, reduction = "tsne", features = "iCAF") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "iCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p21 <- FeaturePlot(fibro, reduction = "tsne", features = "myCAF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "myCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p20 | p21) / p19
```

#### apCAFs

Interestingly, we have a small cluster of 22 cells that does not appear to express any of the fibroblast, CAF, perivascular, or endothelial markers. We'll use differential expression testing to find out what makes it unique, and hopefully annotate it. After performing differential expression analysis, we found that the top 3 markers for cluster 4 are CLU, CD74, and CRYAB. Interestingly, CLU and CD74 were found to be differentially expressed in the apCAF population discovered in the KPC mouse models of CAFs in Elyada *et al*.

```{r}
DefaultAssay(fibro) <- "SCT"
fibro_markers <- FindAllMarkers(fibro, 
                                assay = "SCT", 
                                logfc.threshold = 1.5, 
                                test.use = "wilcox", 
                                only.pos = TRUE, 
                                random.seed = 629, 
                                verbose = FALSE)
fibro_markers %>% 
  filter(cluster == 4) %>% 
  arrange(desc(avg_logFC))
```

We re-run GSEA, again using the `VAM` package and the differentially expressed genes for the apCAF population as defined in Elyada *et al* (with the mouse gene names converted to HGNC symbols). We can see that the apCAF pathway is strongly enriched in cluster 3 as compared to the other CAF clusters.

```{r}
apcaf_genes <- c("HLA-DQB1", "CD74", "SAA3P", "SLPI")
gene_sets <- list(icaf_genes, mycaf_genes, apcaf_genes, pan_caf_genes)
names(gene_sets) <- c("iCAF", "myCAF", "apCAF", "Pan-CAF")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
p22 <- FeaturePlot(fibro, reduction = "tsne", features = "apCAF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "apCAF Enrichment") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p22 / p19
```

### Visualization

Finally, we add cell labels to our identified clusters.

```{r}
fibro$label <- case_when(fibro$seurat_clusters == 0 ~ "myCAF", 
                         fibro$seurat_clusters == 1 ~ "iCAF", 
                         fibro$seurat_clusters == 2 ~ "iCAF", 
                         fibro$seurat_clusters == 3 ~ "apCAF")
p23 <- DimPlot(fibro, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank())
p23
```

## Normal Stroma Cells

We hypothesize that the cluster identified as endothelial cells by `SingleR` contains our normal stroma population, and potentially some other cell types as it contains visible subclusters. 

### Reclustering

```{r}
endo <- ReclusterCells(pdac, 
                       which.clust = 12, 
                       n.variable.genes = 4000, 
                       n.PC = 20, 
                       which.dim.reduc = "tsne", 
                       redo.embedding = TRUE, 
                       random.seed = 629)
endo_pc <- Embeddings(endo, "pca")
```

For consistencies' sake, we'll run Fit-SNE on the reclustered object. 

```{python}
# import data
endo_pc = r.endo_pc
# run Fit-SNE
affin_endo = PerplexityBasedNN(endo_pc, perplexity=60, metric='cosine', random_state=629)
init = initialization.pca(endo_pc, random_state=629)
tsne_e = TSNEEmbedding(init, affin_endo, negative_gradient_method='fft')
embed_e1 = tsne_e.optimize(n_iter=250, exaggeration=10, momentum=0.6) 
embed_e2 = embed_e1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

```{r}
embed_endo <- as.matrix(py$embed_e2)
rownames(embed_endo) <- colnames(endo)
endo@reductions$bh_tsne <- endo@reductions$tsne
endo@reductions$tsne<- CreateDimReducObject(embeddings = embed_endo, 
                                            key = "FitSNE_", 
                                            assay = "SCT", 
                                            global = TRUE)
p24 <- DimPlot(endo, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p24
```

### Perivascular Cells

By examining the marker genes provided in the Elyada paper, we define cluster 0 as being composed of perivascular cells. 

```{r}
p25 <- FeaturePlot(endo, reduction = "tsne", features = "IGFBP7") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "IGFBP7") + 
       theme_yehlab() + 
       NoLegend() +
       theme(axis.title = element_blank())
p26 <- FeaturePlot(endo, reduction = "tsne", features = "ACTA2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "ACTA2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p27 <- FeaturePlot(endo, reduction = "tsne", features = "RGS5") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "RGS5") + 
       theme_yehlab() + 
       NoLegend() +
       theme(axis.title = element_blank())
(p25 | p26 | p27) / p24
```

### Endothelial Cells

Next, we use the gene markers PLVAP and VWF (again provided in Elyada *et al*) to identify cluster 1 as the endothelial cells.

```{r}
p28 <- FeaturePlot(endo, reduction = "tsne", features = "PLVAP") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "PLVAP") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p29 <- FeaturePlot(endo, reduction = "tsne", features = "VWF") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "VWF") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p28 | p29) / p24
```

### Visualization

First we add labels to our two clusters, then we visualize the results. 

```{r}
endo$label <- case_when(endo$seurat_clusters == 0 ~ "Perivascular", 
                        endo$seurat_clusters == 1 ~ "Endothelial")
p30 <- DimPlot(endo, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank())
p30
```

## T Cells

Going back to the main `Seurat` object, we should have a large population of T and NK cells that warrants further investigation. Using CD3D expression we can easily identify clusters 0, 3, and 7 as the mixed T & NK cells. We already see some good separation, so reclustering the cells should have good results. 

```{r}
p31 <- FeaturePlot(pdac, reduction = "tsne", features = "CD3D") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p31 / p4
```

### Tumor

```{r}
nkt_tumor <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 7) & condition == "PDAC")
```

#### Reclustering

Here we run `ReclusterCells()`, while treating the NK & T cell clusters as one large group. This will hopefully allow use to elucidate T cell subtypes. We use fewer PCs for these cells since the differences between immune cells are subtle, and adding more PCs will most likely only contribute noise to our analyses. 

```{r}
nkt_tumor <- ReclusterCells(nkt_tumor, 
                            which.clust = list(0, 3, 7), 
                            merge.clusters = TRUE, 
                            n.variable.genes = 4000, 
                            n.PC = 15, 
                            k.vals = c(30, 40, 50, 60), 
                            resolution.vals = c(.1, .2, .3), 
                            which.dim.reduc = "tsne", 
                            redo.embedding = TRUE, 
                            random.seed = 629)
nkt_tumor_pc <- Embeddings(nkt_tumor, "pca")
```

Once again we'll run Fit-SNE on the reclustered cells. 

```{python}
# import data
nkt_tumor_pc = r.nkt_tumor_pc
# run Fit-SNE
affin_nkt_tumor = PerplexityBasedNN(nkt_tumor_pc, perplexity=180, metric='cosine', random_state=629)
init = initialization.pca(nkt_tumor_pc, random_state=629)
tsne_nkt_tumor = TSNEEmbedding(init, affin_nkt_tumor, negative_gradient_method='fft')
embed_nkt_tumor1 = tsne_nkt_tumor.optimize(n_iter=250, exaggeration=4, momentum=0.6) 
embed_nkt_tumor2 = embed_nkt_tumor1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_nkt_tumor.set_perplexity(75)
embed_nkt_tumor3 = embed_nkt_tumor2.optimize(n_iter=3, exaggeration=2, momentum=0.1)
embed_nkt_tumor4 = embed_nkt_tumor3.optimize(n_iter=500, exaggeration=1, momentum=0.6)
```

The reembedding isn't perfect, which we somewhat expected as immune cells are difficult to tell apart based on the transcriptome alone. 

```{r}
embed_nkt_tumor <- as.matrix(py$embed_nkt_tumor4)
rownames(embed_nkt_tumor) <- colnames(nkt_tumor)
nkt_tumor@reductions$bh_tsne <- nkt_tumor@reductions$tsne
nkt_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_tumor, 
                                                 key = "FitSNE_", 
                                                 assay = "SCT", 
                                                 global = TRUE)
p32 <- DimPlot(nkt_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p32
```

#### Cell Type Identification

##### CD4+ T

CLusters 0 and 1 contain our CD4+ T cells, which we characterize using IL7R as we did in the PBMC3k vignette. It's somewhat outside of our scope here to determine which subtype each cluster belongs to, so we'll simply assign both clusters the same label and move on. 

```{r}
p33 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL7R") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p34 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD69") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p33 | p34) / p32
```

##### T-regs

Cluster 3 contains the regulatory T cells. 

```{r}
p35 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL2RA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p36 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "FOXP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p35 | p36) / p32
```

##### Proliferating T-regs

We can find the proliferating T-regs in cluster 7. 

```{r}
p37 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TOP2A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p37 / p32
```

##### Mast

Mast cells can be identified using TPSAB1 expression in cluster 6. 

```{r}
p38 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TPSAB1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p38 / p32
```

##### NK

NKG7 and PRF1 show us the NK cells in cluster 5. 

```{r}
p39 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "NKG7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p40 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "PRF1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p39 | p40) / p32
```

##### CD8+ T

We use CD8A and CD2 to reveal the CD8+ T cells within clusters 2 and 4. 

```{r}
p41 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD8A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p42 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p41 | p42) / p32
```

#### Intermediate Monocyte

Lastly, we have cluster 8, which doesn't highly express our pan-T/NK cell markers CD3D and CD2. It could be a myeloid cell cluster that was mistakenly grouped with the T/NK cells. We'll start with a Wilcoxon test to determine its differentially expressed genes. 

```{r}
clust8_markers <- FindAllMarkers(nkt_tumor, 
                                 test.use = "wilcox",
                                 min.diff.pct = .2,
                                 logfc.threshold = .5, 
                                 verbose = FALSE, 
                                 random.seed = 629) %>% 
                  filter(cluster == 8, p_val_adj < .05) %>% 
                  arrange(desc(1 - p_val_adj))
clust8_markers
```

Interestingly, several [intermediate monocyte markers](https://www.frontiersin.org/articles/10.3389/fimmu.2019.02035/full#B3) - LYZ, HLA-DRA, CD74, and HLA-DPB1 - are differentially expressed in cluster 8. 

We'll plot some of those markers below. 

```{r}
p43 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "LYZ") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p44 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "HLA-DRA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p45 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD74") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p46 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "HLA-DPB1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
((p43 | p44) / (p45 | p46)) / p32
```


#### Visualization

We add labels to our T cell `Seurat` object and visualize the results. 

```{r}
nkt_tumor$label <- case_when(nkt_tumor$seurat_clusters == 0 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 1 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 2 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 3 ~ "T-reg", 
                             nkt_tumor$seurat_clusters == 4 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 5 ~ "NK", 
                             nkt_tumor$seurat_clusters == 6 ~ "Mast", 
                             nkt_tumor$seurat_clusters == 7 ~ "Proliferating T-reg", 
                             nkt_tumor$seurat_clusters == 8 ~ "Intermediate Monocyte")
p47 <- DimPlot(nkt_tumor, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p47
```

### Adjacent Normal

```{r}
nkt_norm <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 7) & condition == "AdjNorm")
```

#### Reclustering

```{r}
nkt_norm <- ReclusterCells(nkt_norm, 
                           which.clust = list(0, 3, 7), 
                           merge.clusters = TRUE, 
                           n.variable.genes = 4000, 
                           n.PC = 10, 
                           k.vals = c(5, 10, 15, 20), 
                           resolution.vals = c(.1, .2, .3), 
                           which.dim.reduc = "tsne", 
                           redo.embedding = TRUE, 
                           random.seed = 629)
nkt_norm_pc <- Embeddings(nkt_norm, "pca")
```

As with the other reclusterings, we'll run Fit-SNE in order to (hopefully) obtain a better low-dimensional embedding of our cells.

```{python}
# import data
nkt_norm_pc = r.nkt_norm_pc
# run Fit-SNE
affin_nkt_norm = PerplexityBasedNN(nkt_norm_pc, perplexity=100, metric='cosine', random_state=629)
init = initialization.pca(nkt_norm_pc, random_state=629)
tsne_nkt_norm = TSNEEmbedding(init, affin_nkt_norm, negative_gradient_method='fft')
embed_nkt_norm1 = tsne_nkt_norm.optimize(n_iter=250, exaggeration=12, momentum=0.6) 
embed_nkt_norm2 = embed_nkt_norm1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_nkt_norm.set_perplexity(20)
embed_nkt_norm3 = embed_nkt_norm2.optimize(n_iter=20, exaggeration=2, momentum=0.7)
embed_nkt_norm4 = embed_nkt_norm3.optimize(n_iter=500)
```

```{r}
embed_nkt_norm <- as.matrix(py$embed_nkt_norm4)
rownames(embed_nkt_norm) <- colnames(nkt_norm)
nkt_norm@reductions$bh_tsne <- nkt_norm@reductions$tsne
nkt_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_norm, 
                                                key = "FitSNE_", 
                                                assay = "SCT", 
                                                global = TRUE)
p48 <- DimPlot(nkt_norm, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p48
```

#### Cell Type Identification

First we'll use a Wilcoxon test to determine which genes characterize each cluster. 

```{r}
nkt_norm_markers <- FindAllMarkers(nkt_norm, 
                                   logfc.threshold = .5, 
                                   min.diff.pct = .2, 
                                   verbose = FALSE, 
                                   only.pos = TRUE, 
                                   random.seed = 629) %>% 
                    filter(p_val_adj < .05)
nkt_norm_markers  %>% 
  group_by(cluster) %>% 
  top_n(n = 2, wt = avg_logFC)
```

##### CD4+ T

We can use IL7R and S100A4 expression to identify the memory CD4+ T cells in cluster 0. IL7R and CCR7 identify the naive CD4+ T cells in the adjacent cluster 4. 

```{r}
p49 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "IL7R") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p50 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "S100A4") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p51 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "CCR7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p50 | p49 | p51) / p48
```

##### CD8+ T

Next we reveal the CD8+ T cells in clusters 1 and 2 with CD8A, as per usual. 

```{r}
p52 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "CD8A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p52 / p48
```

##### T-reg

The T-reg cluster, cluster 5, is identified using TIGIT and FOXP3. 

```{r}
p53 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TIGIT") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p54 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "FOXP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p53 | p54) / p48
```

##### NK

The natural killers can be found in cluster 4 through their expression of PRF1 and NKG7. The NKG7 expression also confirms the identities of the CD8+ T cells we just annotated. 

```{r}
p55 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "PRF1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p56 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "NKG7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p55 | p56) / p48
```

##### Proliferating T-reg

The tiny proliferating T-reg population in cluster 7 is characterized by TOP2A. 

```{r}
p57 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TOP2A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p57 / p48
```

#### Intermediate Monocyte

After running another Wilcoxon differential expression test specifically for the last cluster, cluster 6, we can see that several of the myeloid marker genes from Elyada *et al* are upregulated - CD68, LYZ, CD14, HLA-DRA. LST1 is also significantly differentially expressed; [LST1 is involved in monocyte differentiation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4015914/). Lastly, TYROBP, aka DAP12, [is associated with inflammation and is highly expressed in activated monocytes](https://www.sciencedirect.com/science/article/abs/pii/S0889159111000742). 

```{r}
FindMarkers(nkt_norm, 
            ident.1 = 6, 
            only.pos = TRUE, 
            min.pct = .3, 
            verbose = FALSE, 
            random.seed = 629) %>% 
  filter(p_val_adj < .05) 
```

We plot some of the intermediate monocyte marker genes. We can say with reasonable confidence that cluster 6 contains intermediate monocytes. 

```{r}
p58 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "LYZ") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p59 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "HLA-DRA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p60 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "LST1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p61 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TYROBP") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
((p58 | p59) / (p60 | p61)) / p48
```

#### Visualization

We add final cell labels to our 8 cell clusters. 

```{r}
nkt_norm$label <- case_when(nkt_norm$seurat_clusters == 0 ~ "Memory CD4+ T", 
                            nkt_norm$seurat_clusters == 1 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 2 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 3 ~ "Naive CD4+ T",
                            nkt_norm$seurat_clusters == 4 ~ "NK",
                            nkt_norm$seurat_clusters == 5 ~ "T-reg", 
                            nkt_norm$seurat_clusters == 6 ~ "Intermediate Monocyte", 
                            nkt_norm$seurat_clusters == 7 ~ "Proliferating T-reg")
p62 <- DimPlot(nkt_norm, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p62
```

## Ductal Cells

### Reclustering

We use KRT8 expression to show the ductal cells residing in clusters 5 and 10. We note that some regions of clusters 5 and 10 have no KRT8 expression, which likely means that they are composed of different cell types. 

```{r}
p63 <- FeaturePlot(pdac, reduction = "tsne", features = "KRT8") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p63 / p4
```

We run `SCISSORS`, then use a Wilcoxon differential expression test to determine potential markers for each cluster. 

```{r}
ductal <- ReclusterCells(pdac, 
                         which.clust = list(5, 10), 
                         merge.clusters = TRUE, 
                         n.variable.genes = 4000, 
                         n.PC = 20, 
                         k.vals = c(10, 20, 30, 40), 
                         resolution.vals = c(.1, .2, .3, .4), 
                         which.dim.reduc = "tsne", 
                         redo.embedding = TRUE, 
                         do.plot = FALSE, 
                         random.seed = 629)
```

```{r}
ductal_markers <- FindAllMarkers(ductal, 
                                 logfc.threshold = .5, 
                                 min.diff.pct = .2, 
                                 only.pos = TRUE, 
                                 verbose = FALSE, 
                                 random.seed = 629) %>% 
                  filter(p_val_adj < .05)
ductal_markers %>% 
  group_by(cluster) %>% 
  top_n(n = 3, wt = avg_logFC)
```

Again, we'll run Fit-SNE on the reclustered cells. 

```{r}
duct_pc <- Embeddings(ductal, reduction = "pca")
```

```{python}
# import data
duct_pc = r.duct_pc
# run Fit-SNE
affin_duct = PerplexityBasedNN(duct_pc, perplexity=200, metric='cosine', random_state=629)
init = initialization.pca(duct_pc, random_state=629)
tsne_duct = TSNEEmbedding(init, affin_duct, negative_gradient_method='fft')
embed_d1 = tsne_duct.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_d2 = embed_d1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_duct.set_perplexity(30)
embed_d3 = embed_d2.optimize(n_iter=60, exaggeration=2, momentum=0.8)
embed_d4 = embed_d3.optimize(n_iter=500)
```

We pull the results into R, making sure to save the Barnes-Hut t-SNE results in another reduction slot.

```{r}
embed_duct <- as.matrix(py$embed_d4)
rownames(embed_duct) <- colnames(ductal)
ductal@reductions$bh_tsne <- ductal@reductions$tsne
ductal@reductions$tsne<- CreateDimReducObject(embeddings = embed_duct, 
                                              key = "FitSNE_", 
                                              assay = "SCT", 
                                              global = TRUE)
p64 <- DimPlot(ductal, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p64
```

### Cell Type Identification 

#### Lipid Processing

We use ANPEP expression to identify the lipid processing ductal cells in cluster 7. 

```{r}
p65 <- FeaturePlot(ductal, reduction = "tsne", features = "ANPEP") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p65 / p64
```

#### Secretory

Expression of SOD3 and CFTR reveals the secretory cells in cluster 1.

```{r}
p66 <- FeaturePlot(ductal, reduction = "tsne", features = "SOD3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p67 <- FeaturePlot(ductal, reduction = "tsne", features = "CFTR") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p66 | p67) / p64
```

#### Classical 1

We use TFF1 and TFF2 expression to annotate the classical 1 epithelial cells in clusters 0 and 5. We can also see that the two classical 1 clusters are split by the sample from which the cells originate. Going forward, we'll simply label both clusters as classical 1. 

```{r}
p68 <- FeaturePlot(ductal, reduction = "tsne", features = "TFF1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p69 <- FeaturePlot(ductal, reduction = "tsne", features = "TFF2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p70 <- DimPlot(ductal, reduction = "tsne", group.by = "sample") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Sample") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p68 | p70 | p69) / p64
```

#### Classical 2

The classical 2 cells are located in cluster 6, as evidenced by their expression of CRISP3. 

```{r}
p71 <- FeaturePlot(ductal, reduction = "tsne", features = "CRISP3") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p71 / p64
```

#### Basal-like

The basal compartment weights from `DECODER` are highest in cluster 3, which we will denote as being composed of basal-like PDAC. 

```{r}
p72 <- FeaturePlot(ductal, reduction = "tsne", features = "basal") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Basal-like") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p72 / p64
```

#### CD16+ Monocyte

As seen earlier in the results of the differential expression text, the top three genes for cluster 2 were HLA-DRA, HLA-DPA1, and HLA-DPB1. This points to cluster 2 being some kind of myeloid cell. We use expression of CD14, CD16 aka FCGR3A, and MS4A7 to identify it as containing CD14+ CD16+ monocytes. 

```{r}
p73 <- FeaturePlot(ductal, reduction = "tsne", features = "CD14") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p74 <- FeaturePlot(ductal, reduction = "tsne", features = "FCGR3A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p75 <- FeaturePlot(ductal, reduction = "tsne", features = "MS4A7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p73 | p74 | p75) / p64
```

All three genes are significantly differentially expressed in cluster 2. 

```{r}
clust2_markers <- FindMarkers(ductal, 
                              ident.1 = 2, 
                              random.seed = 629, 
                              only.pos = TRUE)
clust2_markers %>% 
  filter(rownames(clust2_markers) %in% c("CD14", "MS4A7", "FCGR3A")) %>% 
  arrange(desc(1 - p_val_adj))
```

#### Acinar

Lastly, we show that cluster 4 is composed of acinar cells using CTRB2.

```{r}
p76 <- FeaturePlot(ductal, reduction = "tsne", features = "CTRB2") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p76 / p64
```

### Visualization

We add final cluster labels to our `Seurat` object and visualize the results.

```{r}
ductal$label <- case_when(ductal$seurat_clusters == 0 ~ "Classical 1", 
                          ductal$seurat_clusters == 1 ~ "Secretory", 
                          ductal$seurat_clusters == 2 ~ "CD16+ Monocyte", 
                          ductal$seurat_clusters == 3 ~ "Basal-like", 
                          ductal$seurat_clusters == 4 ~ "Acinar", 
                          ductal$seurat_clusters == 5 ~ "Classical 1", 
                          ductal$seurat_clusters == 6 ~ "Classical 2", 
                          ductal$seurat_clusters == 7 ~ "Lipid proc.")
p77 <- DimPlot(ductal, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p77
```

## Plasma Cells & Plasmacytoid DCs

We'll run `SCISSORS` on cluster 11 in order to reveal its subgroups, the plasma cells and plasmacytoid DCs, as defined by their expression of IGJ (aka JCHAIN) and IRF7, respectively. 

```{r}
p78 <- FeaturePlot(pdac, reduction = "tsne", features = "JCHAIN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "JCHAIN / IGJ") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p79 <- FeaturePlot(pdac, reduction = "tsne", features = "IRF7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p78 | p79) / p4
```

### Reclustering

We run `SCISSORS` with default parameters. 

```{r}
plas <- ReclusterCells(pdac, 
                       which.clust = 11, 
                       n.PC = 20, 
                       which.dim.reduc = "tsne",
                       redo.embedding = TRUE, 
                       do.plot = FALSE, 
                       random.seed = 629)
plas_pc <- Embeddings(plas, reduction = "pca")
```

Even though in this case it's probably not necessary, we'll run Fit-SNE on the cells just so that our axis labels are consistent. 

```{python}
# import data
plas_pc = r.plas_pc
# run Fit-SNE
affin_plas = PerplexityBasedNN(plas_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(plas_pc, random_state=629)
tsne_plas = TSNEEmbedding(init, affin_plas, negative_gradient_method='fft')
embed_p1 = tsne_plas.optimize(n_iter=250, exaggeration=12, momentum=0.6)
embed_p2 = embed_p1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

```{r}
embed_plas <- as.matrix(py$embed_p2)
rownames(embed_plas) <- colnames(plas)
plas@reductions$bh_tsne <- plas@reductions$tsne
plas@reductions$tsne<- CreateDimReducObject(embeddings = embed_plas, 
                                            key = "FitSNE_", 
                                            assay = "SCT", 
                                            global = TRUE)
p80 <- DimPlot(plas, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p80
```

### Cell Type Identification

We use the aforementioned markers to identify our two clusters. 

```{r}
p81 <- FeaturePlot(plas, reduction = "tsne", features = "JCHAIN") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "JCHAIN") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p82 <- FeaturePlot(plas, reduction = "tsne", features = "IRF7") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p81 | p82) / p80
```

### Visualization

As before, we add cell labels and visualize the results. 

```{r}
plas$label <- case_when(plas$seurat_clusters == 0 ~ "Plasma", 
                        plas$seurat_clusters == 1 ~ "pDC")
p83 <- DimPlot(plas, reduction = "tsne", group.by = "label") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p83
```

## B Cells

We can identify the B cells in cluster 4 using joint expression of MS4A1 and CD79A. The cluster is split into two subclusters by tissue type: adjacent normal and PDAC. While it would be interesting to determine the genetic drivers of that separation, it's somewhat outside of our scope here. 

```{r}
p84 <- FeaturePlot(pdac, reduction = "tsne", features = "MS4A1") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MS4A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p85 <- FeaturePlot(pdac, reduction = "tsne", features = "CD79A") +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD79A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p86 <- DimPlot(pdac, reduction = "tsne", group.by = "condition") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Tissue Type") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p84 | p86 | p85) / p4
```

## Myeloid Cells

Lastly, we'll split up the myeloid population by tissue condition (PDAC vs. adjacent normal) just like we did with the NK / T cells. We'll run `SCISSORS`, annotate the clusters, and visualize the results. 

### Tumor 

First we'll attempt to assign broad cell type labels to each of the four putative myeloid clusters. We'll use the marker genes from Elyada *et al* once again. 

```{r}
myo_tumor <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 6, 9) & condition == "PDAC")
p87 <- DimPlot(myo_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p87
```

Cluster 1 appears to be composed of our resident macrophages due to its expression of CD14 and C1QA. 

```{r}
p88 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "CD14") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p89 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "C1QA") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p88 | p89) / p87
```

We can use LYZ and S100A8 expression to reveal the classic monocytes and neutrophils in cluster 2. 

```{r}
p90 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "LYZ") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LYZ") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p91 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "S100A8") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p90 | p91) / p87
```

Cluster 9 seems to house the alternatively activated macrophages, as shown by expression of SPP1. 

```{r}
p92 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "SPP1") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "SPP1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p92 / p87
```

Lastly, we show that cluster 6 contains our various DC subtypes through its expression of FCER1A, a canonical dendritic cell marker. 

```{r}
p93 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "FCER1A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FCER1A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p93 / p87
```

#### Reclustering

Since we've already annotated some of the myeloid cell types, we'll focus on cluster 2, which contains both monocytes and neutrophils, and the dendritic cells in cluster 6. 

```{r}
myo_reclust <- ReclusterCells(myo_tumor, 
                              which.clust = c(2, 6), 
                              n.PC = 10, 
                              merge.clusters = FALSE, 
                              k.vals = c(10, 20, 30, 40), 
                              resolution.vals = c(.2, .3, .4), 
                              n.variable.genes = 4000, 
                              which.dim.reduc = "tsne", 
                              redo.embedding = TRUE, 
                              do.plot = FALSE, 
                              random.seed = 629)
mono_tumor <- myo_reclust[[1]]
dc_tumor <- myo_reclust[[2]]
```

We'll again run Fit-SNE on our reclustered cells. 

```{r}
mono_tumor_pc <- Embeddings(mono_tumor, "pca")
dc_tumor_pc <- Embeddings(dc_tumor, "pca")
```

```{python}
# import data
mono_pc = r.mono_tumor_pc
dc_pc = r.dc_tumor_pc
# Fit-SNE - monocytes
affin_mono = PerplexityBasedNN(mono_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(mono_pc, random_state=629)
tsne_mono = TSNEEmbedding(init, affin_mono, negative_gradient_method='fft')
embed_mono1 = tsne_mono.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_mono2 = embed_mono1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
# Fit-SNE - DC
affin_dc = PerplexityBasedNN(dc_pc, perplexity=60, metric='cosine', random_state=629)
init = initialization.pca(dc_pc, random_state=629)
tsne_dc = TSNEEmbedding(init, affin_dc, negative_gradient_method='fft')
embed_dc1 = tsne_dc.optimize(n_iter=250, exaggeration=8, momentum=0.6)
embed_dc2 = embed_dc1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_dc.set_perplexity(20)
embed_dc3 = embed_dc2.optimize(n_iter=100)
```

We pull the results back into R and visualize them. 

```{r}
embed_mono <- as.matrix(py$embed_mono2)
rownames(embed_mono) <- colnames(mono_tumor)
mono_tumor@reductions$bh_tsne <- mono_tumor@reductions$tsne
mono_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_mono, 
                                                  key = "FitSNE_", 
                                                  assay = "SCT",
                                                  global = TRUE)
p94 <- DimPlot(mono_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p94
```

```{r}
embed_dc <- as.matrix(py$embed_dc3)
rownames(embed_dc) <- colnames(dc_tumor)
dc_tumor@reductions$bh_tsne <- dc_tumor@reductions$tsne
dc_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_dc, 
                                                key = "FitSNE_", 
                                                assay = "SCT",
                                                global = TRUE)
p95 <- DimPlot(dc_tumor, reduction = "tsne") + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p95
```

#### Cell Type Identification

##### Monoctyes & Neutrophils

First we ID the neutrophils in cluster 1 using S100A8 and S100A9 - marker genes used by Elyada *et al*. 

```{r}
p96 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "S100A8") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p97 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "S100A9") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A9") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p96 | p97) / p94
```

We can identify the classical monocytes in cluster 0 through expression of, you guessed it, CD14, as well as expression of CD16 aka FCGR3A. 

```{r}
p98 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "CD14") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p99 <- FeaturePlot(mono_tumor, reduction = "tsne", features = "FCGR3A") + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD16") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p98 | p99) / p94
```

We add cell labels to our `Seurat` object, then we're off to the DCs. 

```{r}
mono_tumor$label <- case_when(mono_tumor$seurat_clusters == 0 ~ "Classical Monocyte", 
                              mono_tumor$seurat_clusters == 1 ~ "Neutrophil")
```

##### Dendritic Cells

We can use CLEC9A to annotate the cDC1 population in cluster 2. 

```{r}
p100 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CLEC9A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CLEC9A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p100 / p95
```

Joint expression of CD1A and CD207 reveals the Langerhans-like DCs in cluster 0. 

```{r}
p101 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CD1A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD1A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p102 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CD207") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD207") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p101 | p102) / p95
```

Next we use LAMP3 to identify the activated DCs in cluster 3. 

```{r}
p103 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "LAMP3") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LAMP3") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p103 / p95
```

Using a Wilcoxon differential expression test shows that cluster 1 is characterized by [several conventionl DC2 marker genes / transcription factors](https://onlinelibrary.wiley.com/doi/pdf/10.1111/imm.12888). 

```{r}
dc_tumor_markers <- FindAllMarkers(dc_tumor, 
                                   logfc.threshold = .3, 
                                   only.pos = TRUE, 
                                   verbose = FALSE, 
                                   random.seed = 629)
dc_tumor_markers %>% 
  filter(p_val_adj < .05 & cluster == 1 & gene %in% c("CLEC10A", "KLF4", "ZEB2"))
```

```{r}
p104 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "CLEC10A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p105 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "KLF4") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p106 <- FeaturePlot(dc_tumor, reduction = "tsne", features = "ZEB2") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p104 | p105 | p106) / p95
```

We add final subcluster cell type labels to our `Seurat` object, and then we're off to the macrophages. 

```{r}
dc_tumor$label <- case_when(dc_tumor$seurat_clusters == 0 ~ "Langerhans-like DC", 
                            dc_tumor$seurat_clusters == 1 ~ "cDC2", 
                            dc_tumor$seurat_clusters == 2 ~ "cDC1", 
                            dc_tumor$seurat_clusters == 3 ~ "Activated DC")
```

#### Visualization

##### Monocytes & Neutrophils

```{r}
p107 <- DimPlot(mono_tumor, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p107
```

##### DCs

```{r}
p108 <- DimPlot(dc_tumor, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p108
```

### Adjacent Normal

We have the same four clusters as in the tumor tissue: 1, 2, 6, and 9. We know that clusters 1 and 9 are composed of the resident and alternatively-activated macrophages, respectively. We'll focus on the monocytes & neutrophils and DCs in clusters 2 and 6, respectively. 

```{r}
myo_norm <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 6, 9) & condition == "AdjNorm")
```

#### Reclustering

```{r}
myo_norm_reclust <- ReclusterCells(myo_norm, 
                                   which.clust = c(2, 6), 
                                   merge.clusters = FALSE, 
                                   n.variable.genes = 4000, 
                                   n.PC = 10, 
                                   which.dim.reduc = "tsne", 
                                   redo.embedding = TRUE, 
                                   random.seed = 629)
mono_norm <- myo_norm_reclust[[1]]
dc_norm <- myo_norm_reclust[[2]]
```

For the last time, we run Fit-SNE in order to obtain a better embedding. 

```{r}
mono_norm_pc <- Embeddings(mono_norm, "pca")
dc_norm_pc <- Embeddings(dc_norm, "pca")
```

```{python}
# import data
mono_pc = r.mono_norm_pc
dc_pc = r.dc_norm_pc
# Fit-SNE - monocytes
affin_mono = PerplexityBasedNN(mono_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(mono_pc, random_state=629)
tsne_mono = TSNEEmbedding(init, affin_mono, negative_gradient_method='fft')
embed_mono1 = tsne_mono.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_mono2 = embed_mono1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
# Fit-SNE - DC
affin_dc = PerplexityBasedNN(dc_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(dc_pc, random_state=629)
tsne_dc = TSNEEmbedding(init, affin_dc, negative_gradient_method='fft')
embed_dc1 = tsne_dc.optimize(n_iter=250, exaggeration=8, momentum=0.6)
embed_dc2 = embed_dc1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

We pull the results back into R and visualize them. 

```{r}
embed_mono <- as.matrix(py$embed_mono2)
rownames(embed_mono) <- colnames(mono_norm)
mono_norm@reductions$bh_tsne <- mono_norm@reductions$tsne
mono_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_mono, 
                                                 key = "FitSNE_", 
                                                 assay = "SCT",
                                                 global = TRUE)
p109 <- DimPlot(mono_norm, reduction = "tsne") + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p109
```

```{r}
embed_dc <- as.matrix(py$embed_dc2)
rownames(embed_dc) <- colnames(dc_norm)
dc_norm@reductions$bh_tsne <- dc_norm@reductions$tsne
dc_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_dc, 
                                               key = "FitSNE_", 
                                               assay = "SCT",
                                               global = TRUE)
p110 <- DimPlot(dc_norm, reduction = "tsne") + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p110
```

#### Cell Type Identification

##### Monocytes & Neutrophils

We use high S100A8 and S100A9 expression to define the neutrophils in cluster 2. 

```{r}
p111 <- FeaturePlot(mono_norm, reduction = "tsne", features = "S100A8") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p112 <- FeaturePlot(mono_norm, reduction = "tsne", features = "S100A9") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p111 | p112) / p109
```

Next up are the classical monocytes in clusters 0 and 1, split by sample. 

```{r}
p113 <- FeaturePlot(mono_norm, reduction = "tsne", features = "CD68") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p114 <- FeaturePlot(mono_norm, reduction = "tsne", features = "CD14") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p115 <- DimPlot(mono_norm, reduction = "tsne", group.by = "sample") + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Sample") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p113 | p115 | p114) / p109
```

Lastly, it seems we have a small group of CD8+ T cells in cluster 3 that snuck into the myeloid cluster, as defined by their expression of CD2, CD3D (marking them as NK / T cells), and CD8A & NKG7. I don't believe they're NK cells as the do not express PRF1 or GZMB. 

```{r}
p116 <- FeaturePlot(mono_norm, reduction = "tsne", features = "CD2") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p117 <- FeaturePlot(mono_norm, reduction = "tsne", features = "CD3D") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p118 <- FeaturePlot(mono_norm, reduction = "tsne", features = "CD8A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p119 <- FeaturePlot(mono_norm, reduction = "tsne", features = "NKG7") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
((p116 | p117) / (p118 | p119)) / p109
```

We add labels to our clusters. 

```{r}
mono_norm$label <- case_when(mono_norm$seurat_clusters == 0 ~ "Classic Monocyte", 
                             mono_norm$seurat_clusters == 1 ~ "Classic Monocyte", 
                             mono_norm$seurat_clusters == 2 ~ "Neutrophil", 
                             mono_norm$seurat_clusters == 3 ~ "CD8+ T")
```

##### Dendritic Cells

We annotate cluster 1 as conventional DC1 and cluster 0 as conventional DC2 through mutually exclusive expression of the canonical markers CLEC9A and CLEC10A, respectively. 

```{r}
p120 <- FeaturePlot(dc_norm, reduction = "tsne", features = "CLEC9A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p121 <- FeaturePlot(dc_norm, reduction = "tsne", features = "CLEC10A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p120 | p121) / p110
```

Labels are added to our adjacent normal tissue DC `Seurat` object. 

```{r}
dc_norm$label <- case_when(dc_norm$seurat_clusters == 0 ~ "cDC2", 
                           dc_norm$seurat_clusters == 1 ~ "cDC1")
```

#### Visualization

Here are the final annotations for the adjacent normal tissue myeloid population. 

```{r}
p122 <- DimPlot(mono_norm, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p122
```

```{r}
p123 <- DimPlot(dc_norm, reduction = "tsne", group.by = "label") + 
        scale_color_manual(values = paletteer_d("awtools::mpalette")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 3)))
p123
```

# Conclusions

# Save Figures & Data

We'll create a quick convenience function to help us save the figures.

```{r}
saveSCISSORS <- function(plot = NULL, 
                         name = NULL, 
                         border = TRUE, 
                         pub.ready = FALSE) {
  if (is.null(plot) | is.null(name)) stop("You forgot some arguments.")
  if (pub.ready) {
    dir <- "~/Desktop/R/SCISSORS/vignettes/figures_pub/Elyada/"
    if (!border) {
      plot <- plot + 
              theme(panel.border = element_blank(), 
                    axis.title = element_blank(), 
                    legend.position = "none")
    } else {
      plot <- plot + 
              theme(axis.title = element_blank(), 
                    legend.position = "none")
    }
    ggsave(filename = paste0(name, ".pdf"), 
           device = "pdf", 
           units = "in",
           path = dir, 
           height = 5, 
           width = 5) 
  } else {
    dir <- "~/Desktop/R/SCISSORS/vignettes/figures_supp/Elyada/"
    ggsave(filename = paste0(name, ".pdf"), 
           device = "pdf", 
           units = "in",
           path = dir, 
           height = 5, 
           width = 5) 
  }
}
```

This section isn't worth reading; it's here solely to prove that the figures we present in our publication were dynamically generated during the knitting of this document.

```{r}
saveSCISSORS(plot = p0, name = "Seurat_Clusters_tSNE.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p1, name = "Seurat_Clusters_UMAP.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p4, name = "Seurat_Clusters_FitSNE.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p5, name = "SingleR_Annos_sc_ref.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p6, name = "SingleR_Annos_bulk_ref.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p7, name = "CONICSmat_Annos.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p8, name = "DECODER_basal.pdf")
saveSCISSORS(plot = p9, name = "DECODER_classical.pdf")
saveSCISSORS(plot = p10, name = "DECODER_exocrine.pdf")
saveSCISSORS(plot = p11, name = "DECODER_endocrine.pdf")
saveSCISSORS(plot = p12, name = "DECODER_immune.pdf")
saveSCISSORS(plot = p13, name = "DECODER_normal_stroma.pdf")
saveSCISSORS(plot = p14, name = "DECODER_act_stroma.pdf")
saveSCISSORS(plot = p15, name = "Fibro_all_cells_COL1A1.pdf")
saveSCISSORS(plot = p16, name = "Fibro_all_cells_COL3A1.pdf")
saveSCISSORS(plot = p17, name = "Fibro_all_cells_LUM.pdf")
saveSCISSORS(plot = p18, name = "Fibro_all_cells_DCN.pdf")
saveSCISSORS(plot = p19, name = "Fibro_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p20, name = "Fibro_iCAF_VAM.pdf")
saveSCISSORS(plot = p21, name = "Fibro_myCAF_VAM.pdf")
saveSCISSORS(plot = p22, name = "Fibro_apCAF_VAM.pdf")
saveSCISSORS(plot = p23, name = "Fibro_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p24, name = "Endo_Perivascular_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p25, name = "Perivascular_IGFBP7.pdf")
saveSCISSORS(plot = p26, name = "Perivascular_ACTA2.pdf")
saveSCISSORS(plot = p27, name = "Perivascular_RGS5.pdf")
saveSCISSORS(plot = p28, name = "Endothelial_PLVAP.pdf")
saveSCISSORS(plot = p29, name = "Endothelial_VWF.pdf")
saveSCISSORS(plot = p30, name = "Endo_Perivascular_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p31, name = "NKT_all_cells_CD3D.pdf")
saveSCISSORS(plot = p32, name = "NKT_Tumor_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p33, name = "NKT_Tumor_CD4T_IL7R.pdf")
saveSCISSORS(plot = p34, name = "NKT_Tumor_CD4T_CD69.pdf")
saveSCISSORS(plot = p35, name = "NKT_Tumor_Treg_IL2RA.pdf")
saveSCISSORS(plot = p36, name = "NKT_Tumor_Treg_FOXP3.pdf")
saveSCISSORS(plot = p37, name = "NKT_Tumor_Prolif_Treg_TOP2A.pdf")
saveSCISSORS(plot = p38, name = "NKT_Tumor_Mast_TPSAB1.pdf")
saveSCISSORS(plot = p39, name = "NKT_Tumor_NK_NKG7.pdf")
saveSCISSORS(plot = p40, name = "NKT_Tumor_NK_PRF1.pdf")
saveSCISSORS(plot = p41, name = "NKT_Tumor_CD8T_CD8A.pdf")
saveSCISSORS(plot = p42, name = "NKT_Tumor_CD8T_CD2.pdf")
saveSCISSORS(plot = p43, name = "NKT_Tumor_Intermediate_Mono_LYZ.pdf")
saveSCISSORS(plot = p44, name = "NKT_Tumor_Intermediate_Mono_HLADRA.pdf")
saveSCISSORS(plot = p45, name = "NKT_Tumor_Intermediate_Mono_CD74.pdf")
saveSCISSORS(plot = p46, name = "NKT_Tumor_Intermediate_Mono_HLADPB1.pdf")
saveSCISSORS(plot = p47, name = "NKT_Tumor_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p48, name = "NKT_Norm_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p49, name = "NKT_Norm_CD4T_IL7R.pdf")
saveSCISSORS(plot = p50, name = "NKT_Norm_CD4T_S100A4.pdf")
saveSCISSORS(plot = p51, name = "NKT_Norm_CD4T_CCR7.pdf")
saveSCISSORS(plot = p52, name = "NKT_Norm_CD8T_CD8A.pdf")
saveSCISSORS(plot = p53, name = "NKT_Norm_Treg_TIGIT.pdf")
saveSCISSORS(plot = p54, name = "NKT_Norm_Treg_FOXP3.pdf")
saveSCISSORS(plot = p55, name = "NKT_Norm_NK_PRF1.pdf")
saveSCISSORS(plot = p56, name = "NKT_Norm_NK_NKG7.pdf")
saveSCISSORS(plot = p57, name = "NKT_Norm_Prolif_Treg_TOP2A.pdf")
saveSCISSORS(plot = p58, name = "NKT_Norm_Intermediate_Mono_LYZ.pdf")
saveSCISSORS(plot = p59, name = "NKT_Norm_Intermediate_Mono_HLADRA.pdf")
saveSCISSORS(plot = p60, name = "NKT_Norm_Intermediate_Mono_LST1.pdf")
saveSCISSORS(plot = p61, name = "NKT_Norm_Intermediate_Mono_TYROBP.pdf")
saveSCISSORS(plot = p62, name = "NKT_Norm_SCISSORS_FitSNE.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p63, name = "Ductal_all_cells_KRT8.pdf")
saveSCISSORS(plot = p64, name = "Ductal_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p65, name = "Ductal_Lipid_Proc_ANPEP.pdf")
saveSCISSORS(plot = p66, name = "Ductal_Secretory_SOD3.pdf")
saveSCISSORS(plot = p67, name = "Ductal_Secretory_CFTR.pdf")
saveSCISSORS(plot = p68, name = "Ductal_Classical1_TFF1.pdf")
saveSCISSORS(plot = p69, name = "Ductal_Classical1_TFF2.pdf")
saveSCISSORS(plot = p70, name = "Ductal_Classical1_SampleID.pdf")
saveSCISSORS(plot = p71, name = "Ductal_Classical2_CRISP3.pdf")
saveSCISSORS(plot = p72, name = "Ductal_basal_DECODER.pdf")
saveSCISSORS(plot = p73, name = "Ductal_CD16_Mono_CD14.pdf")
saveSCISSORS(plot = p74, name = "Ductal_CD16_Mono_FCGR3A.pdf")
saveSCISSORS(plot = p75, name = "Ductal_CD16_Mono_MS4A7.pdf")
saveSCISSORS(plot = p76, name = "Ductal_Acinar_CTRB2.pdf")
saveSCISSORS(plot = p77, name = "Ductal_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p78, name = "Plasma_all_cells_IGJ.pdf")
saveSCISSORS(plot = p79, name = "Plasma_all_cells_IRF7.pdf")
saveSCISSORS(plot = p80, name = "Plasma_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p81, name = "Plasma_Plasma_JCHAIN.pdf")
saveSCISSORS(plot = p82, name = "Plasma_pDC_IGJ.pdf")
saveSCISSORS(plot = p83, name = "Plasma_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p84, name = "B_all_cells_MS4A1.pdf")
saveSCISSORS(plot = p85, name = "B_all_cells_CD79A.pdf")
saveSCISSORS(plot = p86, name = "B_all_cells_tissue_type.pdf")
saveSCISSORS(plot = p87, name = "Myeloid_Tumor_original_FitSNE.pdf")
saveSCISSORS(plot = p88, name = "Myeloid_Tumor_Resident_Macro_CD14.pdf")
saveSCISSORS(plot = p89, name = "Myeloid_Tumor_Resident_Macro_C1QA.pdf")
saveSCISSORS(plot = p90, name = "Myeloid_Tumor_Classic_Mono_LYZ.pdf")
saveSCISSORS(plot = p91, name = "Myeloid_Tumor_Classic_Mono_S100A8.pdf")
saveSCISSORS(plot = p92, name = "Myeloid_Tumor_Alternative_Macro_Spp1.pdf")
saveSCISSORS(plot = p93, name = "Myeloid_Tumor_DC_FCER1A.pdf")
saveSCISSORS(plot = p94, name = "Monocyte_Tumor_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p95, name = "DC_Tumor_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p96, name = "Monocyte_Tumor_Neutrophil_S100A8.pdf")
saveSCISSORS(plot = p97, name = "Monocyte_Tumor_Neutrophil_S100A9.pdf")
saveSCISSORS(plot = p98, name = "Monocyte_Tumor_Classic_Mono_CD14.pdf")
saveSCISSORS(plot = p99, name = "Monocyte_Tumor_Classic_Mono_FCGR3A.pdf")
saveSCISSORS(plot = p100, name = "DC_Tumor_cDC1_CLEC9A.pdf")
saveSCISSORS(plot = p101, name = "DC_Tumor_Langerhans_DC_CD1A.pdf")
saveSCISSORS(plot = p102, name = "DC_Tumor_Langerhans_DC_CD207.pdf")
saveSCISSORS(plot = p103, name = "DC_Tumor_Activated_DC_LAMP3.pdf")
saveSCISSORS(plot = p104, name = "DC_Tumor_cDC2_CLEC10A.pdf")
saveSCISSORS(plot = p105, name = "DC_Tumor_cDC2_KLF4.pdf")
saveSCISSORS(plot = p106, name = "DC_Tumor_cDC2_ZEB2.pdf")
saveSCISSORS(plot = p107, name = "Mono_Tumor_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p108, name = "DC_Tumor_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p109, name = "Mono_Norm_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p110, name = "DC_Norm_SCISSORS_FitSNE.pdf")
saveSCISSORS(plot = p111, name = "Mono_Norm_Neutrophil_S100A8.pdf")
saveSCISSORS(plot = p112, name = "Mono_Norm_Neutrophil_S100A9.pdf")
saveSCISSORS(plot = p113, name = "Mono_Norm_Classic_Monocyte_CD68.pdf")
saveSCISSORS(plot = p114, name = "Mono_Norm_Classic_Monocyte_CD14.pdf")
saveSCISSORS(plot = p115, name = "Mono_Norm_Classic_MOnocyte_SampleID.pdf")
saveSCISSORS(plot = p116, name = "Mono_Norm_CD8T_CD2.pdf")
saveSCISSORS(plot = p117, name = "Mono_Norm_CD8T_CD3D.pdf")
saveSCISSORS(plot = p118, name = "Mono_Norm_CD8T_CD8A.pdf")
saveSCISSORS(plot = p119, name = "Mono_Norm_CD8T_NKG7.pdf")
saveSCISSORS(plot = p120, name = "DC_Norm_cDC1_CLEC9A.pdf")
saveSCISSORS(plot = p121, name = "DC_Norm_cDC2_CLEC10A.pdf")
saveSCISSORS(plot = p122, name = "Mono_Norm_final_labels.pdf", pub.ready = TRUE, border = FALSE)
saveSCISSORS(plot = p123, name = "DC_Norm_final_labels.pdf", pub.ready = TRUE, border = FALSE)
```

And of course:

```{r}
sessionInfo()
```
