---
title: "Reclustering Cancer-Associated Fibroblasts Using SCISSORS"
subtitle: "Jack Leary"
author: 
  - "University of North Carolina at Chapel Hill - Lineberger Comprehensive Cancer Center"
  - "University of Florida - Department of Biostatistics"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: paper
    highlight: tango
    df_print: kable
    toc: true
    toc_float: true
    code_folding: show
    code_download: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE, 
                      warning = FALSE, 
                      results = "hold", 
                      fig.align = "center")
reticulate::use_virtualenv("~/Desktop/Python/science/venv/", required = TRUE)
```

# Introduction

In 2017, the Tuveson Lab at Cold Spring Harbor Cancer Center published a paper written by Elyada *et al* that detailed the discovery of cancer-associated fibroblasts (CAFs) in mice. The subtypes were then validated in human samples affected with PDAC in a subsequent paper released in 2019. Here we will use SCISSORS to identify the CAF subtypes within the larger stroma population, fine-grained immune cell types within broadly-defined immune clusters, and ductal & PDAC subtypes within the ductal group. You can install SCISSORS from [our GitHub repository](https://github.com/jr-leary7/SCISSORS).

# Libraries

## R

```{r, results='hide'}
library(VAM)           # single cell GSEA
library(dplyr)         # tidy data 
library(Seurat)        # single cell infrastructure
library(ggplot2)       # pretty plots
library(SingleR)       # cell type assignment
library(janitor)       # clean data
library(magrittr)      # pipes
library(decoderr)      # de novo deconvolution 
library(SCISSORS)      # our package
library(mixtools)      # Gaussian mixture model estimation
library(ggsignif)      # significance bars
library(patchwork)     # align plots
library(latex2exp)     # LaTeX
library(paletteer)     # color palettes
library(CONICSmat)     # CNV estimation
library(reticulate)    # Python interface
library(kableExtra)    # pretty tables
library(wesanderson)   # more color palettes
```

## Python

```{python}
import numpy as np
from openTSNE import TSNEEmbedding
from openTSNE import initialization
from openTSNE.affinity import Multiscale
from openTSNE.affinity import PerplexityBasedNN
```

# Data

First we load in the $\text{gene} \times \text{cell}$ counts matrix, then create a `Seurat` object to hold it in. Next, we add samplename, tissue type, and patient sex metadata taken from the publicly available dataset.

```{r}
raw_counts <- Read10X(data.dir = "~/Desktop/Data/Elyada Raw/All Human/")
pdac <- CreateSeuratObject(raw_counts, 
                           project = "Elyada", 
                           min.cells = 3, 
                           min.features = 500)
pdac@meta.data$sample <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "SRR9274536", 
                                   grepl("-2", rownames(pdac@meta.data)) ~ "SRR9274537", 
                                   grepl("-3", rownames(pdac@meta.data)) ~ "SRR9274538", 
                                   grepl("-4", rownames(pdac@meta.data)) ~ "SRR9274539",
                                   grepl("-5", rownames(pdac@meta.data)) ~ "SRR9274540", 
                                   grepl("-6", rownames(pdac@meta.data)) ~ "SRR9274541", 
                                   grepl("-7", rownames(pdac@meta.data)) ~ "SRR9274542", 
                                   grepl("-8", rownames(pdac@meta.data)) ~ "SRR9274543", 
                                   grepl("-9", rownames(pdac@meta.data)) ~ "SRR9274544")
pdac@meta.data$condition <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-2", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-3", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-4", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-5", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-6", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-7", rownames(pdac@meta.data)) ~ "PDAC", 
                                      grepl("-8", rownames(pdac@meta.data)) ~ "AdjNorm", 
                                      grepl("-9", rownames(pdac@meta.data)) ~ "PDAC")
pdac@meta.data$sex <- case_when(grepl("-1", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-2", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-3", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-4", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-5", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-6", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-7", rownames(pdac@meta.data)) ~ "female", 
                                grepl("-8", rownames(pdac@meta.data)) ~ "male", 
                                grepl("-9", rownames(pdac@meta.data)) ~ "female")
```

```{r, echo=FALSE}
rm(raw_counts)
```

Next, we read in a dataset of basal-like and classical PDAC marker genes that we'll use later on to perform enrichment analysis. 

```{r}
load("~/Desktop/Data/cmbSubtypes.RData")
pdac_50_genes <- subtypeGeneList[[4]]
basal_genes <- pdac_50_genes[pdac_50_genes$BasalLike, ]$geneSymbol
classical_genes <- pdac_50_genes[!pdac_50_genes$BasalLike, ]$geneSymbol
```

# Preprocessing

We run the typical single cell pre-processing steps on our cells - normalization, dimension reduction, and clustering. 

```{r, warning=FALSE}
pdac <- PrepareData(seurat.object = pdac, 
                    n.HVG = 4000, 
                    n.PC = 20, 
                    regress.mt = FALSE, 
                    regress.cc = FALSE, 
                    which.dim.reduc = "tsne", 
                    initial.resolution = 0.4, 
                    random.seed = 629)
```

Let's check out the t-SNE embedding. It looks decent, but could definitely be improved. Globally, the clusters are arranged in a blob - which isn't very informative - though the local structure seems to have been preserved fairly well. Visually, there are several clusters that contain subgroups - clusters 3, 4, 5, 6, & 9 look like good candidates for reclustering. 

```{r}
p0 <- DimPlot(pdac, reduction = "tsne", pt.size = 0.75) + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      labs(x = "t-SNE 1", y = "t-SNE 2") + 
      theme_yehlab() + 
      theme(legend.text = element_text(size = 10)) + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p0
```

We compute the distribution of the silhouette scores for each cluster. 

```{r}
sil_df <- ComputeSilhouetteScores(pdac, avg = FALSE)
p1 <- ggplot(sil_df, aes(x = Cluster, y = Score, fill = Cluster)) + 
      geom_violin(draw_quantiles = .5, color = "black", scale = "width") + 
      scale_fill_paletteer_d("ggthemes::Classic_20") + 
      labs(y = "Silhouette Score", x = "Louvain Clusters", fill = NULL) + 
      theme_minimal() + 
      theme(panel.grid = element_blank(), 
            panel.border = element_rect(fill = NA, size = 1), 
            legend.position = "none", 
            axis.title.x = element_text(size = 22), 
            axis.title.y = element_text(size = 22), 
            axis.text = element_text(size = 16), 
            axis.ticks = element_line(), 
            plot.subtitle = element_text(face = "italic", size = 10))
```

Looking at the silhouette score distribution for each cluster, we see that Cluster 10 seems to have the best fit, and Clusters 1, 2, & 9 seem to have pretty poor fits. 

```{r}
p1
```

Lastly, we'll identify marker genes for each of the clusters. 

```{r, fig.width=12, fig.height=5}
pdac_markers <- FindAllMarkers(pdac, 
                               logfc.threshold = 2, 
                               test.use = "wilcox", 
                               only.pos = TRUE, 
                               random.seed = 629, 
                               verbose = FALSE) %>% 
                filter(p_val_adj < .05) 
top5_pdac_markers <- pdac_markers %>% 
                     group_by(cluster) %>% 
                     arrange(desc(avg_log2FC)) %>% 
                     slice_head(n = 5)
p2 <- DotPlot(pdac, features = unique(top5_pdac_markers$gene), dot.scale = 15) + 
      scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
      labs(color = "Expression", size = "% Expressed", y = "Louvain Cluster") + 
      theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
            legend.position = "right", legend.justification = "center", 
            panel.border = element_rect(fill = NA, size = 1, color = "black"), 
            axis.line = element_blank(), 
            legend.title = element_text(size = 18), 
            axis.title.x = element_blank(), 
            axis.title.y = element_text(size = 20), 
            axis.text.y = element_text(size = 18)) + 
      guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
             size = guide_legend(title.position = "top", title.hjust = 0.5))
p2
```

## Optimize Dimension Reduction

I think the two-dimensional visualization of the cells could be improved. We'll try using UMAP and the Fast Fourier Transform-accelerated Fit-SNE (as implemented in the `openTSNE` library) to improve the embedding.

### UMAP

It seems like UMAP does a good job of clearly separating our clusters and preserving the global structure of the data. However, it's difficult to see local structure within some of the clusters due to their density.

```{r}
pdac <- RunUMAP(pdac, 
                reduction = "pca", 
                dims = 1:20, 
                umap.method = "uwot", 
                n.components = 2, 
                n.epochs = 750, 
                n.neighbors = 50, 
                metric = "cosine", 
                seed.use = 629, 
                verbose = FALSE)
p3 <- DimPlot(pdac, reduction = "umap", pt.size = 0.75) + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      labs(x = "UMAP 1", y = "UMAP 2") + 
      theme_yehlab() +
      theme(legend.text = element_text(size = 10)) + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p3
```

### Fit-SNE

For information on how to install the `openTSNE` implementation of Fit-SNE and how to run the algorithm, please visit [the excellent GitHub repository of Pavlin Policar](https://github.com/pavlin-policar/openTSNE).

First we'll run a simple, standard Fit-SNE embedding. It's necessary to make the PCA embeddings accessible by Python.

```{r}
pc_df <- Embeddings(pdac, reduction = "pca")
```

```{python}
# import data
pc_df = np.array(r.pc_df)
# run Fit-SNE
affin = PerplexityBasedNN(pc_df, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(pc_df, random_state=629)
tsne1 = TSNEEmbedding(init, affin, negative_gradient_method='fft')
embed1 = tsne1.optimize(n_iter=350, exaggeration=12, momentum=0.6) 
embed2 = embed1.optimize(n_iter=750, momentum=0.8)
```

The embedding looks good, so we'll use it going forwards.

```{r}
embed <- as.matrix(py$embed2)
rownames(embed) <- colnames(pdac)
pdac@reductions$fitsne <- CreateDimReducObject(embeddings = embed, 
                                               key = "FitSNE_", 
                                               assay = "SCT", 
                                               global = TRUE)
p4 <- DimPlot(pdac, reduction = "fitsne", pt.size = 0.75) + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme_yehlab() +
      theme(legend.text = element_text(size = 10)) + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p4
```

Due to the way `Seurat` accesses cell embeddings, we'll need to replace our original t-SNE dimension reduction in our `Seurat` object with the new Fit-SNE version. We'll keep the original Barnes-Hut t-SNE embedding under a separate name.

```{r}
pdac@reductions$bh_tsne <- pdac@reductions$tsne
pdac@reductions$tsne <- pdac@reductions$fitsne
```

```{r, echo=FALSE, results='hide'}
rm(embed, pc_df)
gc(verbose = FALSE)
```

# SingleR Cell Type Identification

## Single Cell RNA-seq Reference Data

Here we use the `SingleR` package to identify broad cell types. The reference dataset we load is an `sctransform`-normalized version of the raw counts available in `scRNAseq::BaronPancreasData()`, which consists of normal pancreas cells that were sequenced and annotated by the researchers.

```{r}
sc_ref <- readRDS("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/Seurat/single_cell_ref_normalized.Rds")
sc_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                    ref = sc_ref, 
                    labels = sc_ref$label, 
                    clusters = pdac$seurat_clusters, 
                    de.method = "wilcox")
pdac[["SingleR.labels.sc"]] <- sc_preds$labels[match(pdac[[]][["seurat_clusters"]], rownames(sc_preds))]
pdac$SingleR.labels.sc <- case_when(pdac$SingleR.labels.sc == "acinar" ~ "Acinar", 
                                    pdac$SingleR.labels.sc == "activated_stellate" ~ "Activated Stellate",
                                    pdac$SingleR.labels.sc == "ductal" ~ "Ductal", 
                                    pdac$SingleR.labels.sc == "macrophage" ~ "Macrophage",
                                    pdac$SingleR.labels.sc == "t_cell" ~ "T")
```

We can see that there's a large immune population, as well as smaller ductal, fibroblast (denoted activated stellate in the reference dataset), and acinar groups. These broad cell types line up with what we expected to see given the authors' original cell cluster annotations. 

```{r}
p5 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.sc", pt.size = 0.75) + 
      scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 4)))
p5
```

## Bulk Tissue RNA-seq Reference Data

This dataset is composed of labeled & log-normalized bulk RNA-seq samples from the Human Primary Cell Atlas.

```{r}
bulk_ref <- celldex::HumanPrimaryCellAtlasData()
bulk_preds <- SingleR(test = data.frame(pdac@assays$SCT@data), 
                      ref = bulk_ref, 
                      labels = bulk_ref$label.main, 
                      clusters = pdac$seurat_clusters, 
                      de.method = "wilcox")
pdac[["SingleR.labels.bulk"]] <- bulk_preds$labels[match(pdac[[]][["seurat_clusters"]], rownames(bulk_preds))]
pdac$SingleR.labels.bulk <- case_when(pdac$SingleR.labels.bulk == "B_cell" ~ "B", 
                                      pdac$SingleR.labels.bulk == "Epithelial_cells" ~ "Epithelial", 
                                      pdac$SingleR.labels.bulk == "B_cell-" ~ "B", 
                                      pdac$SingleR.labels.bulk == "T_cells" ~ "T",
                                      TRUE ~ pdac$SingleR.labels.bulk)
```

The bulk reference gives us somewhat more granular labels for the immune cells, and confirms the identities of the ductal / epithelial and stroma clusters. 

```{r}
p6 <- DimPlot(pdac, reduction = "tsne", group.by = "SingleR.labels.bulk", pt.size = 0.75) + 
      scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 2, override.aes = list(size = 4)))
p6
```

```{r, echo=FALSE, results='hide'}
rm(sc_preds, bulk_preds, bulk_ref, sc_ref)
gc(verbose = FALSE)
```

One shouldn't use `SingleR` as the final authority for cell types, but we were able to confirm the identities of the ductal and fibroblast clusters, which is important for this dataset as the cells we are most interested in are the cancer-associated fibroblasts (CAFs). 

# CONICSmat CNV Estimation

Next we'll attempt to identify malignant cells using single-cell copy number variation estimation as implemented in the `CONCISmat` package. Details of the GMM methodology used can be found at [the Diaz Lab's GitHub repository](https://github.com/diazlab/CONICS). Note: this step is memory-intensive because 1) it requires the sparse counts matrix to be cast to a dense matrix and 2) a lot of Gaussian mixture models get estimated. If your machine doesn't have a lot of RAM it might be best to skip this and manually annotate the malignant cells instead through the usage of canonical markers or the `VAM` single-cell GSEA methodology. 

```{r, results="hide"}
chrom_regions <- read.table("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/chrom_arm_positions.txt", 
                            sep = "\t", 
                            row.names = 1, 
                            header = TRUE)
gene_pos <- getGenePositions(rownames(pdac))
cpm <- t(t(as.matrix(pdac@assays$SCT@counts)) / colSums(as.matrix(pdac@assays$SCT@counts))) * 10^5
cpm <- log2(cpm + 1)
norm_factor <- calcNormFactors(cpm)
cnv_est <- plotAll(mat = cpm, 
                   normFactor = norm_factor, 
                   regions = chrom_regions, 
                   gene_pos = gene_pos, 
                   fname = "Elyada")
```

```{r, echo=FALSE, results='hide'}
gc(verbose = FALSE)
```

## Visualizing CNVs

After estimating CNVs, we cluster the cells into $k = 3$ clusters, with the hope of finding one large cluster of normal cells and two smaller clusters composed of CAFs and PDAC cells. 

```{r}
bic_table <- read.table("./Elyada_BIC_LR.txt", 
                        sep = "\t", 
                        row.names = 1, 
                        header = TRUE, 
                        check.names = FALSE)
cand_regions <- rownames(bic_table[bic_table$`BIC difference` > 1000 & bic_table$`LRT adj. p-val` < .01, ])
pdf("~/Desktop/R/SCISSORS/vignettes/figures_supp/Elyada/CONICSmat_Heatmap.pdf", width = 12, height = 6)
hist1 <- plotHistogram(cnv_est[, cand_regions], 
                       cpm, 
                       clusters = 3, 
                       zscoreThreshold = 3, 
                       celltypes = pdac$SingleR.labels.bulk, 
                       patients = pdac$sample)
dev.off()
```


```{r}
normal <- which(hist1 != 2)
malignant <- which(hist1 == 2)
pdac@meta.data$malig <- ifelse(rownames(pdac@meta.data) %in% names(normal), "Normal", "Malignant")
cbind(hist1, rownames(pdac@meta.data)) %>% 
  as.data.frame() %>% 
  filter(V2 %in% secretory_cells) %>% 
  mutate(malig = ifelse(V2 %in% names(normal), "Normal", "Malignant")) %>% 
  #group_by(hist1) %>% 
  summarise(MeanMalig = mean(case_when(malig == "Malignant" ~ 1, TRUE ~ 0)))

cbind(hist1, rownames(pdac@meta.data)) %>% 
  as.data.frame() %>% 
  group_by(hist1) %>% 
  summarise(MeanSec = mean(case_when(V2 %in% secretory_cells ~ 1, TRUE ~ 0)), 
            MeanSec2 = mean(secretory_cells %in% V2), 
            #MeanMalig = mean(case_when(V2 %in% names(malignant) ~ 1, TRUE ~ 0)), 
            N = n())
```

We add the normal vs. malignant cell labels in to our `Seurat` object's metadata, then visualize the results. As expected, the malignant cells are located in the clusters identified by `SingleR` as ductal cells and fibroblasts. This indicates that `CONICSmat` did a solid job of estimating the CNVs - no easy feat with sparse, noisy single-cell data. 

```{r}
normal <- which(hist1 == 1 | hist1 == 2)
malignant <- which(hist1 != 1 & hist1 != 2)
pdac@meta.data$malig <- ifelse(rownames(pdac@meta.data) %in% names(normal), "Normal", "Malignant")
p7 <- DimPlot(pdac, reduction = "tsne", group.by = "malig", pt.size = 0.75) + 
      scale_color_manual(values = wes_palette("Zissou1", n = 5)[c(5, 2)]) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p7
p4
```

When looking at the `SingleR` labels from the bulk reference, we see that the Epithelial (Ductal) and Mesenchymal Stem Cell (Fibroblast) clusters have the highest proportions of malignant cells, which is exactly what we expected to see. 

```{r}
pdac@meta.data %>% 
  group_by(SingleR.labels.bulk) %>% 
  summarise(MeanMalig = mean(case_when(malig == "Malignant" ~ 1, TRUE ~ 0))) %>% 
  mutate(MeanMalig = formattable::percent(MeanMalig)) %>% 
  kbl(booktabs = TRUE, col.names = c("Bulk Reference SingleR Label", "% Malignant Cells")) %>% 
  kable_minimal(full_width = FALSE)

pdac@meta.data %>% 
  filter(rownames(.) %in% secretory_cells) %>% 
  summarise(MeanMalig = mean(case_when(malig == "Malignant" ~ 1, TRUE ~ 0)))
```

```{r, echo=FALSE, results='hide'}
rm(cpm, bic_table, chrom_regions, cnv_est, gene_pos, cand_regions, hist1, normal, norm_factor, malignant)
gc(verbose = FALSE)
```

# DECODER

Next, we use [Dr. Xianlu Peng's DECODER](https://github.com/laurapeng/decoderr) in order to deconvolve the dataset and assign weights to each cell using non-negative matrix factorization. We calculate basal & classical PDAC, normal & activated stroma, immune, and endocrine & exocrine pancreas compartment weights.

```{r}
sample_wts_unscaled <- Decon_single_sample(refSet = "TCGA_RNAseq_PAAD", 
                                           data = pdac@assays$SCT@data, 
                                           geneIDType = "geneSymbol")
sample_wts <- Norm_PDAC_weights(sample_wts_unscaled)
pdac <- AddMetaData(pdac, sample_wts$Immune, "immune")
pdac <- AddMetaData(pdac, sample_wts$bcRatio, "bc_ratio")
pdac <- AddMetaData(pdac, sample_wts$Exocrine, "exocrine")
pdac <- AddMetaData(pdac, sample_wts$Endocrine, "endocrine")
pdac <- AddMetaData(pdac, sample_wts$BasalTumor, "basal-like")
pdac <- AddMetaData(pdac, sample_wts$ClassicalTumor, "classical")
pdac <- AddMetaData(pdac, sample_wts$NormalStroma, "norm_stroma")
pdac <- AddMetaData(pdac, sample_wts$ActivatedStroma, "act_stroma")
```

## Visualizing DECODER Weights

### Basal PDAC

The basal weights are highest in a subcluster of the ductal group identified through `SingleR`. This is interesting as the authors did not find evidence of the basal subtype in their paper. 

```{r}
p8 <- FeaturePlot(pdac, reduction = "tsne", features = "basal", pt.size = 0.75) + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p8
```

### Classical PDAC

The classical weights are highest in another subcluster of the ductal cluster, and cells with high classical weights do not collocate with those that have high basal weights. The putative classical and basal PDAC cells also align closely with the cells identified through `CONCISmat` as malignant. 

```{r}
p9 <- FeaturePlot(pdac, reduction = "tsne", features = "classical", pt.size = 0.75) + 
      scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
      labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
      theme(plot.title = element_blank()) + 
      theme_yehlab() + 
      NoLegend()
p9
```

### Exocrine Pancreas

The cluster identified through `SingleR` as acinar cells is the only cluster with high exocrine pancreas weights. 

```{r}
p10 <- FeaturePlot(pdac, reduction = "tsne", features = "exocrine", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p10
```

### Endocrine Pancreas

No cells have high endocrine pancreas weights. 

```{r}
p11 <- FeaturePlot(pdac, reduction = "tsne", features = "endocrine", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p11
```

### Immune

Once again, we confirm the largeness of the immune cell population in this dataset. 

```{r}
p12 <- FeaturePlot(pdac, reduction = "tsne", features = "immune", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p12
```

### Normal Stroma

Cells with high normal stroma stroma weights are located in the cluster identified by `SingleR` as being stromal cells. 

```{r}
p13 <- FeaturePlot(pdac, reduction = "tsne", features = "norm_stroma", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p13
```

### Activated Stroma

Cells with high activated stroma weights are also located in the fibroblast cluster, and do not intersect with the cells that have high normal stroma scores. This indicates that `SCISSORS` will most likely perform well on the fibroblast cluster and be able to quickly tease out the cell subtypes.

```{r}
p14 <- FeaturePlot(pdac, reduction = "tsne", features = "act_stroma", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme(plot.title = element_blank()) + 
       theme_yehlab() + 
       NoLegend()
p14
```

# SCISSORS

Now that we have rough labels from `SingleR`, CNVs from `CONICSmat`, and compartment weights from `DECODER`, we should have more than enough cell-level metadata to look for and annotate cell subtypes using `SCISSORS`. 

## Fibroblasts

The fibroblast marker genes provided by Elyada *et al* match the `SingleR` results defining cluster 6 as containing fibroblasts. 

```{r}
p15 <- FeaturePlot(pdac, reduction = "tsne", features = "COL1A1", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL1A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p16 <- FeaturePlot(pdac, reduction = "tsne", features = "COL3A1", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "COL3A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p17 <- FeaturePlot(pdac, reduction = "tsne", features = "LUM", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "LUM") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p18 <- FeaturePlot(pdac, reduction = "tsne", features = "DCN", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "DCN") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p15 | p16) / (p17 | p18)
```

Here's the cells we'll be reclustering. 

```{r}
fibro_cells <- rownames(pdac@meta.data[pdac@meta.data$seurat_clusters == 6, ])
p19 <- DimPlot(pdac, reduction = "tsne", cells.highlight = fibro_cells, cols.highlight = "navy", pt.size = 0.75) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend()
p19 / p4
```

### Reclustering

Here we use `ReclusterCells()` to identify subclusters within cluster 6. We find five distinct subclusters.

```{r, warning=FALSE}
fibro <- ReclusterCells(pdac,
                        which.clust = 6, 
                        n.HVG = 4000, 
                        n.PC = 20, 
                        resolution.vals = c(.03, .05, .1), 
                        k.vals = c(20, 30, 40), 
                        redo.embedding = TRUE)
fibro_pc <- Embeddings(fibro, "pca")
```

We'll again run Fit-SNE on the reclustered cells, for consistencies sake. 

```{python}
# import data
fibro_pc = r.fibro_pc
# run Fit-SNE
affin_fibro = PerplexityBasedNN(fibro_pc, perplexity=40, metric='cosine', random_state=629)
init = initialization.pca(fibro_pc, random_state=629)
tsne_f = TSNEEmbedding(init, affin_fibro, negative_gradient_method='fft')
embed_f1 = tsne_f.optimize(n_iter=250, exaggeration=5, momentum=0.4) 
embed_f2 = embed_f1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

Pulling the results back into R and visualizing them shows clear separation between the subclusters. There's some noise in subcluster 0, but other than that the reembedding looks solid. 

```{r}
embed_fibro <- as.matrix(py$embed_f2)
rownames(embed_fibro) <- colnames(fibro)
fibro@reductions$bh_tsne <- fibro@reductions$tsne
fibro@reductions$tsne<- CreateDimReducObject(embeddings = embed_fibro, 
                                             key = "FitSNE_", 
                                             assay = "SCT", 
                                             global = TRUE)
p20 <- DimPlot(fibro, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p20
```

### Cell Type Identification

First we identify the endothelial and perivascular cells using PLVAP and RGS5 expression. 

```{r}
p21 <- FeaturePlot(fibro, reduction = "tsne", features = "PLVAP", pt.size = 1.5) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "PLVAP") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p22 <- FeaturePlot(fibro, reduction = "tsne", features = "RGS5", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "RGS5") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p21 | p22) / p20
```


Next, we use the `VAM` method of single cell gene set enrichment analysis to determine which clusters are enriched for the iCAF and myCAF marker genes, as well as the general pan-CAF marker set. We use the marker genes identified by the authors. 

```{r}
icaf_genes <- c("IL6", "PDGFRA", "CXCL12", "CFD", "LMNA", "AGTR1", "HAS1", "CXCL1", "CXCL2", "CCL2", "IL8")
mycaf_genes <- c("ACTA2", "TAGLN", "MMP11", "MYL9", "HOPX", "POSTN", "TPM1", "TPM2")
pan_caf_genes <- c("COL1A1", "FAP", "PDPN", "DCN", "VIM")
pan_caf_genes_mouse <- c("COL1A1", "COL1A2", "PDPN", "DCN")
gene_sets <- list(icaf_genes, mycaf_genes, pan_caf_genes, pan_caf_genes_mouse)
names(gene_sets) <- c("iCAF", "myCAF", "Pan-CAF", "Pan-CAF_mouse")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
```

#### iCAFs & myCAFs

We can easily define cluster 0 as the myCAF population, and cluster 2 as the slightly smaller iCAF population. Cluster 4 shows no enrichment whatsoever for either the iCAF or myCAF gene sets. 

```{r}
p23 <- FeaturePlot(fibro, reduction = "tsne", features = "iCAF", pt.size = 1.5) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "iCAF") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p24 <- FeaturePlot(fibro, reduction = "tsne", features = "myCAF", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "myCAF") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p23 | p24) / p20
```

#### apCAFs

So we have a small cluster of 23 cells that does not appear to express any of the fibroblast, CAF, perivascular, or endothelial markers. After performing differential expression analysis, we find that the top 3 markers for cluster 4 are CLU, CD74, and CRYAB. Interestingly, CLU and CD74 were found to be differentially expressed in the apCAF population discovered in the KPC mouse models of CAFs in Elyada *et al*.

```{r}
DefaultAssay(fibro) <- "SCT"
fibro_markers <- FindAllMarkers(fibro, 
                                assay = "SCT", 
                                logfc.threshold = 1.5, 
                                test.use = "wilcox", 
                                only.pos = TRUE, 
                                random.seed = 629, 
                                verbose = FALSE)
fibro_markers %>% 
  filter(cluster == 4) %>% 
  arrange(desc(avg_log2FC)) %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2) %>% 
  slice_head(n = 5) %>% 
  kbl(booktabs = TRUE, digits = 4, row.names = FALSE) %>% 
  kable_minimal("hover", full_width = FALSE)
```

We re-run GSEA, again using the `VAM` package and the differentially expressed genes for the apCAF population as defined in Elyada *et al* (with the mouse gene names converted to HGNC symbols). We can see that the apCAF pathway is strongly enriched in cluster 4 as compared to the other CAF clusters.

```{r}
apcaf_genes <- c("HLA-DQB1", "CD74", "SAA3P", "SLPI")
gene_sets <- list(icaf_genes, mycaf_genes, apcaf_genes, pan_caf_genes)
names(gene_sets) <- c("iCAF", "myCAF", "apCAF", "Pan-CAF")
for (i in seq(gene_sets)) {
  gene_sets[[i]] <- gene_sets[[i]][gene_sets[[i]] %in% rownames(fibro)]
}
fibro <- vamForSeurat(fibro, 
                      gene.set.collection = gene_sets, 
                      gamma = TRUE)
DefaultAssay(fibro) <- "VAMcdf"
p25 <- FeaturePlot(fibro, reduction = "tsne", features = "apCAF", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(title = "apCAF") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p25 / p20
```

### Visualization

Finally, we add cell labels to our identified clusters.

```{r}
fibro$label <- case_when(fibro$seurat_clusters == 0 ~ "myCAF", 
                         fibro$seurat_clusters == 1 ~ "Perivascular", 
                         fibro$seurat_clusters == 2 ~ "iCAF", 
                         fibro$seurat_clusters == 3 ~ "Endothelial", 
                         fibro$seurat_clusters == 4 ~ "apCAF")
Idents(fibro) <- "label"
p26 <- DimPlot(fibro, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p26
```

Here's the marker genes for each cluster. 

```{r, fig.width=8, fig.height=4}
DefaultAssay(fibro) <- "SCT"
fibro_markers2 <- FindAllMarkers(fibro, 
                                 logfc.threshold = 2, 
                                 test.use = "wilcox", 
                                 only.pos = TRUE, 
                                 random.seed = 629, 
                                 verbose = FALSE) %>% 
                  filter(p_val_adj < .05) %>% 
                  mutate(source = "Stroma", 
                         log2fc_cutoff = 2)
top5_fibro_markers <- fibro_markers2 %>% 
                      group_by(cluster) %>% 
                      arrange(desc(avg_log2FC)) %>% 
                      slice_head(n = 5)
p27 <- DotPlot(fibro, features = top5_fibro_markers$gene, dot.scale = 15) + 
       scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
       labs(color = "Expression", size = "% Expressed") + 
       theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
             legend.position = "right", 
             legend.justification = "center", 
             panel.border = element_rect(fill = NA, size = 1, color = "black"), 
             axis.line = element_blank(), 
             legend.title = element_text(size = 18), 
             axis.title.x = element_blank(), 
             axis.title.y = element_blank(), 
             axis.text.y = element_text(size = 18)) + 
       guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
              size = guide_legend(title.position = "top", title.hjust = 0.5))
p27
```

We'll also compute CAF-only marker genes & visualize them. 

```{r, fig.width=6, fig.height=3}
caf <- subset(fibro, subset = label %in% c("iCAF", "myCAF", "apCAF"))
caf_markers <- FindAllMarkers(caf, 
                              logfc.threshold = .5, 
                              test.use = "wilcox", 
                              only.pos = TRUE, 
                              random.seed = 629, 
                              verbose = FALSE) %>% 
               filter(p_val_adj < .05) %>% 
               mutate(source = "CAF", 
                      log2fc_cutoff = .5)
top5_caf_markers <- caf_markers %>% 
                    group_by(cluster) %>% 
                    arrange(desc(avg_log2FC)) %>% 
                    slice_head(n = 5)
p28a <- DotPlot(caf, features = top5_caf_markers$gene, dot.scale = 15) + 
        scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
        labs(color = "Expression", size = "% Expressed") + 
        theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
              legend.position = "right", 
              legend.justification = "center", 
              panel.border = element_rect(fill = NA, size = 1, color = "black"), 
              axis.line = element_blank(), 
              legend.title = element_text(size = 18), 
              axis.title.x = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_text(size = 18)) + 
        guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
               size = guide_legend(title.position = "top", title.hjust = 0.5))
p28a
```

Lastly, here's a violin plot of the CAF gene set `VAM` scores for each of the three CAF clusters. 

```{r}
p28b <- data.frame(t(caf@assays$VAMcdf@data)) %>% 
        bind_cols((caf@meta.data %>% dplyr::select(label))) %>% 
        ggplot(aes(x = label, y = apCAF, fill = label)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1.2) + 
        geom_signif(comparisons = list(c("apCAF", "iCAF"), c("apCAF", "myCAF")), size = 1.2, textsize = 8,
                    margin_top = .075, 
                    test = wilcox.test, color = "black", map_signif_level = TRUE, step_increase = .15) +
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(5, 4, 2)]) +
        scale_y_continuous(breaks = seq(0, 1, 0.25), limits = c(0, 1.29)) + 
        labs(y = "Elyada apCAF", x = NULL, fill = NULL) + 
        theme_minimal() + 
        theme(panel.grid = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              legend.position = "none", 
              axis.title.y = element_text(size = 24), 
              axis.text = element_text(size = 22), 
              axis.text.x = element_text(size = 22, angle = 45, vjust = 1, hjust = 1), 
              axis.ticks = element_line(), 
              plot.subtitle = element_text(face = "italic", size = 10))
p28c <- data.frame(t(caf@assays$VAMcdf@data)) %>% 
        bind_cols((caf@meta.data %>% dplyr::select(label))) %>% 
        ggplot(aes(x = label, y = iCAF, fill = label)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1.2) + 
        geom_signif(comparisons = list(c("iCAF", "apCAF"), c("iCAF", "myCAF")), size = 1.2, textsize = 8,
                    margin_top = .075, 
                    test = wilcox.test, color = "black", map_signif_level = TRUE, step_increase = .15) +
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(5, 4, 2)]) +
        scale_y_continuous(breaks = seq(0, 1, 0.25), limits = c(0, 1.29)) + 
        labs(y = "Elyada iCAF", x = NULL, fill = NULL) + 
        theme_minimal() + 
        theme(panel.grid = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              legend.position = "none", 
              axis.title.y = element_text(size = 24), 
              axis.text = element_text(size = 22), 
              axis.text.x = element_text(size = 22, angle = 45, vjust = 1, hjust = 1), 
              axis.ticks = element_line(), 
              plot.subtitle = element_text(face = "italic", size = 10))
p28d <- data.frame(t(caf@assays$VAMcdf@data)) %>% 
        bind_cols((caf@meta.data %>% dplyr::select(label))) %>% 
        ggplot(aes(x = label, y = myCAF, fill = label)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1.2) + 
        geom_signif(comparisons = list(c("myCAF", "iCAF"), c("myCAF", "apCAF")), size = 1.2, textsize = 8,
                    margin_top = .075, 
                    test = wilcox.test, color = "black", map_signif_level = TRUE, step_increase = .15) +
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(5, 4, 2)]) +
        scale_y_continuous(breaks = seq(0, 1, 0.25), limits = c(0, 1.29)) + 
        labs(y = "Elyada myCAF", x = NULL, fill = NULL) + 
        theme_minimal() + 
        theme(panel.grid = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              legend.position = "none", 
              axis.title.y = element_text(size = 24), 
              axis.text = element_text(size = 22), 
              axis.text.x = element_text(size = 22, angle = 45, vjust = 1, hjust = 1), 
              axis.ticks = element_line(), 
              plot.subtitle = element_text(face = "italic", size = 10))
p28e <- data.frame(t(caf@assays$VAMcdf@data)) %>% 
        bind_cols((caf@meta.data %>% dplyr::select(label))) %>% 
        ggplot(aes(x = label, y = `Pan.CAF`, fill = label)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1.2) +
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(5, 4, 2)]) +
        labs(y = "Elyada Pan-CAF", x = NULL, fill = NULL) + 
        theme_minimal() + 
        theme(panel.grid = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              legend.position = "none", 
              axis.title.y = element_text(size = 24), 
              axis.text = element_text(size = 22), 
              axis.text.x = element_text(size = 22, angle = 45, vjust = 1, hjust = 1), 
              axis.ticks = element_line(), 
              plot.subtitle = element_text(face = "italic", size = 10))
```

Here's the CAF gene set scores as generated by `VAM`.

```{r, fig.width=3, fig.height=4}
p28f <- (p28e | p28c)/ (p28d | p28b)
p28f
```

## T Cells

Going back to the main `Seurat` object, we should have a large population of T and NK cells that warrants further investigation. Using CD3D expression we can easily identify clusters 0, 3, and 7 as the mixed T & NK cells. We already see some good separation, so reclustering the cells should have good results. 

```{r}
p29 <- FeaturePlot(pdac, reduction = "tsne", features = "CD3D", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p29 / p4
```

Here's the cells we'll be reclustering. 

```{r}
nkt_cells <- rownames(pdac@meta.data[pdac@meta.data$seurat_clusters %in% c(0, 3, 8), ])
p30 <- DimPlot(pdac, reduction = "tsne", cells.highlight = nkt_cells, cols.highlight = "navy", pt.size = 0.75) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend()
p30 / p4
```

### Tumor

```{r}
nkt_tumor <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 8) & condition == "PDAC")
```

#### Reclustering

Here we run `ReclusterCells()`, while treating the NK & T cell clusters as one large group. This will hopefully allow use to elucidate T cell subtypes. We use fewer PCs for these cells since the differences between immune cells are subtle, and adding more PCs will most likely only contribute noise to our analyses. 

```{r}
nkt_tumor <- ReclusterCells(nkt_tumor, 
                            which.clust = list(0, 3, 8), 
                            merge.clusters = TRUE, 
                            n.HVG = 4000, 
                            n.PC = 15, 
                            k.vals = c(30, 40, 50, 60), 
                            resolution.vals = c(.1, .2, .3), 
                            redo.embedding = TRUE, 
                            random.seed = 629)
nkt_tumor_pc <- Embeddings(nkt_tumor, "pca")
```

Once again we'll run Fit-SNE on the reclustered cells. 

```{python}
# import data
nkt_tumor_pc = r.nkt_tumor_pc
# run Fit-SNE
affin_nkt_tumor = PerplexityBasedNN(nkt_tumor_pc, perplexity=100, metric='cosine', random_state=629)
init = initialization.pca(nkt_tumor_pc, random_state=629)
tsne_nkt_tumor = TSNEEmbedding(init, affin_nkt_tumor, negative_gradient_method='fft')
embed_nkt_tumor1 = tsne_nkt_tumor.optimize(n_iter=250, exaggeration=4, momentum=0.6) 
embed_nkt_tumor2 = embed_nkt_tumor1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
affin_nkt_tumor.set_perplexity(50)
embed_nkt_tumor3 = embed_nkt_tumor2.optimize(n_iter=500, exaggeration=1, momentum=0.6)
```

The reembedding isn't perfect, which we somewhat expected as immune cells are difficult to tell apart based on the transcriptome alone. 

```{r}
embed_nkt_tumor <- as.matrix(py$embed_nkt_tumor3)
rownames(embed_nkt_tumor) <- colnames(nkt_tumor)
nkt_tumor@reductions$bh_tsne <- nkt_tumor@reductions$tsne
nkt_tumor@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_tumor, 
                                                 key = "FitSNE_", 
                                                 assay = "SCT", 
                                                 global = TRUE)
p31 <- DimPlot(nkt_tumor, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::category20_d3")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p31
```

#### Cell Type Identification

##### CD4+ T

Clusters 0 and 1 contain our CD4+ T cells, which we characterize using IL7R as we did in the PBMC3k vignette. It's somewhat outside of our scope here to determine which subtype each cluster belongs to, so we'll simply assign both clusters the same label and move on. 

```{r}
p32 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL7R", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p33 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD69", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p32 | p33) / p31
```

##### T-regs

Cluster 3 contains the regulatory T cells. 

```{r}
p34 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "IL2RA", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p35 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "FOXP3", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p34 | p35) / p31
```

##### Proliferating T-regs

We can find the proliferating T-regs in cluster 7. 

```{r}
p36 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TOP2A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p36 / p31
```

##### Mast

Mast cells can be identified using TPSAB1 expression in cluster 6. 

```{r}
p37 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "TPSAB1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p37 / p31
```

##### NK

NKG7 and PRF1 show us the NK cells in cluster 5. 

```{r}
p38 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "NKG7", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p39 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "PRF1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p38 | p39) / p31
```

##### CD8+ T

We use CD8A and CD2 to reveal the CD8+ T cells within clusters 2 and 4. 

```{r}
p40 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD8A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p41 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD2", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p40 | p41) / p31
```

#### Intermediate Monocyte

Lastly, we have cluster 8, which doesn't highly express our pan-T/NK cell markers CD3D and CD2. It could be a myeloid cell cluster that was mistakenly grouped with the T/NK cells. We'll start with a Wilcoxon test to determine its differentially expressed genes. Interestingly, several [intermediate monocyte markers](https://www.frontiersin.org/articles/10.3389/fimmu.2019.02035/full#B3) - LYZ, HLA-DRA, CD74, and HLA-DPB1 - are differentially expressed in cluster 8. 

```{r}
clust8_markers <- FindAllMarkers(nkt_tumor, 
                                 test.use = "wilcox",
                                 min.diff.pct = .2,
                                 logfc.threshold = .5, 
                                 verbose = FALSE, 
                                 random.seed = 629) %>% 
                  filter(cluster == 8, p_val_adj < .05) %>% 
                  arrange(desc(1 - p_val_adj))
clust8_markers %>% 
  filter(gene %in% c("LYZ", "HLA-DRA", "CD74", "HLA-DPB1")) %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2) %>% 
  kbl(booktabs = TRUE, digits = 4, row.names = FALSE) %>% 
  kable_minimal("hover", full_width = FALSE)
```

We'll plot some of those markers below. 

```{r}
p42 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "LYZ", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p43 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "HLA-DRA", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p44 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "CD74", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p45 <- FeaturePlot(nkt_tumor, reduction = "tsne", features = "HLA-DPB1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
((p42 | p43) / (p44 | p45)) / p31
```

#### Visualization

We add labels to our T cell `Seurat` object and visualize the results. 

```{r}
nkt_tumor$label <- case_when(nkt_tumor$seurat_clusters == 0 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 1 ~ "CD4+ T", 
                             nkt_tumor$seurat_clusters == 2 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 3 ~ "T-reg", 
                             nkt_tumor$seurat_clusters == 4 ~ "CD8+ T", 
                             nkt_tumor$seurat_clusters == 5 ~ "NK", 
                             nkt_tumor$seurat_clusters == 6 ~ "Mast", 
                             nkt_tumor$seurat_clusters == 7 ~ "Proliferating T-reg", 
                             nkt_tumor$seurat_clusters == 8 ~ "Intermediate Monocyte")
Idents(nkt_tumor) <- "label"
p46 <- DimPlot(nkt_tumor, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 4)))
p46
```

Here's the marker genes. 

```{r, fig.width=8, fig.height=4}
nkt_tumor_markers2 <- FindAllMarkers(nkt_tumor, 
                                     logfc.threshold = .75, 
                                     test.use = "wilcox", 
                                     only.pos = TRUE, 
                                     random.seed = 629, 
                                     verbose = FALSE) %>% 
                      filter(p_val_adj < .05) %>% 
                      mutate(source = "NK/T Tumor", 
                             log2fc_cutoff = .75)
top5_nkt_tumor_markers <- nkt_tumor_markers2 %>% 
                          group_by(cluster) %>% 
                          arrange(desc(avg_log2FC)) %>% 
                          slice_head(n = 5)
p47 <- DotPlot(nkt_tumor, features = unique(top5_nkt_tumor_markers$gene), dot.scale = 15) + 
       scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
       labs(color = "Expression", size = "% Expressed") + 
       theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
             legend.position = "right", 
             legend.justification = "center", 
             panel.border = element_rect(fill = NA, size = 1, color = "black"), 
             axis.line = element_blank(), 
             legend.title = element_text(size = 18), 
             axis.title.x = element_blank(), 
             axis.title.y = element_blank(), 
             axis.text.y = element_text(size = 18)) + 
       guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
              size = guide_legend(title.position = "top", title.hjust = 0.5))
p47
```

### Adjacent Normal

```{r}
nkt_norm <- subset(pdac, subset = seurat_clusters %in% c(0, 3, 8) & condition == "AdjNorm")
```

#### Reclustering

```{r}
nkt_norm <- ReclusterCells(nkt_norm, 
                           which.clust = list(0, 3, 8), 
                           merge.clusters = TRUE, 
                           n.HVG = 4000, 
                           n.PC = 10, 
                           k.vals = c(15, 20, 25), 
                           resolution.vals = c(.2, .3, .4), 
                           nn.metric = "euclidean", 
                           redo.embedding = TRUE, 
                           random.seed = 629)
nkt_norm_pc <- Embeddings(nkt_norm, "pca")
```

As with the other reclusterings, we'll run Fit-SNE in order to (hopefully) obtain a better low-dimensional embedding of our cells.

```{python}
# import data
nkt_norm_pc = r.nkt_norm_pc
# run Fit-SNE
affin_nkt_norm = PerplexityBasedNN(nkt_norm_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(nkt_norm_pc, random_state=629)
tsne_nkt_norm = TSNEEmbedding(init, affin_nkt_norm, negative_gradient_method='fft')
embed_nkt_norm1 = tsne_nkt_norm.optimize(n_iter=250, exaggeration=12, momentum=0.6) 
embed_nkt_norm2 = embed_nkt_norm1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

```{r}
embed_nkt_norm <- as.matrix(py$embed_nkt_norm2)
rownames(embed_nkt_norm) <- colnames(nkt_norm)
nkt_norm@reductions$bh_tsne <- nkt_norm@reductions$tsne
nkt_norm@reductions$tsne<- CreateDimReducObject(embeddings = embed_nkt_norm, 
                                                key = "FitSNE_", 
                                                assay = "SCT", 
                                                global = TRUE)
p48 <- DimPlot(nkt_norm, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p48
```

#### Cell Type Identification

First we'll use a Wilcoxon test to determine which genes characterize each cluster. 

```{r}
nkt_norm_markers <- FindAllMarkers(nkt_norm, 
                                   logfc.threshold = .5, 
                                   min.diff.pct = .2, 
                                   verbose = FALSE, 
                                   only.pos = TRUE, 
                                   random.seed = 629) %>% 
                    filter(p_val_adj < .05)
nkt_norm_markers  %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2) %>% 
  group_by(cluster) %>% 
  top_n(n = 3, wt = avg_log2FC) %>% 
  kbl(booktabs = TRUE, digits = 4) %>% 
  kable_minimal("hover", full_width = FALSE)
```

##### CD4+ T

We can use IL7R and S100A4 expression to identify the memory CD4+ T cells in clusters 0 & 2. IL7R and CCR7 identify the naive CD4+ T cells in the adjacent cluster 6. 

```{r}
p49 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "IL7R", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p50 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "S100A4", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p51 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "CCR7", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p49 | p50 | p51) / p48
```

##### CD8+ T

Next we reveal the CD8+ T cells in clusters 1 and 3 with CD8A, as per usual. 

```{r}
p52 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "CD8A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p52 / p48
```

##### T-reg

The T-reg cluster, cluster 5, is identified using TIGIT and FOXP3. 

```{r}
p53 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TIGIT", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p54 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "FOXP3", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p53 | p54) / p48
```

##### NK

The natural killers can be found in cluster 4 through their expression of PRF1 and NKG7. The NKG7 expression also confirms the identities of the CD8+ T cells we just annotated. 

```{r}
p55 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "PRF1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p56 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "NKG7", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p55 | p56) / p48
```

##### Proliferating T-reg

The tiny proliferating T-reg population in cluster 7 is characterized by TOP2A. 

```{r}
p57 <- FeaturePlot(nkt_norm, reduction = "tsne", features = "TOP2A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p57 / p48
```

#### Visualization

We add final cell labels to our 8 cell clusters. 

```{r}
nkt_norm$label <- case_when(nkt_norm$seurat_clusters == 0 ~ "Memory CD4+ T", 
                            nkt_norm$seurat_clusters == 1 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 2 ~ "Memory CD4+ T",
                            nkt_norm$seurat_clusters == 3 ~ "CD8+ T",
                            nkt_norm$seurat_clusters == 4 ~ "NK",
                            nkt_norm$seurat_clusters == 5 ~ "T-reg", 
                            nkt_norm$seurat_clusters == 6 ~ "Naive CD4+ T", 
                            nkt_norm$seurat_clusters == 7 ~ "Proliferating T-reg")
Idents(nkt_norm) <- "label"
p58 <- DimPlot(nkt_norm, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 2, override.aes = list(size = 4)))
p58
```

Here's the marker genes. 

```{r, fig.width=8, fig.height=4}
nkt_norm_markers2 <- FindAllMarkers(nkt_norm, 
                                    logfc.threshold = .75, 
                                    test.use = "wilcox", 
                                    only.pos = TRUE, 
                                    random.seed = 629, 
                                    verbose = FALSE) %>% 
                      filter(p_val_adj < .05) %>% 
                      mutate(source = "NK/T Adjacent Normal", 
                             log2fc_cutoff = .75)
top5_nkt_norm_markers <- nkt_norm_markers2%>% 
                         group_by(cluster) %>% 
                         arrange(desc(avg_log2FC)) %>% 
                         slice_head(n = 5)
p59 <- DotPlot(nkt_norm, features = unique(top5_nkt_norm_markers$gene), dot.scale = 15) + 
       scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
       labs(color = "Expression", size = "% Expressed") + 
       theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
             legend.position = "right", 
             legend.justification = "center", 
             panel.border = element_rect(fill = NA, size = 1, color = "black"), 
             axis.line = element_blank(), 
             legend.title = element_text(size = 18), 
             axis.title.x = element_blank(), 
             axis.title.y = element_blank(), 
             axis.text.y = element_text(size = 18)) + 
       guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
              size = guide_legend(title.position = "top", title.hjust = 0.5))
p59
```

## Ductal Cells

### Reclustering

We use KRT8 expression to show the ductal cells residing in clusters 5 and 9. We note that some regions of clusters 5 and 9 have no KRT8 expression, which likely means that they are composed of different cell types. 

```{r}
p60 <- FeaturePlot(pdac, reduction = "tsne", features = "KRT8", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p60 / p4
```

These are the cells we'll be reclustering. 

```{r}
ductal_cells <- rownames(pdac@meta.data[pdac@meta.data$seurat_clusters %in% c(5, 9), ])
p61 <- DimPlot(pdac, reduction = "tsne", cells.highlight = ductal_cells, cols.highlight = "navy", pt.size = 0.75) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend()
p61 / p4
```


We run `SCISSORS`, then use a Wilcoxon differential expression test to determine potential markers for each cluster. 

```{r}
ductal <- ReclusterCells(pdac, 
                         which.clust = list(5, 9), 
                         merge.clusters = TRUE, 
                         n.HVG = 4000, 
                         n.PC = 20, 
                         k.vals = c(20, 30, 40, 50), 
                         nn.metric = "euclidean", 
                         resolution.vals = c(.2, .3, .4), 
                         redo.embedding = TRUE, 
                         random.seed = 629)
ductal$cluster_color <- case_when(ductal$seurat_clusters == 0 ~ paletteer_d("ggsci::default_nejm")[1],
                                  ductal$seurat_clusters == 1 ~ paletteer_d("ggsci::default_nejm")[2],
                                  ductal$seurat_clusters == 2 ~ paletteer_d("ggsci::default_nejm")[3],
                                  ductal$seurat_clusters == 3 ~ paletteer_d("ggsci::default_nejm")[4],
                                  ductal$seurat_clusters == 4 ~ paletteer_d("ggsci::default_nejm")[5],
                                  ductal$seurat_clusters == 5 ~ paletteer_d("ggsci::default_nejm")[6])
ductal_markers <- FindAllMarkers(ductal, 
                                 logfc.threshold = .5, 
                                 min.diff.pct = .2, 
                                 only.pos = TRUE, 
                                 verbose = FALSE, 
                                 random.seed = 629) %>% 
                  filter(p_val_adj < .05)
ductal_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2) %>% 
  group_by(cluster) %>% 
  top_n(n = 3, wt = avg_log2FC) %>% 
  kbl(booktabs = TRUE, digits = 4) %>% 
  kable_minimal("hover", full_width = FALSE)
```

Again, we'll run Fit-SNE on the reclustered cells. 

```{r}
duct_pc <- Embeddings(ductal, reduction = "pca")
```

```{python}
# import data
duct_pc = r.duct_pc
# run Fit-SNE
affin_duct = PerplexityBasedNN(duct_pc, perplexity=30, random_state=629)
init = initialization.pca(duct_pc, random_state=629)
tsne_duct = TSNEEmbedding(init, affin_duct, negative_gradient_method='fft')
embed_d1 = tsne_duct.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_d2 = embed_d1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

We pull the results into R, making sure to save the Barnes-Hut t-SNE results in another reduction slot.

```{r}
embed_duct <- as.matrix(py$embed_d2)
rownames(embed_duct) <- colnames(ductal)
ductal@reductions$bh_tsne <- ductal@reductions$tsne
ductal@reductions$tsne<- CreateDimReducObject(embeddings = embed_duct, 
                                              key = "FitSNE_", 
                                              assay = "SCT", 
                                              global = TRUE)
p62 <- DimPlot(ductal, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p62
```

We'll also run `VAM`, using the Yeh Lab's gene sets for basal and classical PDAC. 

```{r}
adex_genes <- subtypeGeneList[[6]][subtypeGeneList[[6]]$ADEX, ]$geneSymbol
collison_exocrine_genes <- subtypeGeneList[[5]][subtypeGeneList[[5]]$Exocrine, ]$geneSymbol
bailey_immune_genes <- subtypeGeneList[[6]][subtypeGeneList[[6]]$Immunogenic, ]$geneSymbol
pdac_gene_sets <- list(classical_genes, basal_genes, adex_genes, collison_exocrine_genes, bailey_immune_genes)
names(pdac_gene_sets) <- c("Classical PDAC", "Basal-like PDAC", "ADEX", "CollisonExocrine", "BaileyImmune")
for (i in seq(pdac_gene_sets)) {
  pdac_gene_sets[[i]] <- pdac_gene_sets[[i]][pdac_gene_sets[[i]] %in% rownames(ductal)]
}
ductal <- vamForSeurat(ductal, 
                       gene.set.collection = pdac_gene_sets, 
                       gamma = TRUE)
DefaultAssay(ductal) <- "VAMcdf"
```

Examining the `VAM` results, we see that clusters 0 & 1 are enriched for Classical PDAC, and cluster 4 has very high Basal PDAC scores. 

```{r}
p63 <- FeaturePlot(ductal, reduction = "tsne", features = "Classical PDAC", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p64 <- FeaturePlot(ductal, reduction = "tsne", features = "Basal-like PDAC", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p63 | p64) / p62
FeaturePlot(ductal, reduction = "tsne", features = "ADEX", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(legend.position = "right", 
             legend.direction = "vertical")
FeaturePlot(ductal, reduction = "tsne", features = "CollisonExocrine", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(legend.position = "right", 
             legend.direction = "vertical")
data.frame(t(ductal@assays$VAMcdf@data)) %>% 
        bind_cols((ductal@meta.data %>% dplyr::select(label))) %>% 
        ggplot(aes(x = label, y = BaileyImmune, fill = label)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1.2) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(2, 5, 1, 4, 6, 3)]) +
        labs(y = "Bailey Immunogenic", x = NULL, fill = NULL) + 
        theme_minimal() + 
        theme(panel.grid = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              legend.position = "none", 
              axis.title.y = element_text(size = 24), 
              axis.text = element_text(size = 22), 
              axis.text.x = element_text(size = 22, angle = 45, vjust = 1, hjust = 1), 
              axis.ticks = element_line(), 
              plot.subtitle = element_text(face = "italic", size = 10))
```

```{r}
ductal@meta.data %>% 
  group_by(condition) %>% 
  summarise(Sum = sum(case_when(malig == "Malignant" ~ 1, TRUE ~ 0)), 
            N = n(), 
            Frac = Sum / N)

table(ductal@meta.data$condition, ductal@meta.data$label)
```


```{r}
ggplot(ductal@meta.data, aes(x = label, fill = label, y = immune)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(2, 5, 1, 4, 6, 3)]) + 
        labs(y = "DECODER exocrine", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
```


### Cell Type Identification 

#### Lipid Processing

We use ANPEP & FABP1 expression to identify the lipid processing ductal cells in cluster 6. 

```{r}
DefaultAssay(ductal) <- "SCT"
p65 <- FeaturePlot(ductal, reduction = "tsne", features = "ANPEP", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p66 <- FeaturePlot(ductal, reduction = "tsne", features = "FABP1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p65 | p66) / p62
```

#### Secretory

Expression of SOD3 and CFTR reveals the secretory cells in cluster 3.

```{r}
p67 <- FeaturePlot(ductal, reduction = "tsne", features = "SOD3", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p68 <- FeaturePlot(ductal, reduction = "tsne", features = "CFTR", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p67 | p68) / p62
```

#### Classical 1

We use TFF1 and TFF2 expression to annotate the classical 1 epithelial cells in clusters 0 and 1. We can also see that the two classical 1 clusters are split by the sample from which the cells originate. Going forward, we'll simply label both clusters as classical 1. 

```{r}
p69 <- FeaturePlot(ductal, reduction = "tsne", features = "TFF1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p70 <- FeaturePlot(ductal, reduction = "tsne", features = "TFF2", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p71 <- DimPlot(ductal, reduction = "tsne", group.by = "sample", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Sample") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p69 | p70 | p71) / p62
```

#### Classical 2

The classical 2 cells are located in cluster 6, as evidenced by their expression of CRISP3. We also see that GATA6, a canonical classical PDAC marker, is expressed in both classical clusters, but not in the putative basal cluster. 

```{r}
p72 <- FeaturePlot(ductal, reduction = "tsne", features = "CRISP3", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p73 <- FeaturePlot(ductal, reduction = "tsne", features = "GATA6", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p72 | p73) / p62
```

#### Basal

The basal compartment weights from `DECODER` are highest in cluster 4, which we will denote as being composed of basal-like PDAC. 

```{r}
p74 <- FeaturePlot(ductal, reduction = "tsne", features = "basal", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Basal") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p75 <- FeaturePlot(ductal, reduction = "tsne", features = "classical", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Classical") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p76 <- FeaturePlot(ductal, reduction = "tsne", features = "bc_ratio", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Basal:Classical") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p77 <- FeaturePlot(ductal, reduction = "tsne", features = "malig", pt.size = 1.5) + 
       scale_color_manual(values = wes_palette("Zissou1", n = 5)[c(5, 2)])  + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Malignant") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p74 | p75 | p76 | p77) / p62
```

#### Acinar

Lastly, we show that cluster 2 is composed of acinar cells using CTRB2.

```{r}
p78 <- FeaturePlot(ductal, reduction = "tsne", features = "CTRB2", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p78 / p62
```

### Visualization

We add final cluster labels to our `Seurat` object and visualize the results.

```{r}
ductal$label <- case_when(ductal$seurat_clusters == 0 ~ "Classical 1", 
                          ductal$seurat_clusters == 1 ~ "Classical 1", 
                          ductal$seurat_clusters == 2 ~ "Acinar", 
                          ductal$seurat_clusters == 3 ~ "Secretory", 
                          ductal$seurat_clusters == 4 ~ "Basal-like", 
                          ductal$seurat_clusters == 5 ~ "Lipid Proc.", 
                          ductal$seurat_clusters == 6 ~ "Classical 2")
Idents(ductal) <- "label"
p79 <- DimPlot(ductal, reduction = "tsne",pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p79
```

The Basal cluster, along with the Classical 2 cluster, shows a high percentage of malignant cells as identified by `CONICSmat`. 

```{r}
ductal@meta.data %>% 
  mutate(malig2 = case_when(malig == "Malignant" ~ 1, TRUE ~ 0)) %>% 
  group_by(label) %>% 
  summarise(M = mean(malig2)) %>% 
  mutate(M = formattable::percent(M, digits = 2)) %>% 
  kbl(booktabs = TRUE, col.names = c("Celltype", "Mean % Malignant")) %>% 
  kable_minimal(full_width = FALSE)
```

Here's the marker genes. 

```{r, fig.width=8, fig.height=4}
ductal_markers2 <- FindAllMarkers(ductal, 
                                  logfc.threshold = 2, 
                                  test.use = "wilcox", 
                                  only.pos = TRUE, 
                                  random.seed = 629, 
                                  verbose = FALSE) %>% 
                   filter(p_val_adj < .05) %>% 
                   mutate(source = "Ductal", 
                          log2fc_cutoff = 2)
top5_ductal_markers <- ductal_markers2 %>% 
                       group_by(cluster) %>% 
                       arrange(desc(avg_log2FC)) %>% 
                       slice_head(n = 5)
p80 <- DotPlot(ductal, features = unique(top5_ductal_markers$gene), dot.scale = 15) + 
       scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
       labs(color = "Expression", size = "% Expressed") + 
       theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
             legend.position = "right", 
             legend.justification = "center", 
             panel.border = element_rect(fill = NA, size = 1, color = "black"), 
             axis.line = element_blank(), 
             legend.title = element_text(size = 18), 
             axis.title.x = element_blank(), 
             axis.title.y = element_blank(), 
             axis.text.y = element_text(size = 18)) + 
       guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
              size = guide_legend(title.position = "top", title.hjust = 0.5))
p80
```

We'll also provide some visualizations of canonical markers for basal vs. classical PDAC. 

```{r}
pdac_ductal <- subset(ductal, subset = label %in% c("Classical 1", "Classical 2", "Basal-like"))
pdac_markers <- FindAllMarkers(pdac_ductal, 
                               logfc.threshold = .4, 
                               test.use = "wilcox", 
                               only.pos = TRUE, 
                               random.seed = 629, 
                               verbose = FALSE) %>% 
                filter(p_val_adj < .05) %>% 
                mutate(source = "PDAC", log2fc_cutoff = .4)
pdac_ductal@meta.data %<>% mutate(label_broad = case_when(label == "Basal-like" ~ label, TRUE ~ "Classical"))
Idents(pdac_ductal) <- "label_broad"
pdac_markers_broad <- FindAllMarkers(pdac_ductal, 
                                     logfc.threshold = .4, 
                                     test.use = "wilcox", 
                                     only.pos = TRUE, 
                                     random.seed = 629, 
                                     verbose = FALSE) %>% 
                      filter(p_val_adj < .05) %>% 
                      mutate(source = "PDAC", log2fc_cutoff = .4)
Idents(pdac_ductal) <- "label"
p80b <- t(data.frame(pdac_ductal@assays$SCT@data)) %>% 
        as.data.frame() %>% 
        mutate(label = pdac_ductal$label) %>% 
        relocate(label, GATA6) %>% 
        filter(label %in% c("Classical 1", "Classical 2", "Basal-like")) %>% 
        dplyr::select(label, GATA6) %>% 
        ggplot(aes(x = label, fill = label, y = GATA6)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) + 
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, 1.7), breaks = c(0, 0.5, 1)) + 
        labs(y = "GATA6", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
p80c <- ggplot(pdac_ductal@meta.data, aes(x = label, fill = label, y = basal)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, .82), breaks = c(0, 0.25, 0.5, 0.75)) + 
        labs(y = "DECODER basal-like", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
p80d <- ggplot(pdac_ductal@meta.data, aes(x = label, fill = label, y = classical)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, .79), breaks = c(0, 0.25, 0.5, 0.75)) + 
        labs(y = "DECODER classical", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
p80e <- ggplot(pdac_ductal@meta.data, aes(x = label, fill = label, y = bc_ratio)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, 6.2)) + 
        labs(y = "DECODER bcRatio", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
p80f <- t(data.frame(pdac_ductal@assays$VAMcdf@data)) %>% 
        as.data.frame() %>% 
        mutate(label = pdac_ductal@meta.data$label) %>% 
        ggplot(aes(x = label, fill = label, y = `Classical PDAC`)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, 1.22), breaks = c(0, 0.25, 0.5, 0.75, 1)) + 
        labs(y = "Moffitt classical", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
p80g <- t(data.frame(pdac_ductal@assays$VAMcdf@data)) %>% 
        as.data.frame() %>% 
        mutate(label = pdac_ductal@meta.data$label) %>% 
        ggplot(aes(x = label, fill = label, y = `Basal-like PDAC`)) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) +
        geom_signif(test = wilcox.test, map_signif_level = TRUE, step_increase = .12, size = 1.2, textsize = 8, 
                    comparisons = list(c("Classical 1", "Basal-like"), c("Classical 2", "Basal-like"))) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(1, 4, 5)]) + 
        scale_y_continuous(limits = c(0, 1.22), breaks = c(0, 0.25, 0.5, 0.75, 1)) + 
        labs(y = "Moffitt basal-like", title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
```

The markers / gene sets match up with our cluster labels. 

```{r, fig.width=6.5, fig.height=3}
p80h <- (p80g | p80f | p80c | p80e | p80d | p80b)
p80h
```

We'll also visualize the `CONICSmat` malignant cell annotations. 

```{r}
p80i <- DimPlot(ductal, reduction = "tsne", group.by = "malig", pt.size = 1.5) + 
        scale_color_manual(values = wes_palette("Zissou1", n = 5)[c(5, 2)]) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme(plot.title = element_blank()) + 
        theme_yehlab() + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p80i
```

Interestingly, we see that the secretory cluster has a high amount of malignant cells. Let's run GSEA on the DEGs from the secretory cluster. 

```{r}
library(clusterProfiler)
library(msigdbr)
m_df <- msigdbr(species = "Homo sapiens")
secretory_deg <- ductal_markers2 %>% 
                 filter(cluster == "Secretory") %>% 
                 select(gene, avg_log2FC) %>% 
                 inner_join((m_df %>% select(human_gene_symbol, entrez_gene) %>% distinct()),
                            by = c("gene" = "human_gene_symbol")) %>% 
                 select(entrez_gene, avg_log2FC) %>% 
                 arrange(desc(avg_log2FC))
secretory_deg_list <- secretory_deg$avg_log2FC
names(secretory_deg_list) <- secretory_deg$entrez_gene
secretory_genes <- m_df %>% 
                   filter(human_gene_symbol %in% ductal_markers2[ductal_markers2$cluster == "Secretory", ]$gene) %>% 
                   dplyr::select(gs_name, entrez_gene)
secretory_gsea <- GSEA(secretory_deg_list, TERM2GENE = secretory_genes)
secretory_enr <- enricher(names(secretory_deg_list), TERM2GENE = secretory_genes)
```


```{r}
decoder_plots <- list()
decoder_compartments <- colnames(ductal@meta.data)[14:21]
for (i in seq_along(decoder_compartments)) {
  p <- ductal@meta.data %>% 
       ggplot(aes_string(x = "label", fill = "label", y = decoder_compartments[i])) + 
        geom_violin(scale = "width", draw_quantiles = 0.5, size = 1) + 
        scale_fill_manual(values = paletteer_d("ggsci::default_nejm")[c(2, 5, 1, 4, 6, 3)]) + 
        labs(y = decoder_compartments[i], title = NULL, x = NULL) + 
        theme(panel.grid = element_blank(), 
              legend.position = "none", 
              panel.background = element_blank(), 
              panel.border = element_rect(fill = NA, size = 1), 
              axis.title.x = element_text(size = 24), 
              axis.title.y = element_text(size = 28), 
              axis.text.y = element_text(size = 20), 
              axis.text.x = element_text(size = 20, angle = 45, vjust = 1, hjust = 1))
  decoder_plots[[i]] <- p
}
```

```{r}
for (i in seq_along(decoder_plots)) {
  print(decoder_plots[[i]])
}
```

We'll also run `VAM` on the Adex-like genes from Bailey & Collison. 

```{r}
adex_genes <- subtypeGeneList[[6]][subtypeGeneList[[6]]$ADEX, ]$geneSymbol
adex_gene_set <- list(adex_genes)
names(pdac_gene_sets) <- c("Classical PDAC", "Basal-like PDAC")
for (i in seq(pdac_gene_sets)) {
  pdac_gene_sets[[i]] <- pdac_gene_sets[[i]][pdac_gene_sets[[i]] %in% rownames(ductal)]
}
ductal <- vamForSeurat(ductal, 
                       gene.set.collection = pdac_gene_sets, 
                       gamma = TRUE)
DefaultAssay(ductal) <- "VAMcdf"
```

## Plasma Cells & Plasmacytoid DCs

We use JCHAIN (denoted IGJ in Elyada *et al*) to reveal the Plasma cells in cluster 10, and IRF7 to identify the plasmacytoid DCs in cluster 11. 

```{r}
p81 <- FeaturePlot(pdac, reduction = "tsne", features = "JCHAIN", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "JCHAIN (IGJ)") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p82 <- FeaturePlot(pdac, reduction = "tsne", features = "IRF7", pt.size = 0.75) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "IRF7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p81 | p82) / p4
```

## B Cells

We can identify the B cells in cluster 4 using joint expression of MS4A1 and CD79A. The cluster is split into two subclusters by tissue type: adjacent normal and PDAC. While it would be interesting to determine the genetic drivers of that separation, it's somewhat outside of our scope here. 

```{r}
p83 <- FeaturePlot(pdac, reduction = "tsne", features = "MS4A1", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MS4A1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p84 <- FeaturePlot(pdac, reduction = "tsne", features = "CD79A", pt.size = 0.75) +
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD79A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p85 <- DimPlot(pdac, reduction = "tsne", group.by = "condition", pt.size = 0.75) + 
       scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Tissue Type") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p83 | p84 | p85) / p4
```

## Myeloid Cells

Lastly, we'll split up the myeloid population by tissue condition (PDAC vs. adjacent normal) just like we did with the NK / T cells. We'll run `SCISSORS`, annotate the clusters, and visualize the results. 

Here's the myeloid and dendritic cells we'll be reclustering. 

```{r}
myeloid_cells <- rownames(pdac@meta.data[pdac@meta.data$seurat_clusters %in% c(1, 2), ])
dendritic_cells <- rownames(pdac@meta.data[pdac@meta.data$seurat_clusters == 7, ])
p86a <- DimPlot(pdac, reduction = "tsne", cells.highlight = myeloid_cells, cols.highlight = "navy", pt.size = 0.75) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend()
p86b <- DimPlot(pdac, reduction = "tsne", cells.highlight = dendritic_cells, cols.highlight = "navy", pt.size = 0.75) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend()
(p86a | p86b) / p4
```

### Tumor 

First we'll attempt to assign broad cell type labels to each of the four putative myeloid clusters. We'll use the marker genes from Elyada *et al* once again. 

```{r}
myo_tumor <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 7) & condition == "PDAC")
p87 <- DimPlot(myo_tumor, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p87
```

Cluster 1 appears to be composed of our resident & alternatively activated macrophages due to its expression of CD14 & C1QA and SPP1, respectively. 

```{r}
p88 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "CD14", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p89 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "C1QA", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p90 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "SPP1", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "SPP1") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p88 | p89 | p90) / p87
```

We can use LYZ and S100A8 expression to reveal the classic monocytes and neutrophils in cluster 2. 

```{r}
p91 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "LYZ", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LYZ") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p92 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "S100A8", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p91 | p92) / p87
```

Lastly, we show that cluster 6 contains our various DC subtypes through its expression of FCER1A, a canonical dendritic cell marker. 

```{r}
p93 <- FeaturePlot(myo_tumor, reduction = "tsne", features = "FCER1A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FCER1A") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p93 / p87
```

#### Reclustering

We'll start with the Macrophages & Monocytes, since the DC populations are small and are best dealt with on their own. 

```{r}
myo_reclust <- ReclusterCells(myo_tumor, 
                              which.clust = c(1, 2), 
                              n.PC = 15, 
                              merge.clusters = TRUE, 
                              k.vals = c(40, 50, 60), 
                              resolution.vals = c(.2, .3, .4), 
                              n.HVG = 4000, 
                              redo.embedding = TRUE, 
                              random.seed = 629)
dc_reclust <- ReclusterCells(myo_tumor, 
                             which.clust = 7, 
                             n.PC = 15, 
                             k.vals = c(20, 30, 40, 50), 
                             resolution.vals = c(.3, .4, .5), 
                             n.HVG = 4000, 
                             nn.metric = "euclidean", 
                             redo.embedding = TRUE, 
                             random.seed = 629)
```

We'll again run Fit-SNE on our reclustered cells. 

```{r}
mono_tumor_pc <- Embeddings(myo_reclust, "pca")
dc_tumor_pc <- Embeddings(dc_reclust, "pca")
```

```{python}
# import data
mono_pc = r.mono_tumor_pc
dc_pc = r.dc_tumor_pc
# Fit-SNE - monocytes & macrophages
affin_mono = PerplexityBasedNN(mono_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(mono_pc, random_state=629)
tsne_mono = TSNEEmbedding(init, affin_mono, negative_gradient_method='fft')
embed_mono1 = tsne_mono.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_mono2 = embed_mono1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
# Fit-SNE - DC
affin_dc = PerplexityBasedNN(dc_pc, perplexity=30, random_state=629)
init = initialization.pca(dc_pc, random_state=629)
tsne_dc = TSNEEmbedding(init, affin_dc, negative_gradient_method='fft')
embed_dc1 = tsne_dc.optimize(n_iter=250, exaggeration=12, momentum=0.6)
embed_dc2 = embed_dc1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

We pull the results back into R and visualize them. 

```{r}
embed_mono <- as.matrix(py$embed_mono2)
rownames(embed_mono) <- colnames(myo_reclust)
myo_reclust@reductions$bh_tsne <- myo_reclust@reductions$tsne
myo_reclust@reductions$tsne<- CreateDimReducObject(embeddings = embed_mono, 
                                                   key = "FitSNE_", 
                                                   assay = "SCT",
                                                   global = TRUE)
p94 <- DimPlot(myo_reclust, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p94
```

```{r}
embed_dc <- as.matrix(py$embed_dc2)
rownames(embed_dc) <- colnames(dc_reclust)
dc_reclust@reductions$bh_tsne <- dc_reclust@reductions$tsne
dc_reclust@reductions$tsne<- CreateDimReducObject(embeddings = embed_dc, 
                                                  key = "FitSNE_", 
                                                  assay = "SCT",
                                                  global = TRUE)
p95 <- DimPlot(dc_reclust, reduction = "tsne", pt.size = 1.5) + 
       scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
       theme_yehlab() + 
       theme(plot.title = element_blank()) + 
       guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p95
```

#### Cell Type Identification

##### Monoctyes, Macrophages, & Neutrophils

First we ID the neutrophils in cluster 2 using S100A8 and S100A9 - marker genes used by Elyada *et al*. 

```{r}
p96 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "S100A8", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A8") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p97 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "S100A9", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A9") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p96 | p97) / p94
```

We can identify the classical monocytes in cluster 1 through their expression of CD14 and lack of expression of CD16 aka FCGR3A, expression of which, alongside that of MS4A7, reveals the group of CD16+ monocytes in cluster 4. Finally, expression of those genes [as well as S100A10](https://www.frontiersin.org/articles/10.3389/fimmu.2019.02035/full#B3) allows us to defined cluster 5 as containing intermediate monocytes. 

```{r}
p98 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "CD14", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD14") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p99 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "FCGR3A", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "FCGR3A (CD16)") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p100 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "MS4A7", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "MS4A7") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
p101 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "S100A10", pt.size = 1.5) + 
       scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
       labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "S100A10") + 
       theme_yehlab() + 
       NoLegend() + 
       theme(axis.title = element_blank())
(p98 | p99 | p100 | p101) / p94
```

Expression of C1QA, APOE, and SPP1 show us the resident and alternatively activated macrophages in clusters 0 and 3, respectively. 

```{r}
p102 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "C1QA", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "C1QA") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p103 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "APOE", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "APOE") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p104 <- FeaturePlot(myo_reclust, reduction = "tsne", features = "SPP1", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "SPP1") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p102 | p103 | p104) / p94
```

We add cell labels to our `Seurat` object, then we're off to the DCs. 

```{r}
myo_reclust$label <- case_when(myo_reclust$seurat_clusters == 0 ~ "Resident Macrophage", 
                               myo_reclust$seurat_clusters == 1 ~ "Classical Monocyte", 
                               myo_reclust$seurat_clusters == 2 ~ "Neutrophil", 
                               myo_reclust$seurat_clusters == 3 ~ "Alt. Activated Macrophage", 
                               myo_reclust$seurat_clusters == 4 ~ "CD16+ Monocyte", 
                               myo_reclust$seurat_clusters == 5 ~ "Intermediate Monocyte")
Idents(myo_reclust) <- "label"
```

##### Dendritic Cells

We can use CLEC9A to annotate the cDC1 population in cluster 5. 

```{r}
p105 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "CLEC9A", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CLEC9A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p105 / p95
```

High and low expression of CD1A and CD207 reveal the Langerhans-like DCB and DCA cells in clusters 3 and 4, respectively. 

```{r}
p106 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "CD1A", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD1A") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p107 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "CD207", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "CD207") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p106 | p107) / p95
```

Next we use LAMP3 to identify the activated DCs in cluster 6. 

```{r}
p108 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "LAMP3", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "LAMP3") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p108 / p95
```

Lastly, we use expression of two [canonical cDC2 marker genes / transcription factors](https://onlinelibrary.wiley.com/doi/pdf/10.1111/imm.12888) to identify clusters 0, 1, and 2 as cDC2 cells, which are split by sample ID. 

```{r}
p109 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "CLEC10A", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p110 <- FeaturePlot(dc_reclust, reduction = "tsne", features = "KLF4", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p111 <- DimPlot(dc_reclust, reduction = "tsne", group.by = "sample", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Sample") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p109 | p110 | p111) / p95
```

We add final subcluster cell type labels to our `Seurat` object, then we're off to the adjacent normal myeloid cells. 

```{r}
dc_reclust$label <- case_when(dc_reclust$seurat_clusters == 0 ~ "cDC2", 
                              dc_reclust$seurat_clusters == 1 ~ "cDC2", 
                              dc_reclust$seurat_clusters == 2 ~ "cDC2", 
                              dc_reclust$seurat_clusters == 3 ~ "Langerhans-like DCB", 
                              dc_reclust$seurat_clusters == 4 ~ "Langerhans-like DCA", 
                              dc_reclust$seurat_clusters == 5 ~ "cDC1", 
                              dc_reclust$seurat_clusters == 6 ~ "Activated DC")
Idents(dc_reclust) <- "label"
```

#### Visualization

##### Monocytes & Neutrophils

```{r}
p112 <- DimPlot(myo_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 4)))
p112
```

Here's the marker genes. 

```{r, fig.width=8, fig.height=4}
myo_reclust_markers2 <- FindAllMarkers(myo_reclust, 
                                       logfc.threshold = 1, 
                                       test.use = "wilcox", 
                                       only.pos = TRUE, 
                                       random.seed = 629, 
                                       verbose = FALSE) %>% 
                        filter(p_val_adj < .05) %>% 
                        mutate(source = "Myeloid Tumor", 
                               log2fc_cutoff = 1)
top5_myo_reclust_markers <- myo_reclust_markers2 %>% 
                            group_by(cluster) %>% 
                            arrange(desc(avg_log2FC)) %>% 
                            slice_head(n = 5)
p113 <- DotPlot(myo_reclust, features = unique(top5_myo_reclust_markers$gene), dot.scale = 15) + 
        scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
        labs(color = "Expression", size = "% Expressed") + 
        theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
              legend.position = "right", 
              legend.justification = "center", 
              panel.border = element_rect(fill = NA, size = 1, color = "black"), 
              axis.line = element_blank(), 
              legend.title = element_text(size = 18), 
              axis.title.x = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_text(size = 18)) + 
        guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
               size = guide_legend(title.position = "top", title.hjust = 0.5))
p113
```

##### DCs

```{r}
p114 <- DimPlot(dc_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 2, override.aes = list(size = 3)))
p114
```

Here's the marker genes. 

```{r, fig.width=8, fig.height=4}
dc_reclust_markers2 <- FindAllMarkers(dc_reclust, 
                                      logfc.threshold = 1, 
                                      test.use = "wilcox", 
                                      only.pos = TRUE, 
                                      random.seed = 629, 
                                      verbose = FALSE) %>% 
                       filter(p_val_adj < .05) %>% 
                       mutate(source = "DC Tumor", 
                              log2fc_cutoff = 1)
top5_dc_reclust_markers <- dc_reclust_markers2 %>% 
                           group_by(cluster) %>% 
                           arrange(desc(avg_log2FC)) %>% 
                           slice_head(n = 5)
p115 <- DotPlot(dc_reclust, features = unique(top5_dc_reclust_markers$gene), dot.scale = 15) + 
        scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
        labs(color = "Expression", size = "% Expressed") + 
        theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
              legend.position = "right", 
              legend.justification = "center", 
              panel.border = element_rect(fill = NA, size = 1, color = "black"), 
              axis.line = element_blank(), 
              legend.title = element_text(size = 18), 
              axis.title.x = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_text(size = 18)) + 
        guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
               size = guide_legend(title.position = "top", title.hjust = 0.5))
p115
```

### Adjacent Normal

We have the same four clusters as in the tumor tissue - 1, 2, & 7 - and we'll run the same analysis steps.  

```{r}
myo_norm <- subset(pdac, subset = seurat_clusters %in% c(1, 2, 7) & condition == "AdjNorm")
```

#### Reclustering

```{r}
myo_norm_reclust <- ReclusterCells(myo_norm, 
                                   which.clust = c(1, 2), 
                                   n.PC = 15, 
                                   merge.clusters = TRUE, 
                                   k.vals = c(20, 30, 40), 
                                   resolution.vals = c(.1, .2), 
                                   n.HVG = 4000, 
                                   redo.embedding = TRUE, 
                                   random.seed = 629)
dc_norm_reclust <- ReclusterCells(myo_norm, 
                                  which.clust = 7, 
                                  n.PC = 15, 
                                  k.vals = c(20, 30, 40, 50), 
                                  resolution.vals = c(.3, .4, .5), 
                                  n.HVG = 4000, 
                                  nn.metric = "euclidean", 
                                  redo.embedding = TRUE, 
                                  random.seed = 629)
```

For the last time, we run Fit-SNE in order to obtain a better embedding. 

```{r}
mono_norm_pc <- Embeddings(myo_norm_reclust, "pca")
dc_norm_pc <- Embeddings(dc_norm_reclust, "pca")
```

```{python}
# import data
mono_pc = r.mono_norm_pc
dc_pc = r.dc_norm_pc
# Fit-SNE - monocytes
affin_mono = PerplexityBasedNN(mono_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(mono_pc, random_state=629)
tsne_mono = TSNEEmbedding(init, affin_mono, negative_gradient_method='fft')
embed_mono1 = tsne_mono.optimize(n_iter=250, exaggeration=10, momentum=0.6)
embed_mono2 = embed_mono1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
# Fit-SNE - DC
affin_dc = PerplexityBasedNN(dc_pc, perplexity=30, metric='cosine', random_state=629)
init = initialization.pca(dc_pc, random_state=629)
tsne_dc = TSNEEmbedding(init, affin_dc, negative_gradient_method='fft')
embed_dc1 = tsne_dc.optimize(n_iter=250, exaggeration=8, momentum=0.6)
embed_dc2 = embed_dc1.optimize(n_iter=750, exaggeration=1, momentum=0.8)
```

We pull the results back into R and visualize them. 

```{r}
embed_mono <- as.matrix(py$embed_mono2)
rownames(embed_mono) <- colnames(myo_norm_reclust)
myo_norm_reclust@reductions$bh_tsne <- myo_norm_reclust@reductions$tsne
myo_norm_reclust@reductions$tsne<- CreateDimReducObject(embeddings = embed_mono, 
                                                        key = "FitSNE_", 
                                                        assay = "SCT",
                                                        global = TRUE)
p116 <- DimPlot(myo_norm_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p116
```

```{r}
embed_dc <- as.matrix(py$embed_dc2)
rownames(embed_dc) <- colnames(dc_norm_reclust)
dc_norm_reclust@reductions$bh_tsne <- dc_norm_reclust@reductions$tsne
dc_norm_reclust@reductions$tsne<- CreateDimReducObject(embeddings = embed_dc, 
                                                       key = "FitSNE_", 
                                                       assay = "SCT",
                                                       global = TRUE)
p117 <- DimPlot(dc_norm_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p117
```

#### Cell Type Identification

##### Monocytes, Macrophages, & Neutrophils

We use high S100A8 and S100A9 expression to define the neutrophils in cluster 4. 

```{r}
p118 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "S100A8", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p119 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "S100A9", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p118 | p119) / p116
```

Next up are the classical monocytes in clusters 0 and 3, split by sample. 

```{r}
p120 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "LYZ", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p121 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "CD14", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p122 <- DimPlot(myo_norm_reclust, reduction = "tsne", group.by = "sample", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("miscpalettes::brightPastel")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2", title = "Sample") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p120 | p121 | p122) / p116
```

C1QA and APOE show us the resident macrophages in clusters 1 and 2. 

```{r}
p123 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "C1QA", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p124 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "APOE", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p123 | p124) / p116
```

Lastly, it seems we have a small group of CD8+ T cells in cluster 5 that snuck into the myeloid cluster, as defined by their expression of CD3D (marking them as NK / T cells), and CD8A & NKG7. I don't believe they're NK cells as they do not express PRF1 or GZMB. 

```{r}
p125 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "CD3D") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p126 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "CD8A") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p127 <- FeaturePlot(myo_norm_reclust, reduction = "tsne", features = "NKG7") + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p125 | p126 | p127) / p116
```

We add labels to our clusters. 

```{r}
myo_norm_reclust$label <- case_when(myo_norm_reclust$seurat_clusters == 0 ~ "Classical Monocyte", 
                                    myo_norm_reclust$seurat_clusters == 1 ~ "Resident Macrophage", 
                                    myo_norm_reclust$seurat_clusters == 2 ~ "Resident Macrophage", 
                                    myo_norm_reclust$seurat_clusters == 3 ~ "Classical Monocyte", 
                                    myo_norm_reclust$seurat_clusters == 4 ~ "Neutrophil",
                                    myo_norm_reclust$seurat_clusters == 5 ~ "CD8+ T")
Idents(myo_norm_reclust) <- "label"
```

##### Dendritic Cells

We annotate cluster 1 as conventional DC1 and cluster 0 as conventional DC2 through mutually exclusive expression of the canonical markers CLEC9A and CLEC10A, respectively. 

```{r}
p128 <- FeaturePlot(dc_norm_reclust, reduction = "tsne", features = "CLEC9A", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
p129 <- FeaturePlot(dc_norm_reclust, reduction = "tsne", features = "CLEC10A", pt.size = 1.5) + 
        scale_color_gradientn(colors = wesanderson::wes_palette("Zissou1")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        NoLegend() + 
        theme(axis.title = element_blank())
(p128 | p129) / p117
```

Labels are added to our adjacent normal tissue DC `Seurat` object. 

```{r}
dc_norm_reclust$label <- case_when(dc_norm_reclust$seurat_clusters == 0 ~ "cDC2", 
                                   dc_norm_reclust$seurat_clusters == 1 ~ "cDC2", 
                                   dc_norm_reclust$seurat_clusters == 2 ~ "cDC1")
Idents(dc_norm_reclust) <- "label"
```

#### Visualization

##### Monocytes, Macrophages, & Neutrophils

Here are the final annotations for the adjacent normal tissue myeloid population. 

```{r}
p130 <- DimPlot(myo_norm_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p130
```

And here's their marker genes. 

```{r, fig.width=8, fig.height=4}
myo_norm_reclust_markers2 <- FindAllMarkers(myo_norm_reclust, 
                                            logfc.threshold = 1, 
                                            test.use = "wilcox", 
                                            only.pos = TRUE, 
                                            random.seed = 629, 
                                            verbose = FALSE) %>% 
                             filter(p_val_adj < .05) %>% 
                             mutate(source = "Myeloid Adjacent Normal", 
                                    log2fc_cutoff = 1)
top5_myo_norm_reclust_markers <- myo_norm_reclust_markers2 %>% 
                                 group_by(cluster) %>% 
                                 arrange(desc(avg_log2FC)) %>% 
                                 slice_head(n = 5)
p131 <- DotPlot(myo_norm_reclust, features = unique(top5_myo_norm_reclust_markers$gene), dot.scale = 15) + 
        scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
        labs(color = "Expression", size = "% Expressed") + 
        theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
              legend.position = "right", 
              legend.justification = "center", 
              panel.border = element_rect(fill = NA, size = 1, color = "black"), 
              axis.line = element_blank(), 
              legend.title = element_text(size = 18), 
              axis.title.x = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_text(size = 18)) + 
        guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
               size = guide_legend(title.position = "top", title.hjust = 0.5))
p131
```

##### Dendritic Cells

Here's the final DC annotations.

```{r}
p132 <- DimPlot(dc_norm_reclust, reduction = "tsne", pt.size = 1.5) + 
        scale_color_manual(values = paletteer_d("ggsci::default_nejm")) + 
        labs(x = "Fit-SNE 1", y = "Fit-SNE 2") + 
        theme_yehlab() + 
        theme(plot.title = element_blank()) + 
        guides(color = guide_legend(nrow = 1, override.aes = list(size = 4)))
p132
```

And here are their marker genes. 

```{r, fig.width=6, fig.height=3}
dc_norm_reclust_markers2 <- FindAllMarkers(dc_norm_reclust, 
                                           logfc.threshold = 1, 
                                           test.use = "wilcox", 
                                           only.pos = TRUE, 
                                           random.seed = 629, 
                                           verbose = FALSE) %>% 
                            filter(p_val_adj < .05) %>% 
                            mutate(source = "DC Adjacent Normal", 
                                   log2fc_cutoff = 1)
top5_dc_norm_reclust_markers <- dc_norm_reclust_markers2 %>% 
                                group_by(cluster) %>% 
                                arrange(desc(avg_log2FC)) %>% 
                                slice_head(n = 5)
p133 <- DotPlot(dc_norm_reclust, features = unique(top5_dc_norm_reclust_markers$gene), dot.scale = 15) + 
        scale_color_gradientn(colors = paletteer_d("wesanderson::Zissou1")) +
        labs(color = "Expression", size = "% Expressed") + 
        theme(axis.text.x = element_text(angle = 90, size = 16, vjust = 0.5), 
              legend.position = "right", 
              legend.justification = "center", 
              panel.border = element_rect(fill = NA, size = 1, color = "black"), 
              axis.line = element_blank(), 
              legend.title = element_text(size = 18), 
              axis.title.x = element_blank(), 
              axis.title.y = element_blank(), 
              axis.text.y = element_text(size = 18)) + 
        guides(color = guide_colorbar(title.position = "top", barheight = unit(3, units = "cm"), title.hjust = 0.5), 
               size = guide_legend(title.position = "top", title.hjust = 0.5))
p133
```

# Final CAF Analysis 

Let's add labels to all the cells based on our analysis. Our goal is to find marker genes for the three CAF subtypes. To do this, we'll use the DEGs we generated earlier when comparing only CAF vs. other CAF. We'll then filter out any genes that are in the top 10th percentile of normalized expression in any of the other major celltypes. 

```{r}
icaf_cells <- names(caf$label)[caf$label == "iCAF"]
mycaf_cells <- names(caf$label)[caf$label == "myCAF"]
apcaf_cells <- names(caf$label)[caf$label == "apCAF"]
basal_cells <- names(pdac_ductal$label)[pdac_ductal$label_broad == "Basal-like"]
classical_cells <- names(pdac_ductal$label)[pdac_ductal$label_broad == "Classical"]
pdac@meta.data %<>% 
  mutate(label = case_when(seurat_clusters == 0 ~ "T", 
                           seurat_clusters == 1 ~ "Myeloid", 
                           seurat_clusters == 2 ~ "Myeloid", 
                           seurat_clusters == 3 ~ "T", 
                           seurat_clusters == 4 ~ "B", 
                           seurat_clusters == 5 ~ "Ductal", 
                           seurat_clusters == 6 ~ "Stroma", 
                           seurat_clusters == 7 ~ "DC", 
                           seurat_clusters == 8 ~ "NK", 
                           seurat_clusters == 9 ~ "Acinar", 
                           seurat_clusters == 10 ~ "Plasma", 
                           seurat_clusters == 11 ~ "pDC"), 
         label_fine = case_when(rownames(pdac@meta.data) %in% icaf_cells ~ "iCAF", 
                                rownames(pdac@meta.data) %in% mycaf_cells ~ "myCAF", 
                                rownames(pdac@meta.data) %in% apcaf_cells ~ "apCAF", 
                                rownames(pdac@meta.data) %in% basal_cells ~ "Basal-like PDAC", 
                                rownames(pdac@meta.data) %in% classical_cells  ~ "Classical PDAC", 
                                TRUE ~ label), 
         CAF_ind = case_when(rownames(pdac@meta.data) %in% colnames(caf) ~ "CAF", TRUE ~ "non-CAF"), 
         basal_ind = case_when(rownames(pdac@meta.data) %in% basal_cells ~ "Basal-like", TRUE ~ "non-Basal-like"), 
         classical_ind = case_when(rownames(pdac@meta.data) %in% classical_cells ~ "Classical", TRUE ~ "non-Classical"))
Idents(pdac) <- "label_fine"
# get mean expression for each gene per non-CAF celltype
gene_means_by_clust <- t(pdac@assays$SCT@data) %>% 
                       as.data.frame() %>% 
                       mutate(label_fine = pdac$label_fine) %>% 
                       filter(!label_fine %in% c("iCAF", "myCAF", "apCAF")) %>% 
                       group_by(label_fine) %>% 
                       summarise(across(where(is.numeric), mean))
# find list of genes w/ mean expression above 90th percentile of expression in each celltype
high_exp_genes <- c()
for (i in gene_means_by_clust$label_fine) {
  top_exp_genes <- gene_means_by_clust %>% 
                   filter(label_fine == i) %>% 
                   dplyr::select(-label_fine) %>% 
                   t() %>% 
                   as.data.frame() %>% 
                   filter(V1 > quantile(V1, .9)) %>% 
                   mutate(gene = rownames(.)) %>% 
                   pull(gene)
  high_exp_genes <- c(high_exp_genes, top_exp_genes)
}
high_exp_genes <- unique(high_exp_genes)
# find highly expressed macrophage genes from Blueprint Encode macrophage RNA-seq samples
# necessary because apCAF have an immune-like transcriptomic profile
bed <- celldex::BlueprintEncodeData()
bed <- bed[, bed$label.main == "Macrophages"]
top_macro_genes <- bed@assays@data@listData[["logcounts"]] %>% 
                   as.data.frame() %>% 
                   t() %>% 
                   as.data.frame() %>% 
                   summarise(across(everything(), mean)) %>% 
                   t() %>% 
                   as.data.frame() %>% 
                   filter(V1 > quantile(V1, .75)) %>% 
                   mutate(gene = rownames(.)) %>% 
                   pull(gene)
# filter out genes highly expressed in any other celltypes or highly expressed in macrophages
caf_specific_markers <- caf_markers %>% 
                        filter(!gene %in% high_exp_genes) %>% 
                        filter(!gene %in% top_macro_genes)
top10_caf_specific_markers <- caf_specific_markers %>% 
                              arrange(desc(avg_log2FC)) %>% 
                              group_by(cluster) %>% 
                              slice_head(n = 10)
top25_caf_specific_markers <- caf_specific_markers %>% 
                              arrange(desc(avg_log2FC)) %>% 
                              group_by(cluster) %>% 
                              slice_head(n = 25)
top10_caf_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  kbl(booktabs = TRUE, digits = 3, row.names = FALSE) %>% 
  kable_minimal("hover", full_width = FALSE)
```

Let's save the CAF gene sets.

```{r}
top10_caf_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top10_CAF_Gene_Set.csv", col_names = TRUE)
top25_caf_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top25_CAF_Gene_Set.csv", col_names = TRUE)
```

Let's perform some further analyses with those gene sets. 

```{r}
library(clusterProfiler)
library(msigdbr)
m_df <- msigdbr(species = "Homo sapiens")
caf_genes <- caf_specific_markers %>% 
             dplyr::select(cluster, gene, avg_log2FC) %>% 
             inner_join((m_df %>% select(human_gene_symbol, entrez_gene) %>% distinct()),
                        by = c("gene" = "human_gene_symbol")) %>% 
             dplyr::select(cluster, gene, entrez_gene, avg_log2FC) %>% 
             group_by(cluster) %>% 
             arrange(desc(avg_log2FC))
icaf_deg_list <- caf_genes %>% filter(cluster == "iCAF") %>% pull(avg_log2FC)
mycaf_deg_list <- caf_genes %>% filter(cluster == "myCAF") %>% pull(avg_log2FC)
apcaf_deg_list <- caf_genes %>% filter(cluster == "apCAF") %>% pull(avg_log2FC)
names(icaf_deg_list) <- caf_genes %>% filter(cluster == "iCAF") %>% pull(entrez_gene)
names(mycaf_deg_list) <- caf_genes %>% filter(cluster == "myCAF") %>% pull(entrez_gene)
names(apcaf_deg_list) <- caf_genes %>% filter(cluster == "apCAF") %>% pull(entrez_gene)
gene_db <- m_df %>% 
           filter(human_gene_symbol %in% caf_genes$gene) %>% 
           dplyr::select(gs_name, entrez_gene)
icaf_enr <- enricher(names(icaf_deg_list), TERM2GENE = gene_db)
mycaf_enr <- enricher(names(mycaf_deg_list), TERM2GENE = gene_db)
apcaf_enr <- enricher(names(apcaf_deg_list), TERM2GENE = gene_db)
```

# Final Basal / Clasical Gene Sets

We'll do what we just did to derive a CAF signature gene set, but this time for the basal / classical cells. 

```{r}
# get mean expression for each gene per non-PDAC celltype
gene_means_by_clust <- t(pdac@assays$SCT@data) %>% 
                       as.data.frame() %>% 
                       mutate(label_fine = pdac$label_fine) %>% 
                       filter(!label_fine %in% c("Classical PDAC", "Basal-like PDAC")) %>% 
                       group_by(label_fine) %>% 
                       summarise(across(where(is.numeric), mean))
# find list of genes w/ mean expression above 90th percentile of expression in each celltype
high_exp_genes <- c()
for (i in gene_means_by_clust$label_fine) {
  top_exp_genes <- gene_means_by_clust %>% 
                   filter(label_fine == i) %>% 
                   dplyr::select(-label_fine) %>% 
                   t() %>% 
                   as.data.frame() %>% 
                   filter(V1 > quantile(V1, .9)) %>% 
                   mutate(gene = rownames(.)) %>% 
                   pull(gene)
  high_exp_genes <- c(high_exp_genes, top_exp_genes)
}
high_exp_genes <- unique(high_exp_genes)
# filter out genes highly expressed in any other celltypes
pdac_specific_markers <- pdac_markers %>% filter(!gene %in% high_exp_genes)
top10_pdac_specific_markers <- pdac_specific_markers %>% 
                               arrange(desc(avg_log2FC)) %>% 
                               group_by(cluster) %>% 
                               slice_head(n = 10)
top25_pdac_specific_markers <- pdac_specific_markers %>% 
                               arrange(desc(avg_log2FC)) %>% 
                               group_by(cluster) %>% 
                               slice_head(n = 25)
top10_pdac_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  kbl(booktabs = TRUE, digits = 3, row.names = FALSE) %>% 
  kable_minimal("hover", full_width = FALSE)
pdac_specific_markers_broad <- pdac_markers_broad %>% filter(!gene %in% high_exp_genes)
top10_pdac_specific_markers_broad <- pdac_specific_markers_broad %>% 
                                     arrange(desc(avg_log2FC)) %>% 
                                     group_by(cluster) %>% 
                                     slice_head(n = 10)
top25_pdac_specific_markers_broad <- pdac_specific_markers_broad %>% 
                                     arrange(desc(avg_log2FC)) %>% 
                                     group_by(cluster) %>% 
                                     slice_head(n = 25)
top10_pdac_specific_markers_broad %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  kbl(booktabs = TRUE, digits = 3, row.names = FALSE) %>% 
  kable_minimal("hover", full_width = FALSE)
```

Let's save the Basal-like / Classical gene sets. 

```{r}
top10_pdac_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top10_PDAC_Fine_Gene_Set.csv", col_names = TRUE)
top25_pdac_specific_markers %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top25_PDAC_Fine_Gene_Set.csv", col_names = TRUE)
top10_pdac_specific_markers_broad %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top10_PDAC_Broad_Gene_Set.csv", col_names = TRUE)
top25_pdac_specific_markers_broad %>% 
  dplyr::select(cluster, gene, avg_log2FC, p_val_adj, pct.1, pct.2, source, log2fc_cutoff) %>% 
  readr::write_csv(file = "~/Desktop/Data/Elyada_Top25_PDAC_Broad_Gene_Set.csv", col_names = TRUE)
```

# Conclusions

`SCISSORS` was able to, with the help of several other methods, identify many celltypes - including celltypes that were not discovered in the original analysis done by Elyada *et al*. It also did so in a reproducible way, which is paramount in the era of computational biology. We believe this analysis shows the promise of `SCISSORS` as a valuable piece of the scRNA-seq data exploration and annotation processes. 

# Save Data & Figures

First we'll save the final PDAC object. 

```{r, eval=FALSE}
saveRDS(pdac, "~/Desktop/Data/Elyada.Rds")
saveRDS(nkt_tumor, "~/Desktop/Data/Elyada_nkt_tumor.Rds")
saveRDS(nkt_norm, "~/Desktop/Data/Elyada_nkt_adjNorm.Rds")
saveRDS(fibro, "~/Desktop/Data/Elyada_fibro.Rds")
saveRDS(ductal, "~/Desktop/Data/Elyada_ductal.Rds")
saveRDS(myo_reclust, "~/Desktop/Data/Elyada_myeloid_tumor.Rds")
saveRDS(myo_norm_reclust, "~/Desktop/Data/Elyada_myeloid_adjNorm.Rds")
saveRDS(dc_reclust, "~/Desktop/Data/Elyada_DC_tumor.Rds")
saveRDS(dc_norm_reclust, "~/Desktop/Data/Elyada_DC_adjNorm.Rds")
```

Next we'll create a final table of all the differentially expressed genes from the `SCISSORS` results for each celltype, and save them into one large Excel document. 

```{r, eval=FALSE}
caf_markers %>% 
  bind_rows(nkt_tumor_markers2) %>% 
  bind_rows(nkt_norm_markers2) %>% 
  bind_rows(ductal_markers2) %>% 
  bind_rows(myo_reclust_markers2) %>% 
  bind_rows(dc_reclust_markers2) %>% 
  bind_rows(myo_norm_reclust_markers2) %>% 
  bind_rows(dc_norm_reclust_markers2) -> SCISSORS_de_results
openxlsx::write.xlsx(SCISSORS_de_results, file = "./Data/Elyada_SCISSORS_Marker_Genes.xlsx")
```

We'll create a quick convenience function to help us save the figures.

```{r}
SaveFigure <- function(my.plot = NULL, name = NULL, height = 8, width = 8) {
  if (is.null(plot) | is.null(name)) stop("You forgot some arguments.")
  # save figure as is - w/ axis labels, titles, etc. 
  dir <- "~/Desktop/R/SCISSORS/vignettes/figures_supp/Elyada"
  ggsave(my.plot, 
         filename = paste0(name, ".pdf"), 
         device = "pdf", 
         units = "in",
         path = dir, 
         height = height, 
         width = width) 
  # save "blank" figure w/ no labels, legends, etc.
  dir <- "~/Desktop/R/SCISSORS/vignettes/figures_pub/Elyada"
  plot_blank <- my.plot + 
                theme(axis.title = element_blank(), 
                      panel.border = element_blank(), 
                      plot.title = element_blank(), 
                      plot.subtitle = element_blank(), 
                      plot.caption = element_blank(), 
                      legend.position = "none")
  ggsave(plot_blank, 
         filename = paste0(name, ".pdf"), 
         device = "pdf", 
         units = "in",
         path = dir, 
         height = height, 
         width = width) 
}
```

This section isn't worth reading; it's here solely to prove that the figures we present in our publication were dynamically generated during the knitting of this document.

```{r, eval=FALSE}
SaveFigure(my.plot = p0, name = "Seurat_Clusters_tSNE")
SaveFigure(my.plot = p1, name = "Seurat_Clusters_Silhouette_Scores")
SaveFigure(my.plot = p2, name = "Seurat_Clusters_Dotplot", height = 8, width = 18)
SaveFigure(my.plot = p3, name = "Seurat_Clusters_UMAP")
SaveFigure(my.plot = p4, name = "Seurat_Clusters_FitSNE")
SaveFigure(my.plot = p5, name = "SingleR_scRNA_Annos_FitSNE")
SaveFigure(my.plot = p6, name = "SingleR_bulkRNA_Annos_FitSNE")
SaveFigure(my.plot = p7, name = "CONICSmat_Annos_FitSNE")
SaveFigure(my.plot = p8, name = "DECODER_Basal_PDAC_FitSNE")
SaveFigure(my.plot = p9, name = "DECODER_Classical_PDAC_FitSNE")
SaveFigure(my.plot = p10, name = "DECODER_Exocrine_FitSNE")
SaveFigure(my.plot = p11, name = "DECODER_Endocrine_FitSNE")
SaveFigure(my.plot = p12, name = "DECODER_Immune_FitSNE")
SaveFigure(my.plot = p13, name = "DECODER_Normal_Stroma_FitSNE")
SaveFigure(my.plot = p14, name = "DECODER_Activated_Stroma_FitSNE")
SaveFigure(my.plot = p15, name = "All_Cells_Stroma_COL1A1_FitSNE")
SaveFigure(my.plot = p16, name = "All_Cells_Stroma_COL3A1_FitSNE")
SaveFigure(my.plot = p17, name = "All_Cells_Stroma_LUM_FitSNE")
SaveFigure(my.plot = p18, name = "All_Cells_Stroma_DCN_FitSNE")
SaveFigure(my.plot = p19, name = "All_Cells_Highlight_Stroma_FitSNE")
SaveFigure(my.plot = p20, name = "SCISSORS_Clusters_Stroma_FitSNE")
SaveFigure(my.plot = p21, name = "SCISSORS_Clusters_Stroma_Endothelial_PLVAP")
SaveFigure(my.plot = p22, name = "SCISSORS_Clusters_Stroma_Perivascular_RGS5")
SaveFigure(my.plot = p23, name = "SCISSORS_Clusters_Stroma_VAM_iCAF")
SaveFigure(my.plot = p24, name = "SCISSORS_Clusters_Stroma_VAM_myCAF")
SaveFigure(my.plot = p25, name = "SCISSORS_Clusters_Stroma_VAM_apCAF")
SaveFigure(my.plot = p26, name = "SCISSORS_Clusters_Stroma_Labels_FitSNE")
SaveFigure(my.plot = p27, name = "SCISSORS_Clusters_Stroma_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p28a, name = "SCISSORS_Clusters_CAF_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p28b, name = "SCISSORS_Clusters_CAF_apCAF_Scores")
SaveFigure(my.plot = p28f, name = "SCISSORS_Clusters_CAF_VAM_Violins")
ggsave(plot = p28f, path = "~/Desktop/R/SCISSORS/vignettes/figures_supp/Elyada", device = "png", 
       units = "in", filename = "SCISSORS_Clusters_CAF_VAM_Violins.png", width = 8, height = 8)
SaveFigure(my.plot = p29, name = "All_Cells_NKT_CD3D_FitSNE")
SaveFigure(my.plot = p30, name = "All_Cells_Highlight_NKT_FitSNE")
SaveFigure(my.plot = p31, name = "SCISSORS_Clusters_NKT_Tumor_FitSNE")
SaveFigure(my.plot = p32, name = "SCISSORS_Clusters_NKT_Tumor_CD4T_IL7R")
SaveFigure(my.plot = p33, name = "SCISSORS_Clusters_NKT_Tumor_CD4T_CD69")
SaveFigure(my.plot = p34, name = "SCISSORS_Clusters_NKT_Tumor_Treg_IL2RA")
SaveFigure(my.plot = p35, name = "SCISSORS_Clusters_NKT_Tumor_Treg_FOXP3")
SaveFigure(my.plot = p36, name = "SCISSORS_Clusters_NKT_Tumor_Prolif_Treg_TOP2A")
SaveFigure(my.plot = p37, name = "SCISSORS_Clusters_NKT_Tumor_Mast_TPSAB1")
SaveFigure(my.plot = p38, name = "SCISSORS_Clusters_NKT_Tumor_NK_NKG7")
SaveFigure(my.plot = p39, name = "SCISSORS_Clusters_NKT_Tumor_NK_PRF1")
SaveFigure(my.plot = p40, name = "SCISSORS_Clusters_NKT_Tumor_CD8T_CD8A")
SaveFigure(my.plot = p41, name = "SCISSORS_Clusters_NKT_Tumor_CD8T_CD2")
SaveFigure(my.plot = p42, name = "SCISSORS_Clusters_NKT_Tumor_Intermediate_Mono_LYZ")
SaveFigure(my.plot = p43, name = "SCISSORS_Clusters_NKT_Tumor_Intermediate_Mono_HLADRA")
SaveFigure(my.plot = p44, name = "SCISSORS_Clusters_NKT_Tumor_Intermediate_Mono_CD74")
SaveFigure(my.plot = p45, name = "SCISSORS_Clusters_NKT_Tumor_Intermediate_Mono_HLADPB1")
SaveFigure(my.plot = p46, name = "SCISSORS_Clusters_NKT_Tumor_Labels_FitSNE")
SaveFigure(my.plot = p47, name = "SCISSORS_Clusters_NKT_Tumor_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p48, name = "SCISSORS_Clusters_NKT_AdjNorm_FitSNE")
SaveFigure(my.plot = p49, name = "SCISSORS_Clusters_NKT_AdjNorm_CD4T_IL7R")
SaveFigure(my.plot = p50, name = "SCISSORS_Clusters_NKT_AdjNorm_CD4T_Memory_S100A4")
SaveFigure(my.plot = p51, name = "SCISSORS_Clusters_NKT_AdjNorm_CD4T_Naive_CCR7")
SaveFigure(my.plot = p52, name = "SCISSORS_Clusters_NKT_AdjNorm_CD8T_CD8A")
SaveFigure(my.plot = p53, name = "SCISSORS_Clusters_NKT_AdjNorm_Treg_TIGIT")
SaveFigure(my.plot = p54, name = "SCISSORS_Clusters_NKT_AdjNorm_Treg_FOXP3")
SaveFigure(my.plot = p55, name = "SCISSORS_Clusters_NKT_AdjNorm_NK_PRF1")
SaveFigure(my.plot = p56, name = "SCISSORS_Clusters_NKT_AdjNorm_NK_NKG7")
SaveFigure(my.plot = p57, name = "SCISSORS_Clusters_NKT_AdjNorm_Prolif_Treg_TOP2A")
SaveFigure(my.plot = p58, name = "SCISSORS_Clusters_NKT_AdjNorm_Labels_FitSNE")
SaveFigure(my.plot = p59, name = "SCISSORS_Clusters_NKT_AdjNorm_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p60, name = "All_Cells_Ductal_KRT8_FitSNE")
SaveFigure(my.plot = p61, name = "All_Cells_Highlight_Ductal_FitSNE")
SaveFigure(my.plot = p62, name = "SCISSORS_Clusters_Ductal_FitSNE")
SaveFigure(my.plot = p63, name = "SCISSORS_Clusters_Ductal_VAM_Classical25_FitSNE")
SaveFigure(my.plot = p64, name = "SCISSORS_Clusters_Ductal_VAM_Basal25_FitSNE")
SaveFigure(my.plot = p65, name = "SCISSORS_Clusters_Ductal_Lipid_Proc_ANPEP")
SaveFigure(my.plot = p66, name = "SCISSORS_Clusters_Ductal_Lipid_Proc_FABP1")
SaveFigure(my.plot = p67, name = "SCISSORS_Clusters_Ductal_Secretory_SOD3")
SaveFigure(my.plot = p68, name = "SCISSORS_Clusters_Ductal_Secretory_CFTR")
SaveFigure(my.plot = p69, name = "SCISSORS_Clusters_Ductal_Classical1_TFF1")
SaveFigure(my.plot = p70, name = "SCISSORS_Clusters_Ductal_Classical1_TFF2")
SaveFigure(my.plot = p71, name = "SCISSORS_Clusters_Ductal_SampleID")
SaveFigure(my.plot = p72, name = "SCISSORS_Clusters_Ductal_Classical2_CRISP3")
SaveFigure(my.plot = p73, name = "SCISSORS_Clusters_Ductal_Classical2_GATA6")
SaveFigure(my.plot = p74, name = "SCISSORS_Clusters_Ductal_DECODER_Basal_PDAC_FitSNE")
SaveFigure(my.plot = p75, name = "SCISSORS_Clusters_Ductal_DECODER_Classical_PDAC_FitSNE")
SaveFigure(my.plot = p76, name = "SCISSORS_Clusters_Ductal_DECODER_BC_Ratio_FitSNE")
SaveFigure(my.plot = p77, name = "SCISSORS_Clusters_Ductal_CONICSmat_Annos_FitSNE")
SaveFigure(my.plot = p78, name = "SCISSORS_Clusters_Ductal_Acinar_CTRB2")
SaveFigure(my.plot = p79, name = "SCISSORS_Clusters_Ductal_Labels_FitSNE")
SaveFigure(my.plot = p80, name = "SCISSORS_Clusters_Ductal_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p80b, name = "SCISSORS_Clusters_Ductal_PDAC_Violin_GATA6")
SaveFigure(my.plot = p80h, name = "SCISSORS_Clusters_Ductal_PDAC_BC_Violins_Combined", height = 7, width = 33)
ggsave(plot = p80h, path = "~/Desktop/R/SCISSORS/vignettes/figures_supp/Elyada", device = "png", 
       units = "in", filename = "SCISSORS_Clusters_Ductal_PDAC_BC_Violins_Combined.png", width = 15, height = 5)
SaveFigure(my.plot = p80i, name = "SCISSORS_Clusters_Ductal_CONICSmat_Annos_FitSNE")
SaveFigure(my.plot = p81, name = "All_Cells_Plasma_JCHAIN_FitSNE")
SaveFigure(my.plot = p82, name = "All_Cells_Plasmacytoid_DC_IRF7_FitSNE")
SaveFigure(my.plot = p83, name = "All_Cells_B_MS4A1_FitSNE")
SaveFigure(my.plot = p84, name = "All_Cells_B_CD79A_FitSNE")
SaveFigure(my.plot = p85, name = "All_Cells_Tissue_Type_FitSNE")
SaveFigure(my.plot = p86a, name = "All_Cells_Highlight_Myeloid_FitSNE")
SaveFigure(my.plot = p86b, name = "All_Cells_Highlight_DC_FitSNE")
SaveFigure(my.plot = p87, name = "All_Cells_Myeloid_Tumor_Subset_FitSNE")
SaveFigure(my.plot = p88, name = "All_Cells_Myeloid_Tumor_Subset_CD14_FitSNE")
SaveFigure(my.plot = p89, name = "All_Cells_Myeloid_Tumor_Subset_C1QA_FitSNE")
SaveFigure(my.plot = p90, name = "All_Cells_Myeloid_Tumor_Subset_SPP1_FitSNE")
SaveFigure(my.plot = p91, name = "All_Cells_Myeloid_Tumor_Subset_LYZ_FitSNE")
SaveFigure(my.plot = p92, name = "All_Cells_Myeloid_Tumor_Subset_S100A8_FitSNE")
SaveFigure(my.plot = p93, name = "All_Cells_Myeloid_Tumor_Subset_FCER1A_FitSNE")
SaveFigure(my.plot = p94, name = "SCISSORS_Clusters_Myeloid_Tumor_FitSNE")
SaveFigure(my.plot = p95, name = "SCISSORS_Clusters_DC_Tumor_FitSNE")
SaveFigure(my.plot = p96, name = "SCISSORS_Clusters_Myeloid_Tumor_Neutrophil_S100A8_FitSNE")
SaveFigure(my.plot = p97, name = "SCISSORS_Clusters_Myeloid_Tumor_Neutrophil_S100A9_FitSNE")
SaveFigure(my.plot = p98, name = "SCISSORS_Clusters_Myeloid_Tumor_Classical_Mono_CD14_FitSNE")
SaveFigure(my.plot = p99, name = "SCISSORS_Clusters_Myeloid_Tumor_CD16_Mono_FCGR3A_FitSNE")
SaveFigure(my.plot = p100, name = "SCISSORS_Clusters_Myeloid_Tumor_CD16_Mono_MS4A7_FitSNE")
SaveFigure(my.plot = p101, name = "SCISSORS_Clusters_Myeloid_Tumor_Intermediate_Mono_MS4A7_FitSNE")
SaveFigure(my.plot = p102, name = "SCISSORS_Clusters_Myeloid_Tumor_Resident_Macro_C1QA_FitSNE")
SaveFigure(my.plot = p103, name = "SCISSORS_Clusters_Myeloid_Tumor_Alt_Active_Macro_APOE_FitSNE")
SaveFigure(my.plot = p104, name = "SCISSORS_Clusters_Myeloid_Tumor_Alt_Active_Macro_SPP1_FitSNE")
SaveFigure(my.plot = p105, name = "SCISSORS_Clusters_DC_Tumor_cDC1_CLEC9A_FitSNE")
SaveFigure(my.plot = p106, name = "SCISSORS_Clusters_DC_Tumor_Langerhans_DC_CD1A_FitSNE")
SaveFigure(my.plot = p107, name = "SCISSORS_Clusters_DC_Tumor_Langerhans_DC_CD207_FitSNE")
SaveFigure(my.plot = p108, name = "SCISSORS_Clusters_DC_Tumor_Active_DC_LAMP3_FitSNE")
SaveFigure(my.plot = p109, name = "SCISSORS_Clusters_DC_Tumor_cDC2_CLEC10A_FitSNE")
SaveFigure(my.plot = p110, name = "SCISSORS_Clusters_DC_Tumor_cDC2_KLF4_FitSNE")
SaveFigure(my.plot = p111, name = "SCISSORS_Clusters_DC_Tumor_SampleID_FitSNE")
SaveFigure(my.plot = p112, name = "SCISSORS_Clusters_Myeloid_Tumor_Labels_FitSNE")
SaveFigure(my.plot = p113, name = "SCISSORS_Clusters_Myeloid_Tumor_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p114, name = "SCISSORS_Clusters_DC_Tumor_Labels_FitSNE")
SaveFigure(my.plot = p115, name = "SCISSORS_Clusters_DC_Tumor_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p116, name = "SCISSORS_Clusters_Myeloid_AdjNorm_FitSNE")
SaveFigure(my.plot = p117, name = "SCISSORS_Clusters_DC_AdjNorm_FitSNE")
SaveFigure(my.plot = p118, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Neutrophil_S100A8_FitSNE")
SaveFigure(my.plot = p119, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Neutrophil_S100A9_FitSNE")
SaveFigure(my.plot = p120, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Classical_Mono_LYZ_FitSNE")
SaveFigure(my.plot = p121, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Classical_Mono_DC14_FitSNE")
SaveFigure(my.plot = p122, name = "SCISSORS_Clusters_Myeloid_AdjNorm_SampleID_FitSNE")
SaveFigure(my.plot = p123, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Resident_Macro_C1QA_FitSNE")
SaveFigure(my.plot = p124, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Resident_Macro_APOE_FitSNE")
SaveFigure(my.plot = p125, name = "SCISSORS_Clusters_Myeloid_AdjNorm_CD8T_CD3D_FitSNE")
SaveFigure(my.plot = p126, name = "SCISSORS_Clusters_Myeloid_AdjNorm_CD8T_CD8A_FitSNE")
SaveFigure(my.plot = p127, name = "SCISSORS_Clusters_Myeloid_AdjNorm_CD8T_NKG7_FitSNE")
SaveFigure(my.plot = p128, name = "SCISSORS_Clusters_DC_AdjNorm_cDC1_CLEC9A_FitSNE")
SaveFigure(my.plot = p129, name = "SCISSORS_Clusters_DC_AdjNorm_cDC2_CLEC10A_FitSNE")
SaveFigure(my.plot = p130, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Labels_FitSNE")
SaveFigure(my.plot = p131, name = "SCISSORS_Clusters_Myeloid_AdjNorm_Dotplot", height = 6, width = 12)
SaveFigure(my.plot = p132, name = "SCISSORS_Clusters_DC_AdjNorm_Labels_FitSNE")
SaveFigure(my.plot = p133, name = "SCISSORS_Clusters_DC_AdjNorm_Dotplot", height = 6, width = 12)
```

And of course:

```{r}
sessionInfo()
```
